{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "c5b9baad",
   "metadata": {},
   "outputs": [],
   "source": [
    "# sample reader to read samples from /data/runs/samples\n",
    "# and build each sample into a matrix\n",
    "\n",
    "# Isaac Berez\n",
    "# 14.01.23\n",
    "\n",
    "from scipy.io import mmread\n",
    "import os\n",
    "import glob\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "from pandas_ods_reader import read_ods\n",
    "from copy import deepcopy\n",
    "import pprint"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d84b85e4",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "c05fc800",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Serial_Number:int</th>\n",
       "      <th>Date_Captured:string</th>\n",
       "      <th>Species:string</th>\n",
       "      <th>Transcriptome:string</th>\n",
       "      <th>Strain:string</th>\n",
       "      <th>Project:string</th>\n",
       "      <th>Group</th>\n",
       "      <th>ChipID:string</th>\n",
       "      <th>SampleID:string</th>\n",
       "      <th>DonorID:string</th>\n",
       "      <th>...</th>\n",
       "      <th>Comments:string</th>\n",
       "      <th>cDNA_Lib_Ok:string</th>\n",
       "      <th>ngperul_cDNA:float64</th>\n",
       "      <th>Avesizebp_cDNAlib</th>\n",
       "      <th>Date</th>\n",
       "      <th>cDNAul</th>\n",
       "      <th>LIbConstructionComment</th>\n",
       "      <th>ngperul_seqlib:float64</th>\n",
       "      <th>lengthbp_seqlib:float64</th>\n",
       "      <th>Sample_Index:string</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>106.0</td>\n",
       "      <td>43993.0</td>\n",
       "      <td>Mm</td>\n",
       "      <td>Mm10</td>\n",
       "      <td>Cntnp-C57Bl/6</td>\n",
       "      <td>Cntnp_KO</td>\n",
       "      <td>Cntnp-KO-F</td>\n",
       "      <td>10X54</td>\n",
       "      <td>10X54_1</td>\n",
       "      <td>Cntnp-KO_2-9F</td>\n",
       "      <td>...</td>\n",
       "      <td>Cntnp-KO female, showing Cntnp \"male\" ASD emot...</td>\n",
       "      <td>None</td>\n",
       "      <td>24.7</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>17.9</td>\n",
       "      <td>473.0</td>\n",
       "      <td>B3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>107.0</td>\n",
       "      <td>43993.0</td>\n",
       "      <td>Mm</td>\n",
       "      <td>Mm10</td>\n",
       "      <td>Cntnp-C57Bl/6</td>\n",
       "      <td>Cntnp_KO</td>\n",
       "      <td>Cntnp-KO-F</td>\n",
       "      <td>10X54</td>\n",
       "      <td>10X54_2</td>\n",
       "      <td>Cntnp-KO_4-10F</td>\n",
       "      <td>...</td>\n",
       "      <td>Cntnp-KO female, showing Cntnp \"male\" ASD emot...</td>\n",
       "      <td>None</td>\n",
       "      <td>37.7</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>20.0</td>\n",
       "      <td>471.0</td>\n",
       "      <td>D3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>212.0</td>\n",
       "      <td>2021-03-22</td>\n",
       "      <td>Mm</td>\n",
       "      <td>Mm10</td>\n",
       "      <td>Cntnp-C57Bl/6</td>\n",
       "      <td>Cntnp_KO</td>\n",
       "      <td>Cntnp-KO-M</td>\n",
       "      <td>10x98</td>\n",
       "      <td>10x98_2</td>\n",
       "      <td>Cntnp-KO_2_M</td>\n",
       "      <td>...</td>\n",
       "      <td>v3! Cntnap KO #2, M; behavioral testing previo...</td>\n",
       "      <td>None</td>\n",
       "      <td>39.7</td>\n",
       "      <td>1873.5</td>\n",
       "      <td>2021-04-04</td>\n",
       "      <td>5.0</td>\n",
       "      <td>Single index; scale down; PCR_14 cycles</td>\n",
       "      <td>17.4</td>\n",
       "      <td>479.2</td>\n",
       "      <td>H9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>213.0</td>\n",
       "      <td>2021-03-22</td>\n",
       "      <td>Mm</td>\n",
       "      <td>Mm10</td>\n",
       "      <td>Cntnp-C57Bl/6</td>\n",
       "      <td>Cntnp_KO</td>\n",
       "      <td>Cntnp-KO-M</td>\n",
       "      <td>10x98</td>\n",
       "      <td>10x98_3</td>\n",
       "      <td>Cntnp-KO_3_M</td>\n",
       "      <td>...</td>\n",
       "      <td>v3! Cntnap KO #3, M; behavioral testing previo...</td>\n",
       "      <td>None</td>\n",
       "      <td>45.5</td>\n",
       "      <td>1841.7</td>\n",
       "      <td>2021-04-04</td>\n",
       "      <td>5.0</td>\n",
       "      <td>Single index; scale down; PCR_14 cycles</td>\n",
       "      <td>21.9</td>\n",
       "      <td>426.4</td>\n",
       "      <td>A10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>98.0</td>\n",
       "      <td>05/31/20</td>\n",
       "      <td>Mm</td>\n",
       "      <td>Mm10</td>\n",
       "      <td>C57Bl/6</td>\n",
       "      <td>Dimorph</td>\n",
       "      <td>Breeder-F</td>\n",
       "      <td>10X51</td>\n",
       "      <td>10X51_2</td>\n",
       "      <td>DI-B1-F</td>\n",
       "      <td>...</td>\n",
       "      <td>Breeders B1, 8pups DOB 27.5 (P4). perfusion aC...</td>\n",
       "      <td>None</td>\n",
       "      <td>29.8</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>22.2</td>\n",
       "      <td>474.0</td>\n",
       "      <td>A2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 27 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  Serial_Number:int Date_Captured:string Species:string Transcriptome:string  \\\n",
       "0             106.0              43993.0             Mm                 Mm10   \n",
       "1             107.0              43993.0             Mm                 Mm10   \n",
       "2             212.0           2021-03-22             Mm                 Mm10   \n",
       "3             213.0           2021-03-22             Mm                 Mm10   \n",
       "4              98.0             05/31/20             Mm                 Mm10   \n",
       "\n",
       "   Strain:string Project:string       Group ChipID:string SampleID:string  \\\n",
       "0  Cntnp-C57Bl/6       Cntnp_KO  Cntnp-KO-F         10X54         10X54_1   \n",
       "1  Cntnp-C57Bl/6       Cntnp_KO  Cntnp-KO-F         10X54         10X54_2   \n",
       "2  Cntnp-C57Bl/6       Cntnp_KO  Cntnp-KO-M         10x98         10x98_2   \n",
       "3  Cntnp-C57Bl/6       Cntnp_KO  Cntnp-KO-M         10x98         10x98_3   \n",
       "4        C57Bl/6        Dimorph   Breeder-F         10X51         10X51_2   \n",
       "\n",
       "   DonorID:string  ...                                    Comments:string  \\\n",
       "0   Cntnp-KO_2-9F  ...  Cntnp-KO female, showing Cntnp \"male\" ASD emot...   \n",
       "1  Cntnp-KO_4-10F  ...  Cntnp-KO female, showing Cntnp \"male\" ASD emot...   \n",
       "2    Cntnp-KO_2_M  ...  v3! Cntnap KO #2, M; behavioral testing previo...   \n",
       "3    Cntnp-KO_3_M  ...  v3! Cntnap KO #3, M; behavioral testing previo...   \n",
       "4         DI-B1-F  ...  Breeders B1, 8pups DOB 27.5 (P4). perfusion aC...   \n",
       "\n",
       "   cDNA_Lib_Ok:string ngperul_cDNA:float64 Avesizebp_cDNAlib        Date  \\\n",
       "0                None                 24.7               NaN        None   \n",
       "1                None                 37.7               NaN        None   \n",
       "2                None                 39.7            1873.5  2021-04-04   \n",
       "3                None                 45.5            1841.7  2021-04-04   \n",
       "4                None                 29.8               NaN        None   \n",
       "\n",
       "   cDNAul                   LIbConstructionComment ngperul_seqlib:float64  \\\n",
       "0     NaN                                     None                   17.9   \n",
       "1     NaN                                     None                   20.0   \n",
       "2     5.0  Single index; scale down; PCR_14 cycles                   17.4   \n",
       "3     5.0  Single index; scale down; PCR_14 cycles                   21.9   \n",
       "4     NaN                                     None                   22.2   \n",
       "\n",
       "  lengthbp_seqlib:float64  Sample_Index:string  \n",
       "0                   473.0                   B3  \n",
       "1                   471.0                   D3  \n",
       "2                   479.2                   H9  \n",
       "3                   426.4                  A10  \n",
       "4                   474.0                   A2  \n",
       "\n",
       "[5 rows x 27 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#get meta data from ods file, only will need sample id, sex, group\n",
    "analysis_dir = '/home/isaac/analysis/scRNA-seq/'\n",
    "dimorph_sample_file = 'Dimorph_samples.ods'\n",
    "dimorph_df = read_ods(analysis_dir+dimorph_sample_file)\n",
    "dimorph_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "d185fba1",
   "metadata": {},
   "outputs": [],
   "source": [
    "#fill in missing data types\n",
    "dimorph_df_1 = dimorph_df.rename(columns={\"Group\":\"Group:string\", \n",
    "                                          \"Avesizebp_cDNAlib\":\"Avesizebp_cDNAlib:float64\",\n",
    "                                          \"Date\":\"Date:string\",\n",
    "                                          \"cDNAul\":\"cDNAul:float64\",\n",
    "                                          \"LIbConstructionComment\": \"LIbConstructionComment:string\"\n",
    "                                         })\n",
    "#list(dimorph_df_1.columns)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "05cf937e",
   "metadata": {},
   "outputs": [],
   "source": [
    "#create lists of meta data header and data types by splitting at colon in each column name\n",
    "mdata_header = []\n",
    "dtype = []\n",
    "for d in dimorph_df_1:\n",
    "    mdata_header.append(d.split(\":\")[0])\n",
    "    dtype.append(d.split(\":\")[1])\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "a16cb28f",
   "metadata": {},
   "outputs": [],
   "source": [
    "#get and return sample id\n",
    "samples = [x for x in list(dimorph_df['SampleID:string']) if x != None]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "a0587247",
   "metadata": {},
   "outputs": [],
   "source": [
    "#create meta data dictionary from sample ids\n",
    "meta_data_dict = dict.fromkeys(samples)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "25ed67d0",
   "metadata": {},
   "outputs": [],
   "source": [
    "#initialize headers for each sample\n",
    "mdata_header_dict = dict.fromkeys(mdata_header)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "3945034a",
   "metadata": {},
   "outputs": [],
   "source": [
    "for i in range(len(meta_data_dict.keys())):\n",
    "    for key,value in zip(mdata_header_dict.keys(),list(dimorph_df_1.loc[i])):\n",
    "        mdata_header_dict.update({key:value})\n",
    "    meta_data_dict.update({list(meta_data_dict.keys())[i]:deepcopy(mdata_header_dict)})\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "188c28f7",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'10X35_1': {'Age': '8w',\n",
      "             'Avesizebp_cDNAlib': nan,\n",
      "             'Cell_Conc': 1000.0,\n",
      "             'ChipID': '10X35',\n",
      "             'Comments': 'Naive 8wo females, pooled, left hemisphere. 5%opti.',\n",
      "             'Date': None,\n",
      "             'Date_Captured': '2019-11-20',\n",
      "             'DonorID': 'DI1,DI2',\n",
      "             'Group': 'Naïve-F',\n",
      "             'LIbConstructionComment': None,\n",
      "             'Num_Pooled_Animals': 2.0,\n",
      "             'PCR_Cycles': 13.0,\n",
      "             'Project': 'Dimorph',\n",
      "             'SampleID': '10X35_1',\n",
      "             'Sample_Index': 'D9',\n",
      "             'Serial_Number': 63.0,\n",
      "             'Sex': 'F',\n",
      "             'Species': 'Mm',\n",
      "             'Strain': None,\n",
      "             'Target_Num_Cells': 5000.0,\n",
      "             'Tissue': 'MeA_L',\n",
      "             'Transcriptome': 'Mm10',\n",
      "             'cDNA_Lib_Ok': 'Y',\n",
      "             'cDNAul': nan,\n",
      "             'lengthbp_seqlib': 416.0,\n",
      "             'ngperul_cDNA': 23.3,\n",
      "             'ngperul_seqlib': 13.6},\n",
      " '10X35_2': {'Age': '8w',\n",
      "             'Avesizebp_cDNAlib': nan,\n",
      "             'Cell_Conc': 1000.0,\n",
      "             'ChipID': '10X35',\n",
      "             'Comments': 'Naive 8wo females, pooled, right hemisphere. 5%opti.',\n",
      "             'Date': None,\n",
      "             'Date_Captured': '2019-11-20',\n",
      "             'DonorID': 'DI1,DI2',\n",
      "             'Group': 'Naïve-F',\n",
      "             'LIbConstructionComment': None,\n",
      "             'Num_Pooled_Animals': 2.0,\n",
      "             'PCR_Cycles': 13.0,\n",
      "             'Project': 'Dimorph',\n",
      "             'SampleID': '10X35_2',\n",
      "             'Sample_Index': 'G8',\n",
      "             'Serial_Number': 64.0,\n",
      "             'Sex': 'F',\n",
      "             'Species': 'Mm',\n",
      "             'Strain': None,\n",
      "             'Target_Num_Cells': 5000.0,\n",
      "             'Tissue': 'MeA_R',\n",
      "             'Transcriptome': 'Mm10',\n",
      "             'cDNA_Lib_Ok': 'Y',\n",
      "             'cDNAul': nan,\n",
      "             'lengthbp_seqlib': 461.0,\n",
      "             'ngperul_cDNA': 26.1,\n",
      "             'ngperul_seqlib': 17.6},\n",
      " '10X36_1': {'Age': '8.5w',\n",
      "             'Avesizebp_cDNAlib': nan,\n",
      "             'Cell_Conc': 1000.0,\n",
      "             'ChipID': '10X36',\n",
      "             'Comments': 'Naive 8.5wo male, left hemisphere. 5%opti.',\n",
      "             'Date': None,\n",
      "             'Date_Captured': '2019-11-24',\n",
      "             'DonorID': 'DI3',\n",
      "             'Group': 'Naïve-M',\n",
      "             'LIbConstructionComment': None,\n",
      "             'Num_Pooled_Animals': 1.0,\n",
      "             'PCR_Cycles': 13.0,\n",
      "             'Project': 'Dimorph',\n",
      "             'SampleID': '10X36_1',\n",
      "             'Sample_Index': 'H8',\n",
      "             'Serial_Number': 65.0,\n",
      "             'Sex': 'M',\n",
      "             'Species': 'Mm',\n",
      "             'Strain': None,\n",
      "             'Target_Num_Cells': 5000.0,\n",
      "             'Tissue': 'MeA_L',\n",
      "             'Transcriptome': 'Mm10',\n",
      "             'cDNA_Lib_Ok': 'Y',\n",
      "             'cDNAul': nan,\n",
      "             'lengthbp_seqlib': 476.0,\n",
      "             'ngperul_cDNA': 26.4,\n",
      "             'ngperul_seqlib': 11.4},\n",
      " '10X36_2': {'Age': '8.5w',\n",
      "             'Avesizebp_cDNAlib': nan,\n",
      "             'Cell_Conc': 1000.0,\n",
      "             'ChipID': '10X36',\n",
      "             'Comments': 'Naive 8.5wo male, right hemisphere. 5%opti.',\n",
      "             'Date': None,\n",
      "             'Date_Captured': '2019-11-24',\n",
      "             'DonorID': 'DI3',\n",
      "             'Group': 'Naïve-M',\n",
      "             'LIbConstructionComment': None,\n",
      "             'Num_Pooled_Animals': 1.0,\n",
      "             'PCR_Cycles': 13.0,\n",
      "             'Project': 'Dimorph',\n",
      "             'SampleID': '10X36_2',\n",
      "             'Sample_Index': 'A9',\n",
      "             'Serial_Number': 66.0,\n",
      "             'Sex': 'M',\n",
      "             'Species': 'Mm',\n",
      "             'Strain': None,\n",
      "             'Target_Num_Cells': 5000.0,\n",
      "             'Tissue': 'MeA_R',\n",
      "             'Transcriptome': 'Mm10',\n",
      "             'cDNA_Lib_Ok': 'Y',\n",
      "             'cDNAul': nan,\n",
      "             'lengthbp_seqlib': 486.0,\n",
      "             'ngperul_cDNA': 16.7,\n",
      "             'ngperul_seqlib': 12.1},\n",
      " '10X37_1': {'Age': '8.5w',\n",
      "             'Avesizebp_cDNAlib': nan,\n",
      "             'Cell_Conc': 1000.0,\n",
      "             'ChipID': '10X37',\n",
      "             'Comments': 'Naive 8.5wo males, pooled, left hemisphere. 5%opti.',\n",
      "             'Date': None,\n",
      "             'Date_Captured': '2019-11-25',\n",
      "             'DonorID': 'DI4,DI5',\n",
      "             'Group': 'Naïve-M',\n",
      "             'LIbConstructionComment': None,\n",
      "             'Num_Pooled_Animals': 2.0,\n",
      "             'PCR_Cycles': 13.0,\n",
      "             'Project': 'Dimorph',\n",
      "             'SampleID': '10X37_1',\n",
      "             'Sample_Index': 'E9',\n",
      "             'Serial_Number': 67.0,\n",
      "             'Sex': 'M',\n",
      "             'Species': 'Mm',\n",
      "             'Strain': None,\n",
      "             'Target_Num_Cells': 5000.0,\n",
      "             'Tissue': 'MeA_L',\n",
      "             'Transcriptome': 'Mm10',\n",
      "             'cDNA_Lib_Ok': 'Y',\n",
      "             'cDNAul': nan,\n",
      "             'lengthbp_seqlib': 440.0,\n",
      "             'ngperul_cDNA': 17.0,\n",
      "             'ngperul_seqlib': 15.9},\n",
      " '10X37_2': {'Age': '8.5w',\n",
      "             'Avesizebp_cDNAlib': nan,\n",
      "             'Cell_Conc': 1000.0,\n",
      "             'ChipID': '10X37',\n",
      "             'Comments': 'Naive 8.5wo males, pooled, right hemisphere. 5%opti.',\n",
      "             'Date': None,\n",
      "             'Date_Captured': '2019-11-25',\n",
      "             'DonorID': 'DI4,DI5',\n",
      "             'Group': 'Naïve-M',\n",
      "             'LIbConstructionComment': None,\n",
      "             'Num_Pooled_Animals': 2.0,\n",
      "             'PCR_Cycles': 13.0,\n",
      "             'Project': 'Dimorph',\n",
      "             'SampleID': '10X37_2',\n",
      "             'Sample_Index': 'F9',\n",
      "             'Serial_Number': 68.0,\n",
      "             'Sex': 'M',\n",
      "             'Species': 'Mm',\n",
      "             'Strain': None,\n",
      "             'Target_Num_Cells': 5000.0,\n",
      "             'Tissue': 'MeA_R',\n",
      "             'Transcriptome': 'Mm10',\n",
      "             'cDNA_Lib_Ok': 'Y',\n",
      "             'cDNAul': nan,\n",
      "             'lengthbp_seqlib': 460.0,\n",
      "             'ngperul_cDNA': 24.4,\n",
      "             'ngperul_seqlib': 16.0},\n",
      " '10X38_1': {'Age': '9w',\n",
      "             'Avesizebp_cDNAlib': 1870.0,\n",
      "             'Cell_Conc': 1000.0,\n",
      "             'ChipID': '10X38',\n",
      "             'Comments': 'Naive 9wo female, left hemisphere. 5%opti.',\n",
      "             'Date': None,\n",
      "             'Date_Captured': '2019-12-01',\n",
      "             'DonorID': 'DI6',\n",
      "             'Group': 'Naïve-F',\n",
      "             'LIbConstructionComment': None,\n",
      "             'Num_Pooled_Animals': 1.0,\n",
      "             'PCR_Cycles': 13.0,\n",
      "             'Project': 'Dimorph',\n",
      "             'SampleID': '10X38_1',\n",
      "             'Sample_Index': 'B9',\n",
      "             'Serial_Number': 69.0,\n",
      "             'Sex': 'F',\n",
      "             'Species': 'Mm',\n",
      "             'Strain': None,\n",
      "             'Target_Num_Cells': 5000.0,\n",
      "             'Tissue': 'MeA_L',\n",
      "             'Transcriptome': 'Mm10',\n",
      "             'cDNA_Lib_Ok': 'Y',\n",
      "             'cDNAul': nan,\n",
      "             'lengthbp_seqlib': 420.0,\n",
      "             'ngperul_cDNA': 21.9,\n",
      "             'ngperul_seqlib': 15.0},\n",
      " '10X38_2': {'Age': '9w',\n",
      "             'Avesizebp_cDNAlib': 1700.0,\n",
      "             'Cell_Conc': 800.0,\n",
      "             'ChipID': '10X38',\n",
      "             'Comments': 'Naive 9wo female, right hemisphere. 1-2 sections '\n",
      "                         'missing, compared to L. 5%opti.',\n",
      "             'Date': None,\n",
      "             'Date_Captured': '2019-12-01',\n",
      "             'DonorID': 'DI6',\n",
      "             'Group': 'Naïve-F',\n",
      "             'LIbConstructionComment': None,\n",
      "             'Num_Pooled_Animals': 1.0,\n",
      "             'PCR_Cycles': 13.0,\n",
      "             'Project': 'Dimorph',\n",
      "             'SampleID': '10X38_2',\n",
      "             'Sample_Index': 'C9',\n",
      "             'Serial_Number': 70.0,\n",
      "             'Sex': 'F',\n",
      "             'Species': 'Mm',\n",
      "             'Strain': None,\n",
      "             'Target_Num_Cells': 5000.0,\n",
      "             'Tissue': 'MeA_R',\n",
      "             'Transcriptome': 'Mm10',\n",
      "             'cDNA_Lib_Ok': 'Y',\n",
      "             'cDNAul': nan,\n",
      "             'lengthbp_seqlib': 415.0,\n",
      "             'ngperul_cDNA': 6.86,\n",
      "             'ngperul_seqlib': 18.7},\n",
      " '10X51_1': {'Age': '11-12w',\n",
      "             'Avesizebp_cDNAlib': nan,\n",
      "             'Cell_Conc': 1000.0,\n",
      "             'ChipID': '10X51',\n",
      "             'Comments': 'Breeders B1, 8pups DOB 27.5 (P4). perfusion aCSF, '\n",
      "                         'MeA, 5% Optiprep clean up, good suspension, good '\n",
      "                         'emulsion',\n",
      "             'Date': None,\n",
      "             'Date_Captured': '05/31/20',\n",
      "             'DonorID': 'DI-B1-F',\n",
      "             'Group': 'Breeder-F',\n",
      "             'LIbConstructionComment': None,\n",
      "             'Num_Pooled_Animals': 1.0,\n",
      "             'PCR_Cycles': 13.0,\n",
      "             'Project': 'Dimorph',\n",
      "             'SampleID': '10X51_1',\n",
      "             'Sample_Index': 'G1',\n",
      "             'Serial_Number': 97.0,\n",
      "             'Sex': 'F',\n",
      "             'Species': 'Mm',\n",
      "             'Strain': 'C57Bl/6',\n",
      "             'Target_Num_Cells': 5000.0,\n",
      "             'Tissue': 'MeA',\n",
      "             'Transcriptome': 'Mm10',\n",
      "             'cDNA_Lib_Ok': None,\n",
      "             'cDNAul': nan,\n",
      "             'lengthbp_seqlib': 478.0,\n",
      "             'ngperul_cDNA': 32.5,\n",
      "             'ngperul_seqlib': 19.2},\n",
      " '10X51_2': {'Age': '11-12w',\n",
      "             'Avesizebp_cDNAlib': nan,\n",
      "             'Cell_Conc': 1000.0,\n",
      "             'ChipID': '10X51',\n",
      "             'Comments': 'Breeders B1, 8pups DOB 27.5 (P4). perfusion aCSF, '\n",
      "                         'MeA, 5% Optiprep clean up, good suspension, good '\n",
      "                         'emulsion',\n",
      "             'Date': None,\n",
      "             'Date_Captured': '05/31/20',\n",
      "             'DonorID': 'DI-B1-F',\n",
      "             'Group': 'Breeder-F',\n",
      "             'LIbConstructionComment': None,\n",
      "             'Num_Pooled_Animals': 1.0,\n",
      "             'PCR_Cycles': 13.0,\n",
      "             'Project': 'Dimorph',\n",
      "             'SampleID': '10X51_2',\n",
      "             'Sample_Index': 'A2',\n",
      "             'Serial_Number': 98.0,\n",
      "             'Sex': 'F',\n",
      "             'Species': 'Mm',\n",
      "             'Strain': 'C57Bl/6',\n",
      "             'Target_Num_Cells': 5000.0,\n",
      "             'Tissue': 'MeA',\n",
      "             'Transcriptome': 'Mm10',\n",
      "             'cDNA_Lib_Ok': None,\n",
      "             'cDNAul': nan,\n",
      "             'lengthbp_seqlib': 474.0,\n",
      "             'ngperul_cDNA': 29.8,\n",
      "             'ngperul_seqlib': 22.2},\n",
      " '10X51_3': {'Age': '11-12w',\n",
      "             'Avesizebp_cDNAlib': nan,\n",
      "             'Cell_Conc': '900 (but loaded like 1000',\n",
      "             'ChipID': '10X51',\n",
      "             'Comments': 'Breeders B1, 8pups DOB 27.5 (P4). perfusion aCSF, '\n",
      "                         'MeA, 5% Optiprep clean up, good suspension, good '\n",
      "                         'emulsion',\n",
      "             'Date': None,\n",
      "             'Date_Captured': '05/31/20',\n",
      "             'DonorID': 'DI-B1-M',\n",
      "             'Group': 'Breeder-M',\n",
      "             'LIbConstructionComment': None,\n",
      "             'Num_Pooled_Animals': 1.0,\n",
      "             'PCR_Cycles': 13.0,\n",
      "             'Project': 'Dimorph',\n",
      "             'SampleID': '10X51_3',\n",
      "             'Sample_Index': 'B2',\n",
      "             'Serial_Number': 99.0,\n",
      "             'Sex': 'M',\n",
      "             'Species': 'Mm',\n",
      "             'Strain': 'C57Bl/6',\n",
      "             'Target_Num_Cells': 5000.0,\n",
      "             'Tissue': 'MeA',\n",
      "             'Transcriptome': 'Mm10',\n",
      "             'cDNA_Lib_Ok': None,\n",
      "             'cDNAul': nan,\n",
      "             'lengthbp_seqlib': 487.0,\n",
      "             'ngperul_cDNA': 11.7,\n",
      "             'ngperul_seqlib': 22.0},\n",
      " '10X51_4': {'Age': '11-12w',\n",
      "             'Avesizebp_cDNAlib': nan,\n",
      "             'Cell_Conc': '900 (but loaded like 1000',\n",
      "             'ChipID': '10X51',\n",
      "             'Comments': 'Breeders B1, 8pups DOB 27.5 (P4). perfusion aCSF, '\n",
      "                         'MeA, 5% Optiprep clean up, good suspension, good '\n",
      "                         'emulsion',\n",
      "             'Date': None,\n",
      "             'Date_Captured': '05/31/20',\n",
      "             'DonorID': 'DI-B1-M',\n",
      "             'Group': 'Breeder-M',\n",
      "             'LIbConstructionComment': None,\n",
      "             'Num_Pooled_Animals': 1.0,\n",
      "             'PCR_Cycles': 13.0,\n",
      "             'Project': 'Dimorph',\n",
      "             'SampleID': '10X51_4',\n",
      "             'Sample_Index': 'C2',\n",
      "             'Serial_Number': 100.0,\n",
      "             'Sex': 'M',\n",
      "             'Species': 'Mm',\n",
      "             'Strain': 'C57Bl/6',\n",
      "             'Target_Num_Cells': 5000.0,\n",
      "             'Tissue': 'MeA',\n",
      "             'Transcriptome': 'Mm10',\n",
      "             'cDNA_Lib_Ok': None,\n",
      "             'cDNAul': nan,\n",
      "             'lengthbp_seqlib': 415.0,\n",
      "             'ngperul_cDNA': 8.17,\n",
      "             'ngperul_seqlib': 23.6},\n",
      " '10X52_1': {'Age': '11-12w',\n",
      "             'Avesizebp_cDNAlib': nan,\n",
      "             'Cell_Conc': 900.0,\n",
      "             'ChipID': '10X52',\n",
      "             'Comments': 'Breeders B3, 5pups DOB 27.5 (P4, larger than B1). '\n",
      "                         'perfusion aCSF, MeA, 5% Optiprep clean up, good '\n",
      "                         'suspension, good emulsion',\n",
      "             'Date': None,\n",
      "             'Date_Captured': '05/31/20',\n",
      "             'DonorID': 'DI-B3-F',\n",
      "             'Group': 'Breeder-F',\n",
      "             'LIbConstructionComment': None,\n",
      "             'Num_Pooled_Animals': 1.0,\n",
      "             'PCR_Cycles': 13.0,\n",
      "             'Project': 'Dimorph',\n",
      "             'SampleID': '10X52_1',\n",
      "             'Sample_Index': 'D12',\n",
      "             'Serial_Number': 101.0,\n",
      "             'Sex': 'F',\n",
      "             'Species': 'Mm',\n",
      "             'Strain': 'C57Bl/6',\n",
      "             'Target_Num_Cells': 5000.0,\n",
      "             'Tissue': 'MeA',\n",
      "             'Transcriptome': 'Mm10',\n",
      "             'cDNA_Lib_Ok': None,\n",
      "             'cDNAul': nan,\n",
      "             'lengthbp_seqlib': 466.0,\n",
      "             'ngperul_cDNA': 13.7,\n",
      "             'ngperul_seqlib': 15.7},\n",
      " '10X52_2': {'Age': '11-12w',\n",
      "             'Avesizebp_cDNAlib': nan,\n",
      "             'Cell_Conc': 900.0,\n",
      "             'ChipID': '10X52',\n",
      "             'Comments': 'Breeders B3, 5pups DOB 27.5 (P4, larger than B1). '\n",
      "                         'perfusion aCSF, MeA, 5% Optiprep clean up, good '\n",
      "                         'suspension, good emulsion',\n",
      "             'Date': None,\n",
      "             'Date_Captured': '05/31/20',\n",
      "             'DonorID': 'DI-B3-F',\n",
      "             'Group': 'Breeder-F',\n",
      "             'LIbConstructionComment': None,\n",
      "             'Num_Pooled_Animals': 1.0,\n",
      "             'PCR_Cycles': 13.0,\n",
      "             'Project': 'Dimorph',\n",
      "             'SampleID': '10X52_2',\n",
      "             'Sample_Index': 'E12',\n",
      "             'Serial_Number': 102.0,\n",
      "             'Sex': 'F',\n",
      "             'Species': 'Mm',\n",
      "             'Strain': 'C57Bl/6',\n",
      "             'Target_Num_Cells': 5000.0,\n",
      "             'Tissue': 'MeA',\n",
      "             'Transcriptome': 'Mm10',\n",
      "             'cDNA_Lib_Ok': None,\n",
      "             'cDNAul': nan,\n",
      "             'lengthbp_seqlib': 440.0,\n",
      "             'ngperul_cDNA': 11.6,\n",
      "             'ngperul_seqlib': 17.5},\n",
      " '10X52_3': {'Age': '11-12w',\n",
      "             'Avesizebp_cDNAlib': nan,\n",
      "             'Cell_Conc': 900.0,\n",
      "             'ChipID': '10X52',\n",
      "             'Comments': 'Breeders B3, 5pups DOB 27.5 (P4, larger than B1). '\n",
      "                         'perfusion aCSF, MeA, 5% Optiprep clean up, good '\n",
      "                         'suspension, good emulsion',\n",
      "             'Date': None,\n",
      "             'Date_Captured': '05/31/20',\n",
      "             'DonorID': 'DI-B3-M',\n",
      "             'Group': 'Breeder-M',\n",
      "             'LIbConstructionComment': None,\n",
      "             'Num_Pooled_Animals': 1.0,\n",
      "             'PCR_Cycles': 13.0,\n",
      "             'Project': 'Dimorph',\n",
      "             'SampleID': '10X52_3',\n",
      "             'Sample_Index': 'F12',\n",
      "             'Serial_Number': 103.0,\n",
      "             'Sex': 'M',\n",
      "             'Species': 'Mm',\n",
      "             'Strain': 'C57Bl/6',\n",
      "             'Target_Num_Cells': 5000.0,\n",
      "             'Tissue': 'MeA',\n",
      "             'Transcriptome': 'Mm10',\n",
      "             'cDNA_Lib_Ok': None,\n",
      "             'cDNAul': nan,\n",
      "             'lengthbp_seqlib': 507.0,\n",
      "             'ngperul_cDNA': 11.2,\n",
      "             'ngperul_seqlib': '20.4(second prep)'},\n",
      " '10X52_4': {'Age': '11-12w',\n",
      "             'Avesizebp_cDNAlib': nan,\n",
      "             'Cell_Conc': 900.0,\n",
      "             'ChipID': '10X52',\n",
      "             'Comments': 'Breeders B3, 5pups DOB 27.5 (P4, larger than B1). '\n",
      "                         'perfusion aCSF, MeA, 5% Optiprep clean up, good '\n",
      "                         'suspension, good emulsion',\n",
      "             'Date': None,\n",
      "             'Date_Captured': '05/31/20',\n",
      "             'DonorID': 'DI-B3-M',\n",
      "             'Group': 'Breeder-M',\n",
      "             'LIbConstructionComment': None,\n",
      "             'Num_Pooled_Animals': 1.0,\n",
      "             'PCR_Cycles': 13.0,\n",
      "             'Project': 'Dimorph',\n",
      "             'SampleID': '10X52_4',\n",
      "             'Sample_Index': 'G12',\n",
      "             'Serial_Number': 104.0,\n",
      "             'Sex': 'M',\n",
      "             'Species': 'Mm',\n",
      "             'Strain': 'C57Bl/6',\n",
      "             'Target_Num_Cells': 5000.0,\n",
      "             'Tissue': 'MeA',\n",
      "             'Transcriptome': 'Mm10',\n",
      "             'cDNA_Lib_Ok': None,\n",
      "             'cDNAul': nan,\n",
      "             'lengthbp_seqlib': 535.0,\n",
      "             'ngperul_cDNA': 14.7,\n",
      "             'ngperul_seqlib': '8.94 (second prep)'},\n",
      " '10X54_1': {'Age': '9.5w',\n",
      "             'Avesizebp_cDNAlib': nan,\n",
      "             'Cell_Conc': 1200.0,\n",
      "             'ChipID': '10X54',\n",
      "             'Comments': 'Cntnp-KO female, showing Cntnp \"male\" ASD emotional '\n",
      "                         'state preference. Perfusion not clear, oxygenated '\n",
      "                         'only at Wagner lab. 5% optiprep, suspension very '\n",
      "                         'good, emulsion OK.',\n",
      "             'Date': None,\n",
      "             'Date_Captured': 43993.0,\n",
      "             'DonorID': 'Cntnp-KO_2-9F',\n",
      "             'Group': 'Cntnp-KO-F',\n",
      "             'LIbConstructionComment': None,\n",
      "             'Num_Pooled_Animals': 1.0,\n",
      "             'PCR_Cycles': 13.0,\n",
      "             'Project': 'Cntnp_KO',\n",
      "             'SampleID': '10X54_1',\n",
      "             'Sample_Index': 'B3',\n",
      "             'Serial_Number': 106.0,\n",
      "             'Sex': 'F',\n",
      "             'Species': 'Mm',\n",
      "             'Strain': 'Cntnp-C57Bl/6',\n",
      "             'Target_Num_Cells': 5000.0,\n",
      "             'Tissue': 'MeA',\n",
      "             'Transcriptome': 'Mm10',\n",
      "             'cDNA_Lib_Ok': None,\n",
      "             'cDNAul': nan,\n",
      "             'lengthbp_seqlib': 473.0,\n",
      "             'ngperul_cDNA': 24.7,\n",
      "             'ngperul_seqlib': 17.9},\n",
      " '10X54_2': {'Age': '9.5w',\n",
      "             'Avesizebp_cDNAlib': nan,\n",
      "             'Cell_Conc': 1200.0,\n",
      "             'ChipID': '10X54',\n",
      "             'Comments': 'Cntnp-KO female, showing Cntnp \"male\" ASD emotional '\n",
      "                         'state preference. Oxygenated only at Wagner lab. 5% '\n",
      "                         'optiprep, suspension very good, emulsion OK.',\n",
      "             'Date': None,\n",
      "             'Date_Captured': 43993.0,\n",
      "             'DonorID': 'Cntnp-KO_4-10F',\n",
      "             'Group': 'Cntnp-KO-F',\n",
      "             'LIbConstructionComment': None,\n",
      "             'Num_Pooled_Animals': 1.0,\n",
      "             'PCR_Cycles': 13.0,\n",
      "             'Project': 'Cntnp_KO',\n",
      "             'SampleID': '10X54_2',\n",
      "             'Sample_Index': 'D3',\n",
      "             'Serial_Number': 107.0,\n",
      "             'Sex': 'F',\n",
      "             'Species': 'Mm',\n",
      "             'Strain': 'Cntnp-C57Bl/6',\n",
      "             'Target_Num_Cells': 5000.0,\n",
      "             'Tissue': 'MeA',\n",
      "             'Transcriptome': 'Mm10',\n",
      "             'cDNA_Lib_Ok': None,\n",
      "             'cDNAul': nan,\n",
      "             'lengthbp_seqlib': 471.0,\n",
      "             'ngperul_cDNA': 37.7,\n",
      "             'ngperul_seqlib': 20.0},\n",
      " '10x98_2': {'Age': None,\n",
      "             'Avesizebp_cDNAlib': 1873.5,\n",
      "             'Cell_Conc': 1000.0,\n",
      "             'ChipID': '10x98',\n",
      "             'Comments': 'v3! Cntnap KO #2, M; behavioral testing previous '\n",
      "                         'day. MeA dissection, dissociation took <40min, 2000g '\n",
      "                         'spin(!), good suspension, good emulsion. ',\n",
      "             'Date': '2021-04-04',\n",
      "             'Date_Captured': '2021-03-22',\n",
      "             'DonorID': 'Cntnp-KO_2_M',\n",
      "             'Group': 'Cntnp-KO-M',\n",
      "             'LIbConstructionComment': 'Single index; scale down; PCR_14 '\n",
      "                                       'cycles',\n",
      "             'Num_Pooled_Animals': 1.0,\n",
      "             'PCR_Cycles': 13.0,\n",
      "             'Project': 'Cntnp_KO',\n",
      "             'SampleID': '10x98_2',\n",
      "             'Sample_Index': 'H9',\n",
      "             'Serial_Number': 212.0,\n",
      "             'Sex': 'M',\n",
      "             'Species': 'Mm',\n",
      "             'Strain': 'Cntnp-C57Bl/6',\n",
      "             'Target_Num_Cells': 6000.0,\n",
      "             'Tissue': 'MeA',\n",
      "             'Transcriptome': 'Mm10',\n",
      "             'cDNA_Lib_Ok': None,\n",
      "             'cDNAul': 5.0,\n",
      "             'lengthbp_seqlib': 479.2,\n",
      "             'ngperul_cDNA': 39.7,\n",
      "             'ngperul_seqlib': 17.4},\n",
      " '10x98_3': {'Age': None,\n",
      "             'Avesizebp_cDNAlib': 1841.7,\n",
      "             'Cell_Conc': 1000.0,\n",
      "             'ChipID': '10x98',\n",
      "             'Comments': 'v3! Cntnap KO #3, M; behavioral testing previous '\n",
      "                         'day. MeA dissection, dissociation took <40min, 2000g '\n",
      "                         'spin(!), good suspension, good emulsion. ',\n",
      "             'Date': '2021-04-04',\n",
      "             'Date_Captured': '2021-03-22',\n",
      "             'DonorID': 'Cntnp-KO_3_M',\n",
      "             'Group': 'Cntnp-KO-M',\n",
      "             'LIbConstructionComment': 'Single index; scale down; PCR_14 '\n",
      "                                       'cycles',\n",
      "             'Num_Pooled_Animals': 1.0,\n",
      "             'PCR_Cycles': 13.0,\n",
      "             'Project': 'Cntnp_KO',\n",
      "             'SampleID': '10x98_3',\n",
      "             'Sample_Index': 'A10',\n",
      "             'Serial_Number': 213.0,\n",
      "             'Sex': 'M',\n",
      "             'Species': 'Mm',\n",
      "             'Strain': 'Cntnp-C57Bl/6',\n",
      "             'Target_Num_Cells': 6000.0,\n",
      "             'Tissue': 'MeA',\n",
      "             'Transcriptome': 'Mm10',\n",
      "             'cDNA_Lib_Ok': None,\n",
      "             'cDNAul': 5.0,\n",
      "             'lengthbp_seqlib': 426.4,\n",
      "             'ngperul_cDNA': 45.5,\n",
      "             'ngperul_seqlib': 21.9}}\n"
     ]
    }
   ],
   "source": [
    "pprint.pprint(meta_data_dict)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "id": "b006589c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "str"
      ]
     },
     "execution_count": 80,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "type(dimorph_df['Transcriptome:string'][0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "89be53cd",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "sample_dir = '/data/runs/samples/'\n",
    "dimorph_samples = samples\n",
    "\n",
    "os.chdir(sample_dir)\n",
    "cwd = os.getcwd()\n",
    "#print (cwd)\n",
    "\n",
    "#for loop would go here to loop over samples\n",
    "dir = str((glob.glob(cwd+'/'+dimorph_samples[0]+\n",
    "                 '/'+'*out*'+'/'+'outs'+\n",
    "                 '/'+'filtered_feature_bc_matrix'))).replace('[','').replace(']','')[1:-1]\n",
    "\n",
    "#print(cwd)\n",
    "#print (type(dir))\n",
    "#print (os.listdir(dir))\n",
    "\n",
    "m = mmread(dir + '/' + 'matrix.mtx')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "c54269d3",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[[0 0 0 ... 0 0 0]\n",
      " [0 0 0 ... 0 0 0]\n",
      " [0 0 0 ... 0 0 0]\n",
      " ...\n",
      " [0 0 0 ... 0 0 0]\n",
      " [0 0 0 ... 0 0 0]\n",
      " [0 0 0 ... 0 0 0]]\n",
      "(27998, 6229)\n",
      "<class 'numpy.ndarray'>\n"
     ]
    }
   ],
   "source": [
    "m_arr = m.toarray()\n",
    "print (m_arr)\n",
    "print (m_arr.shape)\n",
    "print (type(m_arr))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "58d08000",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'pandas.core.frame.DataFrame'>\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AAACCCACAACAGTGG-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AAACCCACATGGCCCA-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AAACCCAGTCCCTGAG-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AAACGAACACTACAGT-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AAACGAATCCCAGCGA-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6224</th>\n",
       "      <td>TTTGGTTTCATCTGTT-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6225</th>\n",
       "      <td>TTTGGTTTCTGTGCTC-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6226</th>\n",
       "      <td>TTTGTTGAGATGTTCC-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6227</th>\n",
       "      <td>TTTGTTGCATAGCACT-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6228</th>\n",
       "      <td>TTTGTTGTCCGCGGAT-1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>6229 rows × 1 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                       0\n",
       "0     AAACCCACAACAGTGG-1\n",
       "1     AAACCCACATGGCCCA-1\n",
       "2     AAACCCAGTCCCTGAG-1\n",
       "3     AAACGAACACTACAGT-1\n",
       "4     AAACGAATCCCAGCGA-1\n",
       "...                  ...\n",
       "6224  TTTGGTTTCATCTGTT-1\n",
       "6225  TTTGGTTTCTGTGCTC-1\n",
       "6226  TTTGTTGAGATGTTCC-1\n",
       "6227  TTTGTTGCATAGCACT-1\n",
       "6228  TTTGTTGTCCGCGGAT-1\n",
       "\n",
       "[6229 rows x 1 columns]"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "barcodes = pd.read_csv(dir+'/'+'barcodes.tsv', sep='\\t',header = None)\n",
    "print(type(barcodes))\n",
    "barcodes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "26577c1e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0       AAACCCACAACAGTGG-1\n",
       "1       AAACCCACATGGCCCA-1\n",
       "2       AAACCCAGTCCCTGAG-1\n",
       "3       AAACGAACACTACAGT-1\n",
       "4       AAACGAATCCCAGCGA-1\n",
       "               ...        \n",
       "6224    TTTGGTTTCATCTGTT-1\n",
       "6225    TTTGGTTTCTGTGCTC-1\n",
       "6226    TTTGTTGAGATGTTCC-1\n",
       "6227    TTTGTTGCATAGCACT-1\n",
       "6228    TTTGTTGTCCGCGGAT-1\n",
       "Name: 0, Length: 6229, dtype: object"
      ]
     },
     "execution_count": 39,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cells = barcodes.iloc[:,0]\n",
    "cells"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "38871a65",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                        0               1                2\n",
      "0      ENSMUSG00000051951            Xkr4  Gene Expression\n",
      "1      ENSMUSG00000089699          Gm1992  Gene Expression\n",
      "2      ENSMUSG00000102343         Gm37381  Gene Expression\n",
      "3      ENSMUSG00000025900             Rp1  Gene Expression\n",
      "4      ENSMUSG00000109048             Rp1  Gene Expression\n",
      "...                   ...             ...              ...\n",
      "27993  ENSMUSG00000079808      AC168977.1  Gene Expression\n",
      "27994  ENSMUSG00000095041            PISD  Gene Expression\n",
      "27995  ENSMUSG00000063897           DHRSX  Gene Expression\n",
      "27996  ENSMUSG00000096730        Vmn2r122  Gene Expression\n",
      "27997  ENSMUSG00000095742  CAAA01147332.1  Gene Expression\n",
      "\n",
      "[27998 rows x 3 columns]\n",
      "0       Xkr4\n",
      "1     Gm1992\n",
      "2    Gm37381\n",
      "3        Rp1\n",
      "Name: 1, dtype: object\n"
     ]
    }
   ],
   "source": [
    "features = pd.read_csv(dir+'/'+'features.tsv', sep='\\t', header=None)\n",
    "print(features)\n",
    "print(features.iloc[0:4,1])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "329d738e",
   "metadata": {},
   "outputs": [
    {
     "ename": "ValueError",
     "evalue": "Data must be 1-dimensional",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mValueError\u001b[0m                                Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[42], line 2\u001b[0m\n\u001b[1;32m      1\u001b[0m gene_labels \u001b[38;5;241m=\u001b[39m features\u001b[38;5;241m.\u001b[39miloc[:,\u001b[38;5;241m1\u001b[39m]\n\u001b[0;32m----> 2\u001b[0m gene_labels \u001b[38;5;241m=\u001b[39m np\u001b[38;5;241m.\u001b[39mreshape(gene_labels ,(gene_labels\u001b[38;5;241m.\u001b[39msize,\u001b[38;5;241m1\u001b[39m))\n\u001b[1;32m      3\u001b[0m gene_labels\u001b[38;5;241m.\u001b[39mshape\n",
      "File \u001b[0;32m<__array_function__ internals>:200\u001b[0m, in \u001b[0;36mreshape\u001b[0;34m(*args, **kwargs)\u001b[0m\n",
      "File \u001b[0;32m~/anaconda3/lib/python3.11/site-packages/numpy/core/fromnumeric.py:298\u001b[0m, in \u001b[0;36mreshape\u001b[0;34m(a, newshape, order)\u001b[0m\n\u001b[1;32m    198\u001b[0m \u001b[38;5;129m@array_function_dispatch\u001b[39m(_reshape_dispatcher)\n\u001b[1;32m    199\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21mreshape\u001b[39m(a, newshape, order\u001b[38;5;241m=\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mC\u001b[39m\u001b[38;5;124m'\u001b[39m):\n\u001b[1;32m    200\u001b[0m \u001b[38;5;250m    \u001b[39m\u001b[38;5;124;03m\"\"\"\u001b[39;00m\n\u001b[1;32m    201\u001b[0m \u001b[38;5;124;03m    Gives a new shape to an array without changing its data.\u001b[39;00m\n\u001b[1;32m    202\u001b[0m \n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m    296\u001b[0m \u001b[38;5;124;03m           [5, 6]])\u001b[39;00m\n\u001b[1;32m    297\u001b[0m \u001b[38;5;124;03m    \"\"\"\u001b[39;00m\n\u001b[0;32m--> 298\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m _wrapfunc(a, \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mreshape\u001b[39m\u001b[38;5;124m'\u001b[39m, newshape, order\u001b[38;5;241m=\u001b[39morder)\n",
      "File \u001b[0;32m~/anaconda3/lib/python3.11/site-packages/numpy/core/fromnumeric.py:54\u001b[0m, in \u001b[0;36m_wrapfunc\u001b[0;34m(obj, method, *args, **kwds)\u001b[0m\n\u001b[1;32m     52\u001b[0m bound \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mgetattr\u001b[39m(obj, method, \u001b[38;5;28;01mNone\u001b[39;00m)\n\u001b[1;32m     53\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m bound \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m:\n\u001b[0;32m---> 54\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m _wrapit(obj, method, \u001b[38;5;241m*\u001b[39margs, \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mkwds)\n\u001b[1;32m     56\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[1;32m     57\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m bound(\u001b[38;5;241m*\u001b[39margs, \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mkwds)\n",
      "File \u001b[0;32m~/anaconda3/lib/python3.11/site-packages/numpy/core/fromnumeric.py:47\u001b[0m, in \u001b[0;36m_wrapit\u001b[0;34m(obj, method, *args, **kwds)\u001b[0m\n\u001b[1;32m     45\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28misinstance\u001b[39m(result, mu\u001b[38;5;241m.\u001b[39mndarray):\n\u001b[1;32m     46\u001b[0m         result \u001b[38;5;241m=\u001b[39m asarray(result)\n\u001b[0;32m---> 47\u001b[0m     result \u001b[38;5;241m=\u001b[39m wrap(result)\n\u001b[1;32m     48\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m result\n",
      "File \u001b[0;32m~/anaconda3/lib/python3.11/site-packages/pandas/core/generic.py:2107\u001b[0m, in \u001b[0;36mNDFrame.__array_wrap__\u001b[0;34m(self, result, context)\u001b[0m\n\u001b[1;32m   2105\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m res\n\u001b[1;32m   2106\u001b[0m d \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_construct_axes_dict(\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_AXIS_ORDERS, copy\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mFalse\u001b[39;00m)\n\u001b[0;32m-> 2107\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_constructor(res, \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39md)\u001b[38;5;241m.\u001b[39m__finalize__(\u001b[38;5;28mself\u001b[39m, method\u001b[38;5;241m=\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124m__array_wrap__\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n",
      "File \u001b[0;32m~/anaconda3/lib/python3.11/site-packages/pandas/core/series.py:470\u001b[0m, in \u001b[0;36mSeries.__init__\u001b[0;34m(self, data, index, dtype, name, copy, fastpath)\u001b[0m\n\u001b[1;32m    468\u001b[0m         data \u001b[38;5;241m=\u001b[39m data\u001b[38;5;241m.\u001b[39mcopy()\n\u001b[1;32m    469\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[0;32m--> 470\u001b[0m     data \u001b[38;5;241m=\u001b[39m sanitize_array(data, index, dtype, copy)\n\u001b[1;32m    472\u001b[0m     manager \u001b[38;5;241m=\u001b[39m get_option(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mmode.data_manager\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n\u001b[1;32m    473\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m manager \u001b[38;5;241m==\u001b[39m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mblock\u001b[39m\u001b[38;5;124m\"\u001b[39m:\n",
      "File \u001b[0;32m~/anaconda3/lib/python3.11/site-packages/pandas/core/construction.py:647\u001b[0m, in \u001b[0;36msanitize_array\u001b[0;34m(data, index, dtype, copy, raise_cast_failure, allow_2d)\u001b[0m\n\u001b[1;32m    644\u001b[0m             subarr \u001b[38;5;241m=\u001b[39m cast(np\u001b[38;5;241m.\u001b[39mndarray, subarr)\n\u001b[1;32m    645\u001b[0m             subarr \u001b[38;5;241m=\u001b[39m maybe_infer_to_datetimelike(subarr)\n\u001b[0;32m--> 647\u001b[0m subarr \u001b[38;5;241m=\u001b[39m _sanitize_ndim(subarr, data, dtype, index, allow_2d\u001b[38;5;241m=\u001b[39mallow_2d)\n\u001b[1;32m    649\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28misinstance\u001b[39m(subarr, np\u001b[38;5;241m.\u001b[39mndarray):\n\u001b[1;32m    650\u001b[0m     \u001b[38;5;66;03m# at this point we should have dtype be None or subarr.dtype == dtype\u001b[39;00m\n\u001b[1;32m    651\u001b[0m     dtype \u001b[38;5;241m=\u001b[39m cast(np\u001b[38;5;241m.\u001b[39mdtype, dtype)\n",
      "File \u001b[0;32m~/anaconda3/lib/python3.11/site-packages/pandas/core/construction.py:698\u001b[0m, in \u001b[0;36m_sanitize_ndim\u001b[0;34m(result, data, dtype, index, allow_2d)\u001b[0m\n\u001b[1;32m    696\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m allow_2d:\n\u001b[1;32m    697\u001b[0m         \u001b[38;5;28;01mreturn\u001b[39;00m result\n\u001b[0;32m--> 698\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mValueError\u001b[39;00m(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mData must be 1-dimensional\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n\u001b[1;32m    699\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m is_object_dtype(dtype) \u001b[38;5;129;01mand\u001b[39;00m \u001b[38;5;28misinstance\u001b[39m(dtype, ExtensionDtype):\n\u001b[1;32m    700\u001b[0m     \u001b[38;5;66;03m# i.e. PandasDtype(\"O\")\u001b[39;00m\n\u001b[1;32m    702\u001b[0m     result \u001b[38;5;241m=\u001b[39m com\u001b[38;5;241m.\u001b[39masarray_tuplesafe(data, dtype\u001b[38;5;241m=\u001b[39mnp\u001b[38;5;241m.\u001b[39mdtype(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mobject\u001b[39m\u001b[38;5;124m\"\u001b[39m))\n",
      "\u001b[0;31mValueError\u001b[0m: Data must be 1-dimensional"
     ]
    }
   ],
   "source": [
    "gene_labels = features.iloc[:,1]\n",
    "gene_labels = np.reshape(gene_labels ,(gene_labels.size,1))\n",
    "gene_labels.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "id": "4cf82e22",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([['Xkr4'],\n",
       "       ['Gm1992'],\n",
       "       ['Gm37381'],\n",
       "       ...,\n",
       "       ['DHRSX'],\n",
       "       ['Vmn2r122'],\n",
       "       ['CAAA01147332.1']], dtype=object)"
      ]
     },
     "execution_count": 74,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_labels"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "e900f1f1",
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'gene_labels' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[23], line 2\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[38;5;66;03m#check for duplicates\u001b[39;00m\n\u001b[0;32m----> 2\u001b[0m gene_labels\u001b[38;5;241m.\u001b[39mis_unique\n",
      "\u001b[0;31mNameError\u001b[0m: name 'gene_labels' is not defined"
     ]
    }
   ],
   "source": [
    "#check for duplicates\n",
    "gene_labels.is_unique"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "id": "b0cbbdaf",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[array(['AAACCCACAACAGTGG-1', 'AAACCCACATGGCCCA-1', 'AAACCCAGTCCCTGAG-1',\n",
       "        ..., 'TTTGTTGAGATGTTCC-1', 'TTTGTTGCATAGCACT-1',\n",
       "        'TTTGTTGTCCGCGGAT-1'], dtype=object)]"
      ]
     },
     "execution_count": 75,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "list(cells)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "id": "70b98905",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead tr th {\n",
       "        text-align: left;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr>\n",
       "      <th></th>\n",
       "      <th>AAACCCACAACAGTGG-1</th>\n",
       "      <th>AAACCCACATGGCCCA-1</th>\n",
       "      <th>AAACCCAGTCCCTGAG-1</th>\n",
       "      <th>AAACGAACACTACAGT-1</th>\n",
       "      <th>AAACGAATCCCAGCGA-1</th>\n",
       "      <th>AAACGAATCTGCTTAT-1</th>\n",
       "      <th>AAACGCTAGCAGATAT-1</th>\n",
       "      <th>AAACGCTAGTGCTCAT-1</th>\n",
       "      <th>AAACGCTGTCCGTACG-1</th>\n",
       "      <th>AAACGCTGTGGCTACC-1</th>\n",
       "      <th>...</th>\n",
       "      <th>TTTGGTTAGAGTCACG-1</th>\n",
       "      <th>TTTGGTTCACAGCCTG-1</th>\n",
       "      <th>TTTGGTTCAGCCTTCT-1</th>\n",
       "      <th>TTTGGTTGTCACTAGT-1</th>\n",
       "      <th>TTTGGTTGTTCCTAAG-1</th>\n",
       "      <th>TTTGGTTTCATCTGTT-1</th>\n",
       "      <th>TTTGGTTTCTGTGCTC-1</th>\n",
       "      <th>TTTGTTGAGATGTTCC-1</th>\n",
       "      <th>TTTGTTGCATAGCACT-1</th>\n",
       "      <th>TTTGTTGTCCGCGGAT-1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>(Xkr4,)</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>(Gm1992,)</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>(Gm37381,)</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>(Rp1,)</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>(Rp1,)</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 6229 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           AAACCCACAACAGTGG-1 AAACCCACATGGCCCA-1 AAACCCAGTCCCTGAG-1  \\\n",
       "(Xkr4,)                     0                  0                  0   \n",
       "(Gm1992,)                   0                  0                  0   \n",
       "(Gm37381,)                  0                  0                  0   \n",
       "(Rp1,)                      0                  0                  0   \n",
       "(Rp1,)                      0                  0                  0   \n",
       "\n",
       "           AAACGAACACTACAGT-1 AAACGAATCCCAGCGA-1 AAACGAATCTGCTTAT-1  \\\n",
       "(Xkr4,)                     0                  0                  1   \n",
       "(Gm1992,)                   0                  0                  0   \n",
       "(Gm37381,)                  0                  0                  0   \n",
       "(Rp1,)                      0                  0                  0   \n",
       "(Rp1,)                      0                  0                  0   \n",
       "\n",
       "           AAACGCTAGCAGATAT-1 AAACGCTAGTGCTCAT-1 AAACGCTGTCCGTACG-1  \\\n",
       "(Xkr4,)                     0                  0                  2   \n",
       "(Gm1992,)                   0                  0                  0   \n",
       "(Gm37381,)                  0                  0                  0   \n",
       "(Rp1,)                      0                  0                  0   \n",
       "(Rp1,)                      0                  0                  0   \n",
       "\n",
       "           AAACGCTGTGGCTACC-1  ... TTTGGTTAGAGTCACG-1 TTTGGTTCACAGCCTG-1  \\\n",
       "(Xkr4,)                     0  ...                  0                  0   \n",
       "(Gm1992,)                   0  ...                  0                  0   \n",
       "(Gm37381,)                  0  ...                  0                  0   \n",
       "(Rp1,)                      0  ...                  0                  0   \n",
       "(Rp1,)                      0  ...                  0                  0   \n",
       "\n",
       "           TTTGGTTCAGCCTTCT-1 TTTGGTTGTCACTAGT-1 TTTGGTTGTTCCTAAG-1  \\\n",
       "(Xkr4,)                     0                  0                  0   \n",
       "(Gm1992,)                   0                  0                  0   \n",
       "(Gm37381,)                  0                  0                  0   \n",
       "(Rp1,)                      0                  0                  0   \n",
       "(Rp1,)                      0                  0                  0   \n",
       "\n",
       "           TTTGGTTTCATCTGTT-1 TTTGGTTTCTGTGCTC-1 TTTGTTGAGATGTTCC-1  \\\n",
       "(Xkr4,)                     0                  0                  0   \n",
       "(Gm1992,)                   0                  0                  0   \n",
       "(Gm37381,)                  0                  0                  0   \n",
       "(Rp1,)                      0                  0                  0   \n",
       "(Rp1,)                      0                  0                  0   \n",
       "\n",
       "           TTTGTTGCATAGCACT-1 TTTGTTGTCCGCGGAT-1  \n",
       "(Xkr4,)                     0                  0  \n",
       "(Gm1992,)                   0                  0  \n",
       "(Gm37381,)                  0                  0  \n",
       "(Rp1,)                      0                  0  \n",
       "(Rp1,)                      0                  0  \n",
       "\n",
       "[5 rows x 6229 columns]"
      ]
     },
     "execution_count": 76,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#combine into single dataframe\n",
    "sample_df = pd.DataFrame(data = m_arr,\n",
    "                         index = gene_labels,\n",
    "                         columns = list(cells))\n",
    "sample_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bd062041",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
