{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "495dc75a",
   "metadata": {},
   "outputs": [],
   "source": [
    "#notebook for integrating spatial and scrna seq data\n",
    "#uses scanpy_env, need to start jupyter notebook in shell with scanpy_env activated, then port forward from local machine"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 183,
   "id": "9db8e73f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "scanpy==1.10.2 anndata==0.10.8 umap==0.5.6 numpy==1.26.4 scipy==1.14.0 pandas==2.2.2 scikit-learn==1.5.1 statsmodels==0.14.2 igraph==0.11.6 pynndescent==0.5.13\n",
      "squidpy==1.5.0\n",
      "The autoreload extension is already loaded. To reload it, use:\n",
      "  %reload_ext autoreload\n"
     ]
    }
   ],
   "source": [
    "import scanpy as sc\n",
    "import squidpy as sq\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "from anndata import AnnData\n",
    "import pathlib\n",
    "import matplotlib.pyplot as plt\n",
    "import matplotlib as mpl\n",
    "import skimage\n",
    "import seaborn as sns\n",
    "import tangram as tg\n",
    "import json\n",
    "\n",
    "sc.logging.print_header()\n",
    "print(f\"squidpy=={sq.__version__}\")\n",
    "\n",
    "%load_ext autoreload\n",
    "%autoreload 2\n",
    "%matplotlib inline"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a9612980",
   "metadata": {},
   "source": [
    "# Load Datasets"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "5ec15384",
   "metadata": {},
   "outputs": [],
   "source": [
    "#example spatial for squidpy\n",
    "adata_st_ex = sq.datasets.visium_fluo_adata_crop()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "846d0231",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 704 × 16562\n",
       "    obs: 'in_tissue', 'array_row', 'array_col', 'n_genes_by_counts', 'log1p_n_genes_by_counts', 'total_counts', 'log1p_total_counts', 'pct_counts_in_top_50_genes', 'pct_counts_in_top_100_genes', 'pct_counts_in_top_200_genes', 'pct_counts_in_top_500_genes', 'total_counts_MT', 'log1p_total_counts_MT', 'pct_counts_MT', 'n_counts', 'leiden', 'cluster'\n",
       "    var: 'gene_ids', 'feature_types', 'genome', 'MT', 'n_cells_by_counts', 'mean_counts', 'log1p_mean_counts', 'pct_dropout_by_counts', 'total_counts', 'log1p_total_counts', 'n_cells', 'highly_variable', 'highly_variable_rank', 'means', 'variances', 'variances_norm'\n",
       "    uns: 'cluster_colors', 'hvg', 'leiden', 'leiden_colors', 'neighbors', 'pca', 'spatial', 'umap'\n",
       "    obsm: 'X_pca', 'X_umap', 'spatial'\n",
       "    varm: 'PCs'\n",
       "    obsp: 'connectivities', 'distances'"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata_st_ex"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "eb4a1ad0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(704, 2)"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata_st_ex.obsm['spatial'].shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "f84113ea",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AAACGAGACGGTTGAT-1    3808\n",
       "AAAGGGATGTAGCAAG-1    5288\n",
       "AAATGGCATGTCTTGT-1    5191\n",
       "AAATGGTCAATGTGCC-1    4923\n",
       "AAATTAACGGGTAGCT-1    4045\n",
       "                      ... \n",
       "TTGTCGTTCAGTTACC-1    3443\n",
       "TTGTGGCCCTGACAGT-1    4971\n",
       "TTGTTAGCAAATTCGA-1    4820\n",
       "TTGTTCAGTGTGCTAC-1    3372\n",
       "TTGTTGTGTGTCAAGA-1    3768\n",
       "Name: n_genes_by_counts, Length: 704, dtype: int32"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata_st_ex.obs['n_genes_by_counts']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "82edbfd7",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "adata_st_test = sc.read_visium()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "2a284379",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|████████████████████████████████████| 3.03G/3.03G [29:28<00:00, 1.84MB/s]\n"
     ]
    }
   ],
   "source": [
    "adata_sc = sq.datasets.sc_mouse_cortex()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "975c9eaa",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 21697 × 36826\n",
       "    obs: 'sample_name', 'organism', 'donor_sex', 'cell_class', 'cell_subclass', 'cell_cluster', 'n_genes_by_counts', 'log1p_n_genes_by_counts', 'total_counts', 'log1p_total_counts', 'pct_counts_in_top_50_genes', 'pct_counts_in_top_100_genes', 'pct_counts_in_top_200_genes', 'pct_counts_in_top_500_genes', 'total_counts_mt', 'log1p_total_counts_mt', 'pct_counts_mt', 'n_counts'\n",
       "    var: 'mt', 'n_cells_by_counts', 'mean_counts', 'log1p_mean_counts', 'pct_dropout_by_counts', 'total_counts', 'log1p_total_counts', 'n_cells', 'highly_variable', 'highly_variable_rank', 'means', 'variances', 'variances_norm'\n",
       "    uns: 'cell_class_colors', 'cell_subclass_colors', 'hvg', 'neighbors', 'pca', 'umap'\n",
       "    obsm: 'X_pca', 'X_umap'\n",
       "    varm: 'PCs'\n",
       "    obsp: 'connectivities', 'distances'"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata_sc"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 103,
   "id": "592b3f1c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>mt</th>\n",
       "      <th>n_cells_by_counts</th>\n",
       "      <th>mean_counts</th>\n",
       "      <th>log1p_mean_counts</th>\n",
       "      <th>pct_dropout_by_counts</th>\n",
       "      <th>total_counts</th>\n",
       "      <th>log1p_total_counts</th>\n",
       "      <th>n_cells</th>\n",
       "      <th>highly_variable</th>\n",
       "      <th>highly_variable_rank</th>\n",
       "      <th>means</th>\n",
       "      <th>variances</th>\n",
       "      <th>variances_norm</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0610005C13Rik</th>\n",
       "      <td>False</td>\n",
       "      <td>872</td>\n",
       "      <td>0.524247</td>\n",
       "      <td>0.421501</td>\n",
       "      <td>96.237812</td>\n",
       "      <td>12151.0</td>\n",
       "      <td>9.405249</td>\n",
       "      <td>867</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.534452</td>\n",
       "      <td>29.752143</td>\n",
       "      <td>1.090297</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0610006L08Rik</th>\n",
       "      <td>False</td>\n",
       "      <td>29</td>\n",
       "      <td>0.049918</td>\n",
       "      <td>0.048712</td>\n",
       "      <td>99.874881</td>\n",
       "      <td>1157.0</td>\n",
       "      <td>7.054450</td>\n",
       "      <td>29</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.050422</td>\n",
       "      <td>5.631491</td>\n",
       "      <td>2.697598</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0610007P14Rik</th>\n",
       "      <td>False</td>\n",
       "      <td>21519</td>\n",
       "      <td>161.839111</td>\n",
       "      <td>5.092762</td>\n",
       "      <td>7.157649</td>\n",
       "      <td>3751107.0</td>\n",
       "      <td>15.137562</td>\n",
       "      <td>21344</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>164.495598</td>\n",
       "      <td>14562.080191</td>\n",
       "      <td>0.629189</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0610009B22Rik</th>\n",
       "      <td>False</td>\n",
       "      <td>21203</td>\n",
       "      <td>150.453445</td>\n",
       "      <td>5.020278</td>\n",
       "      <td>8.521011</td>\n",
       "      <td>3487210.0</td>\n",
       "      <td>15.064612</td>\n",
       "      <td>21044</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>153.827165</td>\n",
       "      <td>13897.739024</td>\n",
       "      <td>0.652549</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0610009E02Rik</th>\n",
       "      <td>False</td>\n",
       "      <td>4185</td>\n",
       "      <td>1.559280</td>\n",
       "      <td>0.939726</td>\n",
       "      <td>81.944085</td>\n",
       "      <td>36141.0</td>\n",
       "      <td>10.495211</td>\n",
       "      <td>4166</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.591464</td>\n",
       "      <td>51.485746</td>\n",
       "      <td>0.484592</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Zzef1</th>\n",
       "      <td>False</td>\n",
       "      <td>16165</td>\n",
       "      <td>38.780094</td>\n",
       "      <td>3.683367</td>\n",
       "      <td>30.257140</td>\n",
       "      <td>898845.0</td>\n",
       "      <td>13.708867</td>\n",
       "      <td>16034</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>38.959856</td>\n",
       "      <td>4371.811764</td>\n",
       "      <td>0.873219</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Zzz3</th>\n",
       "      <td>False</td>\n",
       "      <td>16354</td>\n",
       "      <td>50.861378</td>\n",
       "      <td>3.948574</td>\n",
       "      <td>29.441712</td>\n",
       "      <td>1178865.0</td>\n",
       "      <td>13.980063</td>\n",
       "      <td>16227</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>51.304835</td>\n",
       "      <td>6362.498978</td>\n",
       "      <td>0.978110</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>a</th>\n",
       "      <td>False</td>\n",
       "      <td>188</td>\n",
       "      <td>0.151092</td>\n",
       "      <td>0.140711</td>\n",
       "      <td>99.188886</td>\n",
       "      <td>3502.0</td>\n",
       "      <td>8.161375</td>\n",
       "      <td>185</td>\n",
       "      <td>True</td>\n",
       "      <td>3379.0</td>\n",
       "      <td>0.152878</td>\n",
       "      <td>16.621585</td>\n",
       "      <td>3.142282</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>l7Rn6</th>\n",
       "      <td>False</td>\n",
       "      <td>21904</td>\n",
       "      <td>208.119507</td>\n",
       "      <td>5.342906</td>\n",
       "      <td>5.496592</td>\n",
       "      <td>4823794.0</td>\n",
       "      <td>15.389071</td>\n",
       "      <td>21728</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>213.394617</td>\n",
       "      <td>18848.619805</td>\n",
       "      <td>0.581846</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>n-R5s136</th>\n",
       "      <td>False</td>\n",
       "      <td>39</td>\n",
       "      <td>0.002718</td>\n",
       "      <td>0.002714</td>\n",
       "      <td>99.831737</td>\n",
       "      <td>63.0</td>\n",
       "      <td>4.158883</td>\n",
       "      <td>39</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.002765</td>\n",
       "      <td>0.007183</td>\n",
       "      <td>0.369636</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>36826 rows × 13 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                  mt  n_cells_by_counts  mean_counts  log1p_mean_counts  \\\n",
       "0610005C13Rik  False                872     0.524247           0.421501   \n",
       "0610006L08Rik  False                 29     0.049918           0.048712   \n",
       "0610007P14Rik  False              21519   161.839111           5.092762   \n",
       "0610009B22Rik  False              21203   150.453445           5.020278   \n",
       "0610009E02Rik  False               4185     1.559280           0.939726   \n",
       "...              ...                ...          ...                ...   \n",
       "Zzef1          False              16165    38.780094           3.683367   \n",
       "Zzz3           False              16354    50.861378           3.948574   \n",
       "a              False                188     0.151092           0.140711   \n",
       "l7Rn6          False              21904   208.119507           5.342906   \n",
       "n-R5s136       False                 39     0.002718           0.002714   \n",
       "\n",
       "               pct_dropout_by_counts  total_counts  log1p_total_counts  \\\n",
       "0610005C13Rik              96.237812       12151.0            9.405249   \n",
       "0610006L08Rik              99.874881        1157.0            7.054450   \n",
       "0610007P14Rik               7.157649     3751107.0           15.137562   \n",
       "0610009B22Rik               8.521011     3487210.0           15.064612   \n",
       "0610009E02Rik              81.944085       36141.0           10.495211   \n",
       "...                              ...           ...                 ...   \n",
       "Zzef1                      30.257140      898845.0           13.708867   \n",
       "Zzz3                       29.441712     1178865.0           13.980063   \n",
       "a                          99.188886        3502.0            8.161375   \n",
       "l7Rn6                       5.496592     4823794.0           15.389071   \n",
       "n-R5s136                   99.831737          63.0            4.158883   \n",
       "\n",
       "               n_cells  highly_variable  highly_variable_rank       means  \\\n",
       "0610005C13Rik      867            False                   NaN    0.534452   \n",
       "0610006L08Rik       29            False                   NaN    0.050422   \n",
       "0610007P14Rik    21344            False                   NaN  164.495598   \n",
       "0610009B22Rik    21044            False                   NaN  153.827165   \n",
       "0610009E02Rik     4166            False                   NaN    1.591464   \n",
       "...                ...              ...                   ...         ...   \n",
       "Zzef1            16034            False                   NaN   38.959856   \n",
       "Zzz3             16227            False                   NaN   51.304835   \n",
       "a                  185             True                3379.0    0.152878   \n",
       "l7Rn6            21728            False                   NaN  213.394617   \n",
       "n-R5s136            39            False                   NaN    0.002765   \n",
       "\n",
       "                  variances  variances_norm  \n",
       "0610005C13Rik     29.752143        1.090297  \n",
       "0610006L08Rik      5.631491        2.697598  \n",
       "0610007P14Rik  14562.080191        0.629189  \n",
       "0610009B22Rik  13897.739024        0.652549  \n",
       "0610009E02Rik     51.485746        0.484592  \n",
       "...                     ...             ...  \n",
       "Zzef1           4371.811764        0.873219  \n",
       "Zzz3            6362.498978        0.978110  \n",
       "a                 16.621585        3.142282  \n",
       "l7Rn6          18848.619805        0.581846  \n",
       "n-R5s136           0.007183        0.369636  \n",
       "\n",
       "[36826 rows x 13 columns]"
      ]
     },
     "execution_count": 103,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata_sc.var"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "c4dfbd34",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "F2S4_151217_005_B01    Pvalb\n",
       "F2S4_151217_005_C01       L4\n",
       "F2S4_151217_005_E01       L4\n",
       "F2S4_151217_005_F01       L4\n",
       "F2S4_151217_005_G01       L4\n",
       "                       ...  \n",
       "F1S4_180124_317_D01    Lamp5\n",
       "F1S4_180124_317_E01     Sncg\n",
       "F1S4_180124_317_F01      Sst\n",
       "F1S4_180124_317_G01     Sncg\n",
       "F1S4_180124_317_H01     Sncg\n",
       "Name: cell_subclass, Length: 21697, dtype: category\n",
       "Categories (23, object): ['Astro', 'CR', 'Endo', 'L2/3 IT', ..., 'Sncg', 'Sst', 'VLMC', 'Vip']"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata_sc.obs['cell_subclass']"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "89a554e5",
   "metadata": {},
   "source": [
    "# Load and Format SD Data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 280,
   "id": "b92dddb4",
   "metadata": {},
   "outputs": [],
   "source": [
    "gaba_folder = '/bigdata/isaac/gaba_files/'\n",
    "#load in all filtered metadata with subclass labels\n",
    "with open(gaba_folder + 'GABAmeta_data_df_plis_filtered_markers_2024-09-02.json') as json_data:\n",
    "    gaba_metadata = json.load(json_data)\n",
    "gaba_metadata = pd.DataFrame.from_dict(gaba_metadata, orient='columns')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "28fff4ba",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(30, 10022)"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gaba_metadata.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 281,
   "id": "58f028ec",
   "metadata": {},
   "outputs": [],
   "source": [
    "vglut1_folder = '/bigdata/isaac/Vglut1_files/'\n",
    "#load in all filtered metadata with subclass labels\n",
    "with open(vglut1_folder + 'VGLUT1meta_data_df_plis_filtered_markers_2024-09-01.json') as json_data:\n",
    "    vglut1_metadata = json.load(json_data)\n",
    "vglut1_metadata = pd.DataFrame.from_dict(vglut1_metadata, orient='columns')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "9f903ff1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(30, 9550)"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "vglut1_metadata.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 282,
   "id": "e571014c",
   "metadata": {},
   "outputs": [],
   "source": [
    "vglut2_folder = '/bigdata/isaac/Vglut2_files/'\n",
    "#load in all filtered metadata with subclass labels\n",
    "with open(vglut2_folder + 'VGLUT2meta_data_df_plis_filtered_markers_2024-09-01.json') as json_data:\n",
    "    vglut2_metadata = json.load(json_data)\n",
    "vglut2_metadata = pd.DataFrame.from_dict(vglut2_metadata, orient='columns')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "726c99c2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(30, 5276)"
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "vglut2_metadata.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 283,
   "id": "356938ce",
   "metadata": {},
   "outputs": [],
   "source": [
    "nn_folder = '/bigdata/isaac/Nonneuronal_files/'\n",
    "#load in all filtered metadata with subclass labels\n",
    "with open(nn_folder + 'NNmeta_data_df_plis_filtered_markers_2024-09-18.json') as json_data:\n",
    "    nn_metadata = json.load(json_data)\n",
    "nn_metadata = pd.DataFrame.from_dict(nn_metadata, orient='columns')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "7a6bc8a6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(30, 1976)"
      ]
     },
     "execution_count": 43,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "nn_metadata.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5a7f72de",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e8151e4f",
   "metadata": {},
   "outputs": [],
   "source": [
    "#add dash at end of single cluster labels for each class to fix obs.columns matching var names issue downstream"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 241,
   "id": "c1b189c2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19,\n",
       "       20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35, 36,\n",
       "       37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50],\n",
       "      dtype=object)"
      ]
     },
     "execution_count": 241,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.unique(gaba_metadata.loc['cluster_label'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 251,
   "id": "1da0b211",
   "metadata": {},
   "outputs": [],
   "source": [
    "gaba_metadata.loc['markers' ,gaba_metadata.loc['markers'] == 'Tmem100'] = 'Tmem100-'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 253,
   "id": "27575c46",
   "metadata": {},
   "outputs": [],
   "source": [
    "gaba_metadata.loc['markers' ,gaba_metadata.loc['markers'] == 'Lpl'] = 'Lpl-'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 254,
   "id": "2c15e442",
   "metadata": {},
   "outputs": [],
   "source": [
    "gaba_metadata.loc['markers' ,gaba_metadata.loc['markers'] == 'Npy2r'] = 'Npy2r-'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 255,
   "id": "bdfa8a57",
   "metadata": {},
   "outputs": [],
   "source": [
    "gaba_metadata.loc['markers' ,gaba_metadata.loc['markers'] == 'Nwd2'] = 'Nwd2-'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 260,
   "id": "b16ee526",
   "metadata": {},
   "outputs": [],
   "source": [
    "gaba_metadata.loc['markers' ,gaba_metadata.loc['markers'] == 'Myh7'] = 'Myh7-'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 266,
   "id": "ed440ab6",
   "metadata": {},
   "outputs": [],
   "source": [
    "vglut1_metadata.loc['markers' ,vglut1_metadata.loc['markers'] == 'Lypd6'] = 'Lypd6-'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 267,
   "id": "f6a941ce",
   "metadata": {},
   "outputs": [],
   "source": [
    "vglut1_metadata.loc['markers' ,vglut1_metadata.loc['markers'] == 'Moxd1'] = 'Moxd1-'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 268,
   "id": "c3f33bb3",
   "metadata": {},
   "outputs": [],
   "source": [
    "vglut1_metadata.loc['markers' ,vglut1_metadata.loc['markers'] == 'Gfod2'] = 'Gfod2-'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 269,
   "id": "3fb69820",
   "metadata": {},
   "outputs": [],
   "source": [
    "vglut1_metadata.loc['markers' ,vglut1_metadata.loc['markers'] == 'Rab3b'] = 'Rab3b-'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bd828883",
   "metadata": {},
   "outputs": [],
   "source": [
    "vglut2_metadata.loc['markers' ,vglut2_metadata.loc['markers'] == 'Lmo1'] = 'Lmo1-'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 270,
   "id": "ab1b8364",
   "metadata": {},
   "outputs": [],
   "source": [
    "vglut2_metadata.loc['markers' ,vglut2_metadata.loc['markers'] == 'Ebf1'] = 'Ebf1-'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 271,
   "id": "8a795d8c",
   "metadata": {},
   "outputs": [],
   "source": [
    "vglut2_metadata.loc['markers' ,vglut2_metadata.loc['markers'] == 'Pgr15l'] = 'Pgr15l-'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ffbdb73c",
   "metadata": {},
   "outputs": [],
   "source": [
    "vglut2_metadata.loc['markers' ,vglut2_metadata.loc['markers'] == 'Pgr15l'] = 'Pgr15l-'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 272,
   "id": "47c27ad8",
   "metadata": {},
   "outputs": [],
   "source": [
    "nn_metadata.loc['markers' ,nn_metadata.loc['markers'] == 'Ccl2'] = 'Ccl2-'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "514ef26e",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7357ac6a",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3db7ef06",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "470f924f",
   "metadata": {},
   "outputs": [],
   "source": [
    "#concat into one "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 284,
   "id": "454d59c5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CAACAACAGACATATG-1_10X35_2</th>\n",
       "      <th>ATGACCATCGTGAGAG-1_10X51_1</th>\n",
       "      <th>AACCCAAAGAAATGGG-1_10X35_2</th>\n",
       "      <th>GAGGGATGTATGCTAC-1_10X35_1</th>\n",
       "      <th>AAGATAGGTAACACCT-1_10X37_2</th>\n",
       "      <th>CAGGGCTGTATCTTCT-1_10X51_1</th>\n",
       "      <th>CTGCTCATCCGATCGG-1_10X38_1</th>\n",
       "      <th>TGCTCCAAGAGGCGTT-1_10X51_1</th>\n",
       "      <th>TTGCCTGTCGTAGGGA-1_10X35_2</th>\n",
       "      <th>GCAACATAGATCGACG-1_10X51_2</th>\n",
       "      <th>...</th>\n",
       "      <th>CAAGAGGGTCAATCTG-1_10X36_2</th>\n",
       "      <th>TATACCTAGAGTAACT-1_10X35_1</th>\n",
       "      <th>GCACGTGGTTATTCCT-1_10X37_1</th>\n",
       "      <th>GATGATCTCAAGCCTA-1_10X36_1</th>\n",
       "      <th>GGTGTCGAGAAGCTGC-1_10X36_2</th>\n",
       "      <th>GAGAGGTCACATCATG-1_10X37_2</th>\n",
       "      <th>ATCAGGTTCCACTAGA-1_10X35_2</th>\n",
       "      <th>TCCACGTAGAGCATCG-1_10X37_1</th>\n",
       "      <th>CTGCGAGCACTACGGC-1_10X35_2</th>\n",
       "      <th>CTCTGGTTCGAAGCAG-1_10X37_1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Serial_Number</th>\n",
       "      <td>64.0</td>\n",
       "      <td>97.0</td>\n",
       "      <td>64.0</td>\n",
       "      <td>63.0</td>\n",
       "      <td>68.0</td>\n",
       "      <td>97.0</td>\n",
       "      <td>69.0</td>\n",
       "      <td>97.0</td>\n",
       "      <td>64.0</td>\n",
       "      <td>98.0</td>\n",
       "      <td>...</td>\n",
       "      <td>66.0</td>\n",
       "      <td>63.0</td>\n",
       "      <td>67.0</td>\n",
       "      <td>65.0</td>\n",
       "      <td>66.0</td>\n",
       "      <td>68.0</td>\n",
       "      <td>64.0</td>\n",
       "      <td>67.0</td>\n",
       "      <td>64.0</td>\n",
       "      <td>67.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Date_Captured</th>\n",
       "      <td>2019-11-20</td>\n",
       "      <td>05/31/20</td>\n",
       "      <td>2019-11-20</td>\n",
       "      <td>2019-11-20</td>\n",
       "      <td>2019-11-25</td>\n",
       "      <td>05/31/20</td>\n",
       "      <td>2019-12-01</td>\n",
       "      <td>05/31/20</td>\n",
       "      <td>2019-11-20</td>\n",
       "      <td>05/31/20</td>\n",
       "      <td>...</td>\n",
       "      <td>2019-11-24</td>\n",
       "      <td>2019-11-20</td>\n",
       "      <td>2019-11-25</td>\n",
       "      <td>2019-11-24</td>\n",
       "      <td>2019-11-24</td>\n",
       "      <td>2019-11-25</td>\n",
       "      <td>2019-11-20</td>\n",
       "      <td>2019-11-25</td>\n",
       "      <td>2019-11-20</td>\n",
       "      <td>2019-11-25</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Species</th>\n",
       "      <td>Mm</td>\n",
       "      <td>Mm</td>\n",
       "      <td>Mm</td>\n",
       "      <td>Mm</td>\n",
       "      <td>Mm</td>\n",
       "      <td>Mm</td>\n",
       "      <td>Mm</td>\n",
       "      <td>Mm</td>\n",
       "      <td>Mm</td>\n",
       "      <td>Mm</td>\n",
       "      <td>...</td>\n",
       "      <td>Mm</td>\n",
       "      <td>Mm</td>\n",
       "      <td>Mm</td>\n",
       "      <td>Mm</td>\n",
       "      <td>Mm</td>\n",
       "      <td>Mm</td>\n",
       "      <td>Mm</td>\n",
       "      <td>Mm</td>\n",
       "      <td>Mm</td>\n",
       "      <td>Mm</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Transcriptome</th>\n",
       "      <td>Mm10</td>\n",
       "      <td>Mm10</td>\n",
       "      <td>Mm10</td>\n",
       "      <td>Mm10</td>\n",
       "      <td>Mm10</td>\n",
       "      <td>Mm10</td>\n",
       "      <td>Mm10</td>\n",
       "      <td>Mm10</td>\n",
       "      <td>Mm10</td>\n",
       "      <td>Mm10</td>\n",
       "      <td>...</td>\n",
       "      <td>Mm10</td>\n",
       "      <td>Mm10</td>\n",
       "      <td>Mm10</td>\n",
       "      <td>Mm10</td>\n",
       "      <td>Mm10</td>\n",
       "      <td>Mm10</td>\n",
       "      <td>Mm10</td>\n",
       "      <td>Mm10</td>\n",
       "      <td>Mm10</td>\n",
       "      <td>Mm10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Strain</th>\n",
       "      <td></td>\n",
       "      <td>C57Bl/6</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>C57Bl/6</td>\n",
       "      <td></td>\n",
       "      <td>C57Bl/6</td>\n",
       "      <td></td>\n",
       "      <td>C57Bl/6</td>\n",
       "      <td>...</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Project</th>\n",
       "      <td>Dimorph</td>\n",
       "      <td>Dimorph</td>\n",
       "      <td>Dimorph</td>\n",
       "      <td>Dimorph</td>\n",
       "      <td>Dimorph</td>\n",
       "      <td>Dimorph</td>\n",
       "      <td>Dimorph</td>\n",
       "      <td>Dimorph</td>\n",
       "      <td>Dimorph</td>\n",
       "      <td>Dimorph</td>\n",
       "      <td>...</td>\n",
       "      <td>Dimorph</td>\n",
       "      <td>Dimorph</td>\n",
       "      <td>Dimorph</td>\n",
       "      <td>Dimorph</td>\n",
       "      <td>Dimorph</td>\n",
       "      <td>Dimorph</td>\n",
       "      <td>Dimorph</td>\n",
       "      <td>Dimorph</td>\n",
       "      <td>Dimorph</td>\n",
       "      <td>Dimorph</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Group</th>\n",
       "      <td>Naïve-F</td>\n",
       "      <td>Breeder-F</td>\n",
       "      <td>Naïve-F</td>\n",
       "      <td>Naïve-F</td>\n",
       "      <td>Naïve-M</td>\n",
       "      <td>Breeder-F</td>\n",
       "      <td>Naïve-F</td>\n",
       "      <td>Breeder-F</td>\n",
       "      <td>Naïve-F</td>\n",
       "      <td>Breeder-F</td>\n",
       "      <td>...</td>\n",
       "      <td>Naïve-M</td>\n",
       "      <td>Naïve-F</td>\n",
       "      <td>Naïve-M</td>\n",
       "      <td>Naïve-M</td>\n",
       "      <td>Naïve-M</td>\n",
       "      <td>Naïve-M</td>\n",
       "      <td>Naïve-F</td>\n",
       "      <td>Naïve-M</td>\n",
       "      <td>Naïve-F</td>\n",
       "      <td>Naïve-M</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ChipID</th>\n",
       "      <td>10X35</td>\n",
       "      <td>10X51</td>\n",
       "      <td>10X35</td>\n",
       "      <td>10X35</td>\n",
       "      <td>10X37</td>\n",
       "      <td>10X51</td>\n",
       "      <td>10X38</td>\n",
       "      <td>10X51</td>\n",
       "      <td>10X35</td>\n",
       "      <td>10X51</td>\n",
       "      <td>...</td>\n",
       "      <td>10X36</td>\n",
       "      <td>10X35</td>\n",
       "      <td>10X37</td>\n",
       "      <td>10X36</td>\n",
       "      <td>10X36</td>\n",
       "      <td>10X37</td>\n",
       "      <td>10X35</td>\n",
       "      <td>10X37</td>\n",
       "      <td>10X35</td>\n",
       "      <td>10X37</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SampleID</th>\n",
       "      <td>10X35_2</td>\n",
       "      <td>10X51_1</td>\n",
       "      <td>10X35_2</td>\n",
       "      <td>10X35_1</td>\n",
       "      <td>10X37_2</td>\n",
       "      <td>10X51_1</td>\n",
       "      <td>10X38_1</td>\n",
       "      <td>10X51_1</td>\n",
       "      <td>10X35_2</td>\n",
       "      <td>10X51_2</td>\n",
       "      <td>...</td>\n",
       "      <td>10X36_2</td>\n",
       "      <td>10X35_1</td>\n",
       "      <td>10X37_1</td>\n",
       "      <td>10X36_1</td>\n",
       "      <td>10X36_2</td>\n",
       "      <td>10X37_2</td>\n",
       "      <td>10X35_2</td>\n",
       "      <td>10X37_1</td>\n",
       "      <td>10X35_2</td>\n",
       "      <td>10X37_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DonorID</th>\n",
       "      <td>DI1,DI2</td>\n",
       "      <td>DI-B1-F</td>\n",
       "      <td>DI1,DI2</td>\n",
       "      <td>DI1,DI2</td>\n",
       "      <td>DI4,DI5</td>\n",
       "      <td>DI-B1-F</td>\n",
       "      <td>DI6</td>\n",
       "      <td>DI-B1-F</td>\n",
       "      <td>DI1,DI2</td>\n",
       "      <td>DI-B1-F</td>\n",
       "      <td>...</td>\n",
       "      <td>DI3</td>\n",
       "      <td>DI1,DI2</td>\n",
       "      <td>DI4,DI5</td>\n",
       "      <td>DI3</td>\n",
       "      <td>DI3</td>\n",
       "      <td>DI4,DI5</td>\n",
       "      <td>DI1,DI2</td>\n",
       "      <td>DI4,DI5</td>\n",
       "      <td>DI1,DI2</td>\n",
       "      <td>DI4,DI5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Age</th>\n",
       "      <td>8w</td>\n",
       "      <td>11-12w</td>\n",
       "      <td>8w</td>\n",
       "      <td>8w</td>\n",
       "      <td>8.5w</td>\n",
       "      <td>11-12w</td>\n",
       "      <td>9w</td>\n",
       "      <td>11-12w</td>\n",
       "      <td>8w</td>\n",
       "      <td>11-12w</td>\n",
       "      <td>...</td>\n",
       "      <td>8.5w</td>\n",
       "      <td>8w</td>\n",
       "      <td>8.5w</td>\n",
       "      <td>8.5w</td>\n",
       "      <td>8.5w</td>\n",
       "      <td>8.5w</td>\n",
       "      <td>8w</td>\n",
       "      <td>8.5w</td>\n",
       "      <td>8w</td>\n",
       "      <td>8.5w</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Num_Pooled_Animals</th>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Sex</th>\n",
       "      <td>F</td>\n",
       "      <td>F</td>\n",
       "      <td>F</td>\n",
       "      <td>F</td>\n",
       "      <td>M</td>\n",
       "      <td>F</td>\n",
       "      <td>F</td>\n",
       "      <td>F</td>\n",
       "      <td>F</td>\n",
       "      <td>F</td>\n",
       "      <td>...</td>\n",
       "      <td>M</td>\n",
       "      <td>F</td>\n",
       "      <td>M</td>\n",
       "      <td>M</td>\n",
       "      <td>M</td>\n",
       "      <td>M</td>\n",
       "      <td>F</td>\n",
       "      <td>M</td>\n",
       "      <td>F</td>\n",
       "      <td>M</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Tissue</th>\n",
       "      <td>MeA_R</td>\n",
       "      <td>MeA</td>\n",
       "      <td>MeA_R</td>\n",
       "      <td>MeA_L</td>\n",
       "      <td>MeA_R</td>\n",
       "      <td>MeA</td>\n",
       "      <td>MeA_L</td>\n",
       "      <td>MeA</td>\n",
       "      <td>MeA_R</td>\n",
       "      <td>MeA</td>\n",
       "      <td>...</td>\n",
       "      <td>MeA_R</td>\n",
       "      <td>MeA_L</td>\n",
       "      <td>MeA_L</td>\n",
       "      <td>MeA_L</td>\n",
       "      <td>MeA_R</td>\n",
       "      <td>MeA_R</td>\n",
       "      <td>MeA_R</td>\n",
       "      <td>MeA_L</td>\n",
       "      <td>MeA_R</td>\n",
       "      <td>MeA_L</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Cell_Conc</th>\n",
       "      <td>1000.0</td>\n",
       "      <td>1000.0</td>\n",
       "      <td>1000.0</td>\n",
       "      <td>1000.0</td>\n",
       "      <td>1000.0</td>\n",
       "      <td>1000.0</td>\n",
       "      <td>1000.0</td>\n",
       "      <td>1000.0</td>\n",
       "      <td>1000.0</td>\n",
       "      <td>1000.0</td>\n",
       "      <td>...</td>\n",
       "      <td>1000.0</td>\n",
       "      <td>1000.0</td>\n",
       "      <td>1000.0</td>\n",
       "      <td>1000.0</td>\n",
       "      <td>1000.0</td>\n",
       "      <td>1000.0</td>\n",
       "      <td>1000.0</td>\n",
       "      <td>1000.0</td>\n",
       "      <td>1000.0</td>\n",
       "      <td>1000.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Target_Num_Cells</th>\n",
       "      <td>5000.0</td>\n",
       "      <td>5000.0</td>\n",
       "      <td>5000.0</td>\n",
       "      <td>5000.0</td>\n",
       "      <td>5000.0</td>\n",
       "      <td>5000.0</td>\n",
       "      <td>5000.0</td>\n",
       "      <td>5000.0</td>\n",
       "      <td>5000.0</td>\n",
       "      <td>5000.0</td>\n",
       "      <td>...</td>\n",
       "      <td>5000.0</td>\n",
       "      <td>5000.0</td>\n",
       "      <td>5000.0</td>\n",
       "      <td>5000.0</td>\n",
       "      <td>5000.0</td>\n",
       "      <td>5000.0</td>\n",
       "      <td>5000.0</td>\n",
       "      <td>5000.0</td>\n",
       "      <td>5000.0</td>\n",
       "      <td>5000.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PCR_Cycles</th>\n",
       "      <td>13.0</td>\n",
       "      <td>13.0</td>\n",
       "      <td>13.0</td>\n",
       "      <td>13.0</td>\n",
       "      <td>13.0</td>\n",
       "      <td>13.0</td>\n",
       "      <td>13.0</td>\n",
       "      <td>13.0</td>\n",
       "      <td>13.0</td>\n",
       "      <td>13.0</td>\n",
       "      <td>...</td>\n",
       "      <td>13.0</td>\n",
       "      <td>13.0</td>\n",
       "      <td>13.0</td>\n",
       "      <td>13.0</td>\n",
       "      <td>13.0</td>\n",
       "      <td>13.0</td>\n",
       "      <td>13.0</td>\n",
       "      <td>13.0</td>\n",
       "      <td>13.0</td>\n",
       "      <td>13.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Comments</th>\n",
       "      <td>Naive 8wo females, pooled, right hemisphere. 5...</td>\n",
       "      <td>Breeders B1, 8pups DOB 27.5 (P4). perfusion aC...</td>\n",
       "      <td>Naive 8wo females, pooled, right hemisphere. 5...</td>\n",
       "      <td>Naive 8wo females, pooled, left hemisphere. 5%...</td>\n",
       "      <td>Naive 8.5wo males, pooled, right hemisphere. 5...</td>\n",
       "      <td>Breeders B1, 8pups DOB 27.5 (P4). perfusion aC...</td>\n",
       "      <td>Naive 9wo female, left hemisphere. 5%opti.</td>\n",
       "      <td>Breeders B1, 8pups DOB 27.5 (P4). perfusion aC...</td>\n",
       "      <td>Naive 8wo females, pooled, right hemisphere. 5...</td>\n",
       "      <td>Breeders B1, 8pups DOB 27.5 (P4). perfusion aC...</td>\n",
       "      <td>...</td>\n",
       "      <td>Naive 8.5wo male, right hemisphere. 5%opti.</td>\n",
       "      <td>Naive 8wo females, pooled, left hemisphere. 5%...</td>\n",
       "      <td>Naive 8.5wo males, pooled, left hemisphere. 5%...</td>\n",
       "      <td>Naive 8.5wo male, left hemisphere. 5%opti.</td>\n",
       "      <td>Naive 8.5wo male, right hemisphere. 5%opti.</td>\n",
       "      <td>Naive 8.5wo males, pooled, right hemisphere. 5...</td>\n",
       "      <td>Naive 8wo females, pooled, right hemisphere. 5...</td>\n",
       "      <td>Naive 8.5wo males, pooled, left hemisphere. 5%...</td>\n",
       "      <td>Naive 8wo females, pooled, right hemisphere. 5...</td>\n",
       "      <td>Naive 8.5wo males, pooled, left hemisphere. 5%...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cDNA_Lib_Ok</th>\n",
       "      <td>Y</td>\n",
       "      <td></td>\n",
       "      <td>Y</td>\n",
       "      <td>Y</td>\n",
       "      <td>Y</td>\n",
       "      <td></td>\n",
       "      <td>Y</td>\n",
       "      <td></td>\n",
       "      <td>Y</td>\n",
       "      <td></td>\n",
       "      <td>...</td>\n",
       "      <td>Y</td>\n",
       "      <td>Y</td>\n",
       "      <td>Y</td>\n",
       "      <td>Y</td>\n",
       "      <td>Y</td>\n",
       "      <td>Y</td>\n",
       "      <td>Y</td>\n",
       "      <td>Y</td>\n",
       "      <td>Y</td>\n",
       "      <td>Y</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ngperul_cDNA</th>\n",
       "      <td>26.1</td>\n",
       "      <td>32.5</td>\n",
       "      <td>26.1</td>\n",
       "      <td>23.3</td>\n",
       "      <td>24.4</td>\n",
       "      <td>32.5</td>\n",
       "      <td>21.9</td>\n",
       "      <td>32.5</td>\n",
       "      <td>26.1</td>\n",
       "      <td>29.8</td>\n",
       "      <td>...</td>\n",
       "      <td>16.7</td>\n",
       "      <td>23.3</td>\n",
       "      <td>17.0</td>\n",
       "      <td>26.4</td>\n",
       "      <td>16.7</td>\n",
       "      <td>24.4</td>\n",
       "      <td>26.1</td>\n",
       "      <td>17.0</td>\n",
       "      <td>26.1</td>\n",
       "      <td>17.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Avesizebp_cDNAlib</th>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>1870.0</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>...</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Date</th>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>...</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cDNAul</th>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>...</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>LIbConstructionComment</th>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>...</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ngperul_seqlib</th>\n",
       "      <td>17.6</td>\n",
       "      <td>19.2</td>\n",
       "      <td>17.6</td>\n",
       "      <td>13.6</td>\n",
       "      <td>16.0</td>\n",
       "      <td>19.2</td>\n",
       "      <td>15.0</td>\n",
       "      <td>19.2</td>\n",
       "      <td>17.6</td>\n",
       "      <td>22.2</td>\n",
       "      <td>...</td>\n",
       "      <td>12.1</td>\n",
       "      <td>13.6</td>\n",
       "      <td>15.9</td>\n",
       "      <td>11.4</td>\n",
       "      <td>12.1</td>\n",
       "      <td>16.0</td>\n",
       "      <td>17.6</td>\n",
       "      <td>15.9</td>\n",
       "      <td>17.6</td>\n",
       "      <td>15.9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>lengthbp_seqlib</th>\n",
       "      <td>461.0</td>\n",
       "      <td>478.0</td>\n",
       "      <td>461.0</td>\n",
       "      <td>416.0</td>\n",
       "      <td>460.0</td>\n",
       "      <td>478.0</td>\n",
       "      <td>420.0</td>\n",
       "      <td>478.0</td>\n",
       "      <td>461.0</td>\n",
       "      <td>474.0</td>\n",
       "      <td>...</td>\n",
       "      <td>486.0</td>\n",
       "      <td>416.0</td>\n",
       "      <td>440.0</td>\n",
       "      <td>476.0</td>\n",
       "      <td>486.0</td>\n",
       "      <td>460.0</td>\n",
       "      <td>461.0</td>\n",
       "      <td>440.0</td>\n",
       "      <td>461.0</td>\n",
       "      <td>440.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Sample_Index</th>\n",
       "      <td>G8</td>\n",
       "      <td>G1</td>\n",
       "      <td>G8</td>\n",
       "      <td>D9</td>\n",
       "      <td>F9</td>\n",
       "      <td>G1</td>\n",
       "      <td>B9</td>\n",
       "      <td>G1</td>\n",
       "      <td>G8</td>\n",
       "      <td>A2</td>\n",
       "      <td>...</td>\n",
       "      <td>A9</td>\n",
       "      <td>D9</td>\n",
       "      <td>E9</td>\n",
       "      <td>H8</td>\n",
       "      <td>A9</td>\n",
       "      <td>F9</td>\n",
       "      <td>G8</td>\n",
       "      <td>E9</td>\n",
       "      <td>G8</td>\n",
       "      <td>E9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cell_class</th>\n",
       "      <td>GABA</td>\n",
       "      <td>GABA</td>\n",
       "      <td>GABA</td>\n",
       "      <td>GABA</td>\n",
       "      <td>GABA</td>\n",
       "      <td>GABA</td>\n",
       "      <td>GABA</td>\n",
       "      <td>GABA</td>\n",
       "      <td>GABA</td>\n",
       "      <td>GABA</td>\n",
       "      <td>...</td>\n",
       "      <td>Nonneuronal</td>\n",
       "      <td>Nonneuronal</td>\n",
       "      <td>Nonneuronal</td>\n",
       "      <td>Nonneuronal</td>\n",
       "      <td>Nonneuronal</td>\n",
       "      <td>Nonneuronal</td>\n",
       "      <td>Nonneuronal</td>\n",
       "      <td>Nonneuronal</td>\n",
       "      <td>Nonneuronal</td>\n",
       "      <td>Nonneuronal</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cluster_label</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>11</td>\n",
       "      <td>11</td>\n",
       "      <td>11</td>\n",
       "      <td>11</td>\n",
       "      <td>11</td>\n",
       "      <td>11</td>\n",
       "      <td>11</td>\n",
       "      <td>11</td>\n",
       "      <td>11</td>\n",
       "      <td>11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>markers</th>\n",
       "      <td>Chat-Isl1</td>\n",
       "      <td>Chat-Isl1</td>\n",
       "      <td>Chat-Isl1</td>\n",
       "      <td>Chat-Isl1</td>\n",
       "      <td>Chat-Isl1</td>\n",
       "      <td>Chat-Isl1</td>\n",
       "      <td>Chat-Isl1</td>\n",
       "      <td>Chat-Isl1</td>\n",
       "      <td>Chat-Isl1</td>\n",
       "      <td>Chat-Isl1</td>\n",
       "      <td>...</td>\n",
       "      <td>OL_2</td>\n",
       "      <td>OL_2</td>\n",
       "      <td>OL_2</td>\n",
       "      <td>OL_2</td>\n",
       "      <td>OL_2</td>\n",
       "      <td>OL_2</td>\n",
       "      <td>OL_2</td>\n",
       "      <td>OL_2</td>\n",
       "      <td>OL_2</td>\n",
       "      <td>OL_2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>30 rows × 26824 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                               CAACAACAGACATATG-1_10X35_2  \\\n",
       "Serial_Number                                                        64.0   \n",
       "Date_Captured                                                  2019-11-20   \n",
       "Species                                                                Mm   \n",
       "Transcriptome                                                        Mm10   \n",
       "Strain                                                                      \n",
       "Project                                                           Dimorph   \n",
       "Group                                                             Naïve-F   \n",
       "ChipID                                                              10X35   \n",
       "SampleID                                                          10X35_2   \n",
       "DonorID                                                           DI1,DI2   \n",
       "Age                                                                    8w   \n",
       "Num_Pooled_Animals                                                    2.0   \n",
       "Sex                                                                     F   \n",
       "Tissue                                                              MeA_R   \n",
       "Cell_Conc                                                          1000.0   \n",
       "Target_Num_Cells                                                   5000.0   \n",
       "PCR_Cycles                                                           13.0   \n",
       "Comments                Naive 8wo females, pooled, right hemisphere. 5...   \n",
       "cDNA_Lib_Ok                                                             Y   \n",
       "ngperul_cDNA                                                         26.1   \n",
       "Avesizebp_cDNAlib                                                           \n",
       "Date                                                                        \n",
       "cDNAul                                                                      \n",
       "LIbConstructionComment                                                      \n",
       "ngperul_seqlib                                                       17.6   \n",
       "lengthbp_seqlib                                                     461.0   \n",
       "Sample_Index                                                           G8   \n",
       "cell_class                                                           GABA   \n",
       "cluster_label                                                           1   \n",
       "markers                                                         Chat-Isl1   \n",
       "\n",
       "                                               ATGACCATCGTGAGAG-1_10X51_1  \\\n",
       "Serial_Number                                                        97.0   \n",
       "Date_Captured                                                    05/31/20   \n",
       "Species                                                                Mm   \n",
       "Transcriptome                                                        Mm10   \n",
       "Strain                                                            C57Bl/6   \n",
       "Project                                                           Dimorph   \n",
       "Group                                                           Breeder-F   \n",
       "ChipID                                                              10X51   \n",
       "SampleID                                                          10X51_1   \n",
       "DonorID                                                           DI-B1-F   \n",
       "Age                                                                11-12w   \n",
       "Num_Pooled_Animals                                                    1.0   \n",
       "Sex                                                                     F   \n",
       "Tissue                                                                MeA   \n",
       "Cell_Conc                                                          1000.0   \n",
       "Target_Num_Cells                                                   5000.0   \n",
       "PCR_Cycles                                                           13.0   \n",
       "Comments                Breeders B1, 8pups DOB 27.5 (P4). perfusion aC...   \n",
       "cDNA_Lib_Ok                                                                 \n",
       "ngperul_cDNA                                                         32.5   \n",
       "Avesizebp_cDNAlib                                                           \n",
       "Date                                                                        \n",
       "cDNAul                                                                      \n",
       "LIbConstructionComment                                                      \n",
       "ngperul_seqlib                                                       19.2   \n",
       "lengthbp_seqlib                                                     478.0   \n",
       "Sample_Index                                                           G1   \n",
       "cell_class                                                           GABA   \n",
       "cluster_label                                                           1   \n",
       "markers                                                         Chat-Isl1   \n",
       "\n",
       "                                               AACCCAAAGAAATGGG-1_10X35_2  \\\n",
       "Serial_Number                                                        64.0   \n",
       "Date_Captured                                                  2019-11-20   \n",
       "Species                                                                Mm   \n",
       "Transcriptome                                                        Mm10   \n",
       "Strain                                                                      \n",
       "Project                                                           Dimorph   \n",
       "Group                                                             Naïve-F   \n",
       "ChipID                                                              10X35   \n",
       "SampleID                                                          10X35_2   \n",
       "DonorID                                                           DI1,DI2   \n",
       "Age                                                                    8w   \n",
       "Num_Pooled_Animals                                                    2.0   \n",
       "Sex                                                                     F   \n",
       "Tissue                                                              MeA_R   \n",
       "Cell_Conc                                                          1000.0   \n",
       "Target_Num_Cells                                                   5000.0   \n",
       "PCR_Cycles                                                           13.0   \n",
       "Comments                Naive 8wo females, pooled, right hemisphere. 5...   \n",
       "cDNA_Lib_Ok                                                             Y   \n",
       "ngperul_cDNA                                                         26.1   \n",
       "Avesizebp_cDNAlib                                                           \n",
       "Date                                                                        \n",
       "cDNAul                                                                      \n",
       "LIbConstructionComment                                                      \n",
       "ngperul_seqlib                                                       17.6   \n",
       "lengthbp_seqlib                                                     461.0   \n",
       "Sample_Index                                                           G8   \n",
       "cell_class                                                           GABA   \n",
       "cluster_label                                                           1   \n",
       "markers                                                         Chat-Isl1   \n",
       "\n",
       "                                               GAGGGATGTATGCTAC-1_10X35_1  \\\n",
       "Serial_Number                                                        63.0   \n",
       "Date_Captured                                                  2019-11-20   \n",
       "Species                                                                Mm   \n",
       "Transcriptome                                                        Mm10   \n",
       "Strain                                                                      \n",
       "Project                                                           Dimorph   \n",
       "Group                                                             Naïve-F   \n",
       "ChipID                                                              10X35   \n",
       "SampleID                                                          10X35_1   \n",
       "DonorID                                                           DI1,DI2   \n",
       "Age                                                                    8w   \n",
       "Num_Pooled_Animals                                                    2.0   \n",
       "Sex                                                                     F   \n",
       "Tissue                                                              MeA_L   \n",
       "Cell_Conc                                                          1000.0   \n",
       "Target_Num_Cells                                                   5000.0   \n",
       "PCR_Cycles                                                           13.0   \n",
       "Comments                Naive 8wo females, pooled, left hemisphere. 5%...   \n",
       "cDNA_Lib_Ok                                                             Y   \n",
       "ngperul_cDNA                                                         23.3   \n",
       "Avesizebp_cDNAlib                                                           \n",
       "Date                                                                        \n",
       "cDNAul                                                                      \n",
       "LIbConstructionComment                                                      \n",
       "ngperul_seqlib                                                       13.6   \n",
       "lengthbp_seqlib                                                     416.0   \n",
       "Sample_Index                                                           D9   \n",
       "cell_class                                                           GABA   \n",
       "cluster_label                                                           1   \n",
       "markers                                                         Chat-Isl1   \n",
       "\n",
       "                                               AAGATAGGTAACACCT-1_10X37_2  \\\n",
       "Serial_Number                                                        68.0   \n",
       "Date_Captured                                                  2019-11-25   \n",
       "Species                                                                Mm   \n",
       "Transcriptome                                                        Mm10   \n",
       "Strain                                                                      \n",
       "Project                                                           Dimorph   \n",
       "Group                                                             Naïve-M   \n",
       "ChipID                                                              10X37   \n",
       "SampleID                                                          10X37_2   \n",
       "DonorID                                                           DI4,DI5   \n",
       "Age                                                                  8.5w   \n",
       "Num_Pooled_Animals                                                    2.0   \n",
       "Sex                                                                     M   \n",
       "Tissue                                                              MeA_R   \n",
       "Cell_Conc                                                          1000.0   \n",
       "Target_Num_Cells                                                   5000.0   \n",
       "PCR_Cycles                                                           13.0   \n",
       "Comments                Naive 8.5wo males, pooled, right hemisphere. 5...   \n",
       "cDNA_Lib_Ok                                                             Y   \n",
       "ngperul_cDNA                                                         24.4   \n",
       "Avesizebp_cDNAlib                                                           \n",
       "Date                                                                        \n",
       "cDNAul                                                                      \n",
       "LIbConstructionComment                                                      \n",
       "ngperul_seqlib                                                       16.0   \n",
       "lengthbp_seqlib                                                     460.0   \n",
       "Sample_Index                                                           F9   \n",
       "cell_class                                                           GABA   \n",
       "cluster_label                                                           1   \n",
       "markers                                                         Chat-Isl1   \n",
       "\n",
       "                                               CAGGGCTGTATCTTCT-1_10X51_1  \\\n",
       "Serial_Number                                                        97.0   \n",
       "Date_Captured                                                    05/31/20   \n",
       "Species                                                                Mm   \n",
       "Transcriptome                                                        Mm10   \n",
       "Strain                                                            C57Bl/6   \n",
       "Project                                                           Dimorph   \n",
       "Group                                                           Breeder-F   \n",
       "ChipID                                                              10X51   \n",
       "SampleID                                                          10X51_1   \n",
       "DonorID                                                           DI-B1-F   \n",
       "Age                                                                11-12w   \n",
       "Num_Pooled_Animals                                                    1.0   \n",
       "Sex                                                                     F   \n",
       "Tissue                                                                MeA   \n",
       "Cell_Conc                                                          1000.0   \n",
       "Target_Num_Cells                                                   5000.0   \n",
       "PCR_Cycles                                                           13.0   \n",
       "Comments                Breeders B1, 8pups DOB 27.5 (P4). perfusion aC...   \n",
       "cDNA_Lib_Ok                                                                 \n",
       "ngperul_cDNA                                                         32.5   \n",
       "Avesizebp_cDNAlib                                                           \n",
       "Date                                                                        \n",
       "cDNAul                                                                      \n",
       "LIbConstructionComment                                                      \n",
       "ngperul_seqlib                                                       19.2   \n",
       "lengthbp_seqlib                                                     478.0   \n",
       "Sample_Index                                                           G1   \n",
       "cell_class                                                           GABA   \n",
       "cluster_label                                                           1   \n",
       "markers                                                         Chat-Isl1   \n",
       "\n",
       "                                        CTGCTCATCCGATCGG-1_10X38_1  \\\n",
       "Serial_Number                                                 69.0   \n",
       "Date_Captured                                           2019-12-01   \n",
       "Species                                                         Mm   \n",
       "Transcriptome                                                 Mm10   \n",
       "Strain                                                               \n",
       "Project                                                    Dimorph   \n",
       "Group                                                      Naïve-F   \n",
       "ChipID                                                       10X38   \n",
       "SampleID                                                   10X38_1   \n",
       "DonorID                                                        DI6   \n",
       "Age                                                             9w   \n",
       "Num_Pooled_Animals                                             1.0   \n",
       "Sex                                                              F   \n",
       "Tissue                                                       MeA_L   \n",
       "Cell_Conc                                                   1000.0   \n",
       "Target_Num_Cells                                            5000.0   \n",
       "PCR_Cycles                                                    13.0   \n",
       "Comments                Naive 9wo female, left hemisphere. 5%opti.   \n",
       "cDNA_Lib_Ok                                                      Y   \n",
       "ngperul_cDNA                                                  21.9   \n",
       "Avesizebp_cDNAlib                                           1870.0   \n",
       "Date                                                                 \n",
       "cDNAul                                                               \n",
       "LIbConstructionComment                                               \n",
       "ngperul_seqlib                                                15.0   \n",
       "lengthbp_seqlib                                              420.0   \n",
       "Sample_Index                                                    B9   \n",
       "cell_class                                                    GABA   \n",
       "cluster_label                                                    1   \n",
       "markers                                                  Chat-Isl1   \n",
       "\n",
       "                                               TGCTCCAAGAGGCGTT-1_10X51_1  \\\n",
       "Serial_Number                                                        97.0   \n",
       "Date_Captured                                                    05/31/20   \n",
       "Species                                                                Mm   \n",
       "Transcriptome                                                        Mm10   \n",
       "Strain                                                            C57Bl/6   \n",
       "Project                                                           Dimorph   \n",
       "Group                                                           Breeder-F   \n",
       "ChipID                                                              10X51   \n",
       "SampleID                                                          10X51_1   \n",
       "DonorID                                                           DI-B1-F   \n",
       "Age                                                                11-12w   \n",
       "Num_Pooled_Animals                                                    1.0   \n",
       "Sex                                                                     F   \n",
       "Tissue                                                                MeA   \n",
       "Cell_Conc                                                          1000.0   \n",
       "Target_Num_Cells                                                   5000.0   \n",
       "PCR_Cycles                                                           13.0   \n",
       "Comments                Breeders B1, 8pups DOB 27.5 (P4). perfusion aC...   \n",
       "cDNA_Lib_Ok                                                                 \n",
       "ngperul_cDNA                                                         32.5   \n",
       "Avesizebp_cDNAlib                                                           \n",
       "Date                                                                        \n",
       "cDNAul                                                                      \n",
       "LIbConstructionComment                                                      \n",
       "ngperul_seqlib                                                       19.2   \n",
       "lengthbp_seqlib                                                     478.0   \n",
       "Sample_Index                                                           G1   \n",
       "cell_class                                                           GABA   \n",
       "cluster_label                                                           1   \n",
       "markers                                                         Chat-Isl1   \n",
       "\n",
       "                                               TTGCCTGTCGTAGGGA-1_10X35_2  \\\n",
       "Serial_Number                                                        64.0   \n",
       "Date_Captured                                                  2019-11-20   \n",
       "Species                                                                Mm   \n",
       "Transcriptome                                                        Mm10   \n",
       "Strain                                                                      \n",
       "Project                                                           Dimorph   \n",
       "Group                                                             Naïve-F   \n",
       "ChipID                                                              10X35   \n",
       "SampleID                                                          10X35_2   \n",
       "DonorID                                                           DI1,DI2   \n",
       "Age                                                                    8w   \n",
       "Num_Pooled_Animals                                                    2.0   \n",
       "Sex                                                                     F   \n",
       "Tissue                                                              MeA_R   \n",
       "Cell_Conc                                                          1000.0   \n",
       "Target_Num_Cells                                                   5000.0   \n",
       "PCR_Cycles                                                           13.0   \n",
       "Comments                Naive 8wo females, pooled, right hemisphere. 5...   \n",
       "cDNA_Lib_Ok                                                             Y   \n",
       "ngperul_cDNA                                                         26.1   \n",
       "Avesizebp_cDNAlib                                                           \n",
       "Date                                                                        \n",
       "cDNAul                                                                      \n",
       "LIbConstructionComment                                                      \n",
       "ngperul_seqlib                                                       17.6   \n",
       "lengthbp_seqlib                                                     461.0   \n",
       "Sample_Index                                                           G8   \n",
       "cell_class                                                           GABA   \n",
       "cluster_label                                                           1   \n",
       "markers                                                         Chat-Isl1   \n",
       "\n",
       "                                               GCAACATAGATCGACG-1_10X51_2  \\\n",
       "Serial_Number                                                        98.0   \n",
       "Date_Captured                                                    05/31/20   \n",
       "Species                                                                Mm   \n",
       "Transcriptome                                                        Mm10   \n",
       "Strain                                                            C57Bl/6   \n",
       "Project                                                           Dimorph   \n",
       "Group                                                           Breeder-F   \n",
       "ChipID                                                              10X51   \n",
       "SampleID                                                          10X51_2   \n",
       "DonorID                                                           DI-B1-F   \n",
       "Age                                                                11-12w   \n",
       "Num_Pooled_Animals                                                    1.0   \n",
       "Sex                                                                     F   \n",
       "Tissue                                                                MeA   \n",
       "Cell_Conc                                                          1000.0   \n",
       "Target_Num_Cells                                                   5000.0   \n",
       "PCR_Cycles                                                           13.0   \n",
       "Comments                Breeders B1, 8pups DOB 27.5 (P4). perfusion aC...   \n",
       "cDNA_Lib_Ok                                                                 \n",
       "ngperul_cDNA                                                         29.8   \n",
       "Avesizebp_cDNAlib                                                           \n",
       "Date                                                                        \n",
       "cDNAul                                                                      \n",
       "LIbConstructionComment                                                      \n",
       "ngperul_seqlib                                                       22.2   \n",
       "lengthbp_seqlib                                                     474.0   \n",
       "Sample_Index                                                           A2   \n",
       "cell_class                                                           GABA   \n",
       "cluster_label                                                           1   \n",
       "markers                                                         Chat-Isl1   \n",
       "\n",
       "                        ...                   CAAGAGGGTCAATCTG-1_10X36_2  \\\n",
       "Serial_Number           ...                                         66.0   \n",
       "Date_Captured           ...                                   2019-11-24   \n",
       "Species                 ...                                           Mm   \n",
       "Transcriptome           ...                                         Mm10   \n",
       "Strain                  ...                                                \n",
       "Project                 ...                                      Dimorph   \n",
       "Group                   ...                                      Naïve-M   \n",
       "ChipID                  ...                                        10X36   \n",
       "SampleID                ...                                      10X36_2   \n",
       "DonorID                 ...                                          DI3   \n",
       "Age                     ...                                         8.5w   \n",
       "Num_Pooled_Animals      ...                                          1.0   \n",
       "Sex                     ...                                            M   \n",
       "Tissue                  ...                                        MeA_R   \n",
       "Cell_Conc               ...                                       1000.0   \n",
       "Target_Num_Cells        ...                                       5000.0   \n",
       "PCR_Cycles              ...                                         13.0   \n",
       "Comments                ...  Naive 8.5wo male, right hemisphere. 5%opti.   \n",
       "cDNA_Lib_Ok             ...                                            Y   \n",
       "ngperul_cDNA            ...                                         16.7   \n",
       "Avesizebp_cDNAlib       ...                                                \n",
       "Date                    ...                                                \n",
       "cDNAul                  ...                                                \n",
       "LIbConstructionComment  ...                                                \n",
       "ngperul_seqlib          ...                                         12.1   \n",
       "lengthbp_seqlib         ...                                        486.0   \n",
       "Sample_Index            ...                                           A9   \n",
       "cell_class              ...                                  Nonneuronal   \n",
       "cluster_label           ...                                           11   \n",
       "markers                 ...                                         OL_2   \n",
       "\n",
       "                                               TATACCTAGAGTAACT-1_10X35_1  \\\n",
       "Serial_Number                                                        63.0   \n",
       "Date_Captured                                                  2019-11-20   \n",
       "Species                                                                Mm   \n",
       "Transcriptome                                                        Mm10   \n",
       "Strain                                                                      \n",
       "Project                                                           Dimorph   \n",
       "Group                                                             Naïve-F   \n",
       "ChipID                                                              10X35   \n",
       "SampleID                                                          10X35_1   \n",
       "DonorID                                                           DI1,DI2   \n",
       "Age                                                                    8w   \n",
       "Num_Pooled_Animals                                                    2.0   \n",
       "Sex                                                                     F   \n",
       "Tissue                                                              MeA_L   \n",
       "Cell_Conc                                                          1000.0   \n",
       "Target_Num_Cells                                                   5000.0   \n",
       "PCR_Cycles                                                           13.0   \n",
       "Comments                Naive 8wo females, pooled, left hemisphere. 5%...   \n",
       "cDNA_Lib_Ok                                                             Y   \n",
       "ngperul_cDNA                                                         23.3   \n",
       "Avesizebp_cDNAlib                                                           \n",
       "Date                                                                        \n",
       "cDNAul                                                                      \n",
       "LIbConstructionComment                                                      \n",
       "ngperul_seqlib                                                       13.6   \n",
       "lengthbp_seqlib                                                     416.0   \n",
       "Sample_Index                                                           D9   \n",
       "cell_class                                                    Nonneuronal   \n",
       "cluster_label                                                          11   \n",
       "markers                                                              OL_2   \n",
       "\n",
       "                                               GCACGTGGTTATTCCT-1_10X37_1  \\\n",
       "Serial_Number                                                        67.0   \n",
       "Date_Captured                                                  2019-11-25   \n",
       "Species                                                                Mm   \n",
       "Transcriptome                                                        Mm10   \n",
       "Strain                                                                      \n",
       "Project                                                           Dimorph   \n",
       "Group                                                             Naïve-M   \n",
       "ChipID                                                              10X37   \n",
       "SampleID                                                          10X37_1   \n",
       "DonorID                                                           DI4,DI5   \n",
       "Age                                                                  8.5w   \n",
       "Num_Pooled_Animals                                                    2.0   \n",
       "Sex                                                                     M   \n",
       "Tissue                                                              MeA_L   \n",
       "Cell_Conc                                                          1000.0   \n",
       "Target_Num_Cells                                                   5000.0   \n",
       "PCR_Cycles                                                           13.0   \n",
       "Comments                Naive 8.5wo males, pooled, left hemisphere. 5%...   \n",
       "cDNA_Lib_Ok                                                             Y   \n",
       "ngperul_cDNA                                                         17.0   \n",
       "Avesizebp_cDNAlib                                                           \n",
       "Date                                                                        \n",
       "cDNAul                                                                      \n",
       "LIbConstructionComment                                                      \n",
       "ngperul_seqlib                                                       15.9   \n",
       "lengthbp_seqlib                                                     440.0   \n",
       "Sample_Index                                                           E9   \n",
       "cell_class                                                    Nonneuronal   \n",
       "cluster_label                                                          11   \n",
       "markers                                                              OL_2   \n",
       "\n",
       "                                        GATGATCTCAAGCCTA-1_10X36_1  \\\n",
       "Serial_Number                                                 65.0   \n",
       "Date_Captured                                           2019-11-24   \n",
       "Species                                                         Mm   \n",
       "Transcriptome                                                 Mm10   \n",
       "Strain                                                               \n",
       "Project                                                    Dimorph   \n",
       "Group                                                      Naïve-M   \n",
       "ChipID                                                       10X36   \n",
       "SampleID                                                   10X36_1   \n",
       "DonorID                                                        DI3   \n",
       "Age                                                           8.5w   \n",
       "Num_Pooled_Animals                                             1.0   \n",
       "Sex                                                              M   \n",
       "Tissue                                                       MeA_L   \n",
       "Cell_Conc                                                   1000.0   \n",
       "Target_Num_Cells                                            5000.0   \n",
       "PCR_Cycles                                                    13.0   \n",
       "Comments                Naive 8.5wo male, left hemisphere. 5%opti.   \n",
       "cDNA_Lib_Ok                                                      Y   \n",
       "ngperul_cDNA                                                  26.4   \n",
       "Avesizebp_cDNAlib                                                    \n",
       "Date                                                                 \n",
       "cDNAul                                                               \n",
       "LIbConstructionComment                                               \n",
       "ngperul_seqlib                                                11.4   \n",
       "lengthbp_seqlib                                              476.0   \n",
       "Sample_Index                                                    H8   \n",
       "cell_class                                             Nonneuronal   \n",
       "cluster_label                                                   11   \n",
       "markers                                                       OL_2   \n",
       "\n",
       "                                         GGTGTCGAGAAGCTGC-1_10X36_2  \\\n",
       "Serial_Number                                                  66.0   \n",
       "Date_Captured                                            2019-11-24   \n",
       "Species                                                          Mm   \n",
       "Transcriptome                                                  Mm10   \n",
       "Strain                                                                \n",
       "Project                                                     Dimorph   \n",
       "Group                                                       Naïve-M   \n",
       "ChipID                                                        10X36   \n",
       "SampleID                                                    10X36_2   \n",
       "DonorID                                                         DI3   \n",
       "Age                                                            8.5w   \n",
       "Num_Pooled_Animals                                              1.0   \n",
       "Sex                                                               M   \n",
       "Tissue                                                        MeA_R   \n",
       "Cell_Conc                                                    1000.0   \n",
       "Target_Num_Cells                                             5000.0   \n",
       "PCR_Cycles                                                     13.0   \n",
       "Comments                Naive 8.5wo male, right hemisphere. 5%opti.   \n",
       "cDNA_Lib_Ok                                                       Y   \n",
       "ngperul_cDNA                                                   16.7   \n",
       "Avesizebp_cDNAlib                                                     \n",
       "Date                                                                  \n",
       "cDNAul                                                                \n",
       "LIbConstructionComment                                                \n",
       "ngperul_seqlib                                                 12.1   \n",
       "lengthbp_seqlib                                               486.0   \n",
       "Sample_Index                                                     A9   \n",
       "cell_class                                              Nonneuronal   \n",
       "cluster_label                                                    11   \n",
       "markers                                                        OL_2   \n",
       "\n",
       "                                               GAGAGGTCACATCATG-1_10X37_2  \\\n",
       "Serial_Number                                                        68.0   \n",
       "Date_Captured                                                  2019-11-25   \n",
       "Species                                                                Mm   \n",
       "Transcriptome                                                        Mm10   \n",
       "Strain                                                                      \n",
       "Project                                                           Dimorph   \n",
       "Group                                                             Naïve-M   \n",
       "ChipID                                                              10X37   \n",
       "SampleID                                                          10X37_2   \n",
       "DonorID                                                           DI4,DI5   \n",
       "Age                                                                  8.5w   \n",
       "Num_Pooled_Animals                                                    2.0   \n",
       "Sex                                                                     M   \n",
       "Tissue                                                              MeA_R   \n",
       "Cell_Conc                                                          1000.0   \n",
       "Target_Num_Cells                                                   5000.0   \n",
       "PCR_Cycles                                                           13.0   \n",
       "Comments                Naive 8.5wo males, pooled, right hemisphere. 5...   \n",
       "cDNA_Lib_Ok                                                             Y   \n",
       "ngperul_cDNA                                                         24.4   \n",
       "Avesizebp_cDNAlib                                                           \n",
       "Date                                                                        \n",
       "cDNAul                                                                      \n",
       "LIbConstructionComment                                                      \n",
       "ngperul_seqlib                                                       16.0   \n",
       "lengthbp_seqlib                                                     460.0   \n",
       "Sample_Index                                                           F9   \n",
       "cell_class                                                    Nonneuronal   \n",
       "cluster_label                                                          11   \n",
       "markers                                                              OL_2   \n",
       "\n",
       "                                               ATCAGGTTCCACTAGA-1_10X35_2  \\\n",
       "Serial_Number                                                        64.0   \n",
       "Date_Captured                                                  2019-11-20   \n",
       "Species                                                                Mm   \n",
       "Transcriptome                                                        Mm10   \n",
       "Strain                                                                      \n",
       "Project                                                           Dimorph   \n",
       "Group                                                             Naïve-F   \n",
       "ChipID                                                              10X35   \n",
       "SampleID                                                          10X35_2   \n",
       "DonorID                                                           DI1,DI2   \n",
       "Age                                                                    8w   \n",
       "Num_Pooled_Animals                                                    2.0   \n",
       "Sex                                                                     F   \n",
       "Tissue                                                              MeA_R   \n",
       "Cell_Conc                                                          1000.0   \n",
       "Target_Num_Cells                                                   5000.0   \n",
       "PCR_Cycles                                                           13.0   \n",
       "Comments                Naive 8wo females, pooled, right hemisphere. 5...   \n",
       "cDNA_Lib_Ok                                                             Y   \n",
       "ngperul_cDNA                                                         26.1   \n",
       "Avesizebp_cDNAlib                                                           \n",
       "Date                                                                        \n",
       "cDNAul                                                                      \n",
       "LIbConstructionComment                                                      \n",
       "ngperul_seqlib                                                       17.6   \n",
       "lengthbp_seqlib                                                     461.0   \n",
       "Sample_Index                                                           G8   \n",
       "cell_class                                                    Nonneuronal   \n",
       "cluster_label                                                          11   \n",
       "markers                                                              OL_2   \n",
       "\n",
       "                                               TCCACGTAGAGCATCG-1_10X37_1  \\\n",
       "Serial_Number                                                        67.0   \n",
       "Date_Captured                                                  2019-11-25   \n",
       "Species                                                                Mm   \n",
       "Transcriptome                                                        Mm10   \n",
       "Strain                                                                      \n",
       "Project                                                           Dimorph   \n",
       "Group                                                             Naïve-M   \n",
       "ChipID                                                              10X37   \n",
       "SampleID                                                          10X37_1   \n",
       "DonorID                                                           DI4,DI5   \n",
       "Age                                                                  8.5w   \n",
       "Num_Pooled_Animals                                                    2.0   \n",
       "Sex                                                                     M   \n",
       "Tissue                                                              MeA_L   \n",
       "Cell_Conc                                                          1000.0   \n",
       "Target_Num_Cells                                                   5000.0   \n",
       "PCR_Cycles                                                           13.0   \n",
       "Comments                Naive 8.5wo males, pooled, left hemisphere. 5%...   \n",
       "cDNA_Lib_Ok                                                             Y   \n",
       "ngperul_cDNA                                                         17.0   \n",
       "Avesizebp_cDNAlib                                                           \n",
       "Date                                                                        \n",
       "cDNAul                                                                      \n",
       "LIbConstructionComment                                                      \n",
       "ngperul_seqlib                                                       15.9   \n",
       "lengthbp_seqlib                                                     440.0   \n",
       "Sample_Index                                                           E9   \n",
       "cell_class                                                    Nonneuronal   \n",
       "cluster_label                                                          11   \n",
       "markers                                                              OL_2   \n",
       "\n",
       "                                               CTGCGAGCACTACGGC-1_10X35_2  \\\n",
       "Serial_Number                                                        64.0   \n",
       "Date_Captured                                                  2019-11-20   \n",
       "Species                                                                Mm   \n",
       "Transcriptome                                                        Mm10   \n",
       "Strain                                                                      \n",
       "Project                                                           Dimorph   \n",
       "Group                                                             Naïve-F   \n",
       "ChipID                                                              10X35   \n",
       "SampleID                                                          10X35_2   \n",
       "DonorID                                                           DI1,DI2   \n",
       "Age                                                                    8w   \n",
       "Num_Pooled_Animals                                                    2.0   \n",
       "Sex                                                                     F   \n",
       "Tissue                                                              MeA_R   \n",
       "Cell_Conc                                                          1000.0   \n",
       "Target_Num_Cells                                                   5000.0   \n",
       "PCR_Cycles                                                           13.0   \n",
       "Comments                Naive 8wo females, pooled, right hemisphere. 5...   \n",
       "cDNA_Lib_Ok                                                             Y   \n",
       "ngperul_cDNA                                                         26.1   \n",
       "Avesizebp_cDNAlib                                                           \n",
       "Date                                                                        \n",
       "cDNAul                                                                      \n",
       "LIbConstructionComment                                                      \n",
       "ngperul_seqlib                                                       17.6   \n",
       "lengthbp_seqlib                                                     461.0   \n",
       "Sample_Index                                                           G8   \n",
       "cell_class                                                    Nonneuronal   \n",
       "cluster_label                                                          11   \n",
       "markers                                                              OL_2   \n",
       "\n",
       "                                               CTCTGGTTCGAAGCAG-1_10X37_1  \n",
       "Serial_Number                                                        67.0  \n",
       "Date_Captured                                                  2019-11-25  \n",
       "Species                                                                Mm  \n",
       "Transcriptome                                                        Mm10  \n",
       "Strain                                                                     \n",
       "Project                                                           Dimorph  \n",
       "Group                                                             Naïve-M  \n",
       "ChipID                                                              10X37  \n",
       "SampleID                                                          10X37_1  \n",
       "DonorID                                                           DI4,DI5  \n",
       "Age                                                                  8.5w  \n",
       "Num_Pooled_Animals                                                    2.0  \n",
       "Sex                                                                     M  \n",
       "Tissue                                                              MeA_L  \n",
       "Cell_Conc                                                          1000.0  \n",
       "Target_Num_Cells                                                   5000.0  \n",
       "PCR_Cycles                                                           13.0  \n",
       "Comments                Naive 8.5wo males, pooled, left hemisphere. 5%...  \n",
       "cDNA_Lib_Ok                                                             Y  \n",
       "ngperul_cDNA                                                         17.0  \n",
       "Avesizebp_cDNAlib                                                          \n",
       "Date                                                                       \n",
       "cDNAul                                                                     \n",
       "LIbConstructionComment                                                     \n",
       "ngperul_seqlib                                                       15.9  \n",
       "lengthbp_seqlib                                                     440.0  \n",
       "Sample_Index                                                           E9  \n",
       "cell_class                                                    Nonneuronal  \n",
       "cluster_label                                                          11  \n",
       "markers                                                              OL_2  \n",
       "\n",
       "[30 rows x 26824 columns]"
      ]
     },
     "execution_count": 284,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "combined_metadata = pd.concat([gaba_metadata,vglut1_metadata,vglut2_metadata,nn_metadata],axis=1)\n",
    "combined_metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b04f156b",
   "metadata": {},
   "outputs": [],
   "source": [
    "#load original expression data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "dd5d47de",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_orig = pd.read_feather('/bigdata/isaac/df_orig.feather')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 98,
   "id": "1d77e58b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>AAACCCAAGAGCATAT-1_10X51_2</th>\n",
       "      <th>AAACCCAAGCCTCAGC-1_10X51_2</th>\n",
       "      <th>AAACCCAAGTACTGTC-1_10X51_2</th>\n",
       "      <th>AAACCCACACAACGTT-1_10X51_2</th>\n",
       "      <th>AAACCCAGTAACATAG-1_10X51_2</th>\n",
       "      <th>AAACCCATCACTAGCA-1_10X51_2</th>\n",
       "      <th>AAACGAACACTCCGAG-1_10X51_2</th>\n",
       "      <th>AAACGAACAGCGCGTT-1_10X51_2</th>\n",
       "      <th>AAACGAACAGGACTTT-1_10X51_2</th>\n",
       "      <th>AAACGAAGTGAATAAC-1_10X51_2</th>\n",
       "      <th>...</th>\n",
       "      <th>TTTGGTTTCCACGAAT-1_10X37_2</th>\n",
       "      <th>TTTGGTTTCTTTCTAG-1_10X37_2</th>\n",
       "      <th>TTTGTTGAGAAATTCG-1_10X37_2</th>\n",
       "      <th>TTTGTTGAGTGCTCGC-1_10X37_2</th>\n",
       "      <th>TTTGTTGCAAGGTCAG-1_10X37_2</th>\n",
       "      <th>TTTGTTGCACCGCTGA-1_10X37_2</th>\n",
       "      <th>TTTGTTGGTCCATAGT-1_10X37_2</th>\n",
       "      <th>TTTGTTGGTTGCGAAG-1_10X37_2</th>\n",
       "      <th>TTTGTTGTCGAAGGAC-1_10X37_2</th>\n",
       "      <th>TTTGTTGTCTCTGCTG-1_10X37_2</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0610007P14Rik</th>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0610009B22Rik</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0610009L18Rik</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0610009O20Rik</th>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0610010F05Rik</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mt-Nd3</th>\n",
       "      <td>31</td>\n",
       "      <td>24</td>\n",
       "      <td>7</td>\n",
       "      <td>13</td>\n",
       "      <td>7</td>\n",
       "      <td>11</td>\n",
       "      <td>12</td>\n",
       "      <td>30</td>\n",
       "      <td>29</td>\n",
       "      <td>14</td>\n",
       "      <td>...</td>\n",
       "      <td>10</td>\n",
       "      <td>6</td>\n",
       "      <td>34</td>\n",
       "      <td>15</td>\n",
       "      <td>33</td>\n",
       "      <td>12</td>\n",
       "      <td>2</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mt-Nd4</th>\n",
       "      <td>64</td>\n",
       "      <td>90</td>\n",
       "      <td>12</td>\n",
       "      <td>50</td>\n",
       "      <td>16</td>\n",
       "      <td>58</td>\n",
       "      <td>71</td>\n",
       "      <td>87</td>\n",
       "      <td>88</td>\n",
       "      <td>81</td>\n",
       "      <td>...</td>\n",
       "      <td>90</td>\n",
       "      <td>40</td>\n",
       "      <td>170</td>\n",
       "      <td>69</td>\n",
       "      <td>130</td>\n",
       "      <td>67</td>\n",
       "      <td>33</td>\n",
       "      <td>31</td>\n",
       "      <td>38</td>\n",
       "      <td>31</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mt-Nd4l</th>\n",
       "      <td>8</td>\n",
       "      <td>12</td>\n",
       "      <td>2</td>\n",
       "      <td>5</td>\n",
       "      <td>4</td>\n",
       "      <td>10</td>\n",
       "      <td>7</td>\n",
       "      <td>5</td>\n",
       "      <td>12</td>\n",
       "      <td>9</td>\n",
       "      <td>...</td>\n",
       "      <td>14</td>\n",
       "      <td>4</td>\n",
       "      <td>13</td>\n",
       "      <td>4</td>\n",
       "      <td>8</td>\n",
       "      <td>5</td>\n",
       "      <td>9</td>\n",
       "      <td>7</td>\n",
       "      <td>3</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mt-Nd5</th>\n",
       "      <td>14</td>\n",
       "      <td>20</td>\n",
       "      <td>6</td>\n",
       "      <td>9</td>\n",
       "      <td>1</td>\n",
       "      <td>12</td>\n",
       "      <td>8</td>\n",
       "      <td>10</td>\n",
       "      <td>11</td>\n",
       "      <td>10</td>\n",
       "      <td>...</td>\n",
       "      <td>17</td>\n",
       "      <td>5</td>\n",
       "      <td>24</td>\n",
       "      <td>4</td>\n",
       "      <td>18</td>\n",
       "      <td>13</td>\n",
       "      <td>6</td>\n",
       "      <td>2</td>\n",
       "      <td>4</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mt-Nd6</th>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>27933 rows × 79613 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "               AAACCCAAGAGCATAT-1_10X51_2  AAACCCAAGCCTCAGC-1_10X51_2  \\\n",
       "gene                                                                    \n",
       "0610007P14Rik                           0                           2   \n",
       "0610009B22Rik                           0                           0   \n",
       "0610009L18Rik                           0                           0   \n",
       "0610009O20Rik                           1                           0   \n",
       "0610010F05Rik                           0                           0   \n",
       "...                                   ...                         ...   \n",
       "mt-Nd3                                 31                          24   \n",
       "mt-Nd4                                 64                          90   \n",
       "mt-Nd4l                                 8                          12   \n",
       "mt-Nd5                                 14                          20   \n",
       "mt-Nd6                                  1                           2   \n",
       "\n",
       "               AAACCCAAGTACTGTC-1_10X51_2  AAACCCACACAACGTT-1_10X51_2  \\\n",
       "gene                                                                    \n",
       "0610007P14Rik                           0                           2   \n",
       "0610009B22Rik                           0                           1   \n",
       "0610009L18Rik                           0                           0   \n",
       "0610009O20Rik                           0                           0   \n",
       "0610010F05Rik                           1                           0   \n",
       "...                                   ...                         ...   \n",
       "mt-Nd3                                  7                          13   \n",
       "mt-Nd4                                 12                          50   \n",
       "mt-Nd4l                                 2                           5   \n",
       "mt-Nd5                                  6                           9   \n",
       "mt-Nd6                                  0                           1   \n",
       "\n",
       "               AAACCCAGTAACATAG-1_10X51_2  AAACCCATCACTAGCA-1_10X51_2  \\\n",
       "gene                                                                    \n",
       "0610007P14Rik                           0                           0   \n",
       "0610009B22Rik                           0                           2   \n",
       "0610009L18Rik                           0                           0   \n",
       "0610009O20Rik                           0                           1   \n",
       "0610010F05Rik                           0                           0   \n",
       "...                                   ...                         ...   \n",
       "mt-Nd3                                  7                          11   \n",
       "mt-Nd4                                 16                          58   \n",
       "mt-Nd4l                                 4                          10   \n",
       "mt-Nd5                                  1                          12   \n",
       "mt-Nd6                                  0                           0   \n",
       "\n",
       "               AAACGAACACTCCGAG-1_10X51_2  AAACGAACAGCGCGTT-1_10X51_2  \\\n",
       "gene                                                                    \n",
       "0610007P14Rik                           1                           3   \n",
       "0610009B22Rik                           2                           0   \n",
       "0610009L18Rik                           1                           1   \n",
       "0610009O20Rik                           1                           0   \n",
       "0610010F05Rik                           2                           1   \n",
       "...                                   ...                         ...   \n",
       "mt-Nd3                                 12                          30   \n",
       "mt-Nd4                                 71                          87   \n",
       "mt-Nd4l                                 7                           5   \n",
       "mt-Nd5                                  8                          10   \n",
       "mt-Nd6                                  3                           3   \n",
       "\n",
       "               AAACGAACAGGACTTT-1_10X51_2  AAACGAAGTGAATAAC-1_10X51_2  ...  \\\n",
       "gene                                                                   ...   \n",
       "0610007P14Rik                           2                           0  ...   \n",
       "0610009B22Rik                           1                           0  ...   \n",
       "0610009L18Rik                           0                           0  ...   \n",
       "0610009O20Rik                           0                           0  ...   \n",
       "0610010F05Rik                           1                           0  ...   \n",
       "...                                   ...                         ...  ...   \n",
       "mt-Nd3                                 29                          14  ...   \n",
       "mt-Nd4                                 88                          81  ...   \n",
       "mt-Nd4l                                12                           9  ...   \n",
       "mt-Nd5                                 11                          10  ...   \n",
       "mt-Nd6                                  2                           1  ...   \n",
       "\n",
       "               TTTGGTTTCCACGAAT-1_10X37_2  TTTGGTTTCTTTCTAG-1_10X37_2  \\\n",
       "gene                                                                    \n",
       "0610007P14Rik                           1                           0   \n",
       "0610009B22Rik                           0                           0   \n",
       "0610009L18Rik                           0                           0   \n",
       "0610009O20Rik                           0                           0   \n",
       "0610010F05Rik                           1                           0   \n",
       "...                                   ...                         ...   \n",
       "mt-Nd3                                 10                           6   \n",
       "mt-Nd4                                 90                          40   \n",
       "mt-Nd4l                                14                           4   \n",
       "mt-Nd5                                 17                           5   \n",
       "mt-Nd6                                  1                           1   \n",
       "\n",
       "               TTTGTTGAGAAATTCG-1_10X37_2  TTTGTTGAGTGCTCGC-1_10X37_2  \\\n",
       "gene                                                                    \n",
       "0610007P14Rik                           1                           0   \n",
       "0610009B22Rik                           0                           0   \n",
       "0610009L18Rik                           0                           0   \n",
       "0610009O20Rik                           0                           0   \n",
       "0610010F05Rik                           2                           0   \n",
       "...                                   ...                         ...   \n",
       "mt-Nd3                                 34                          15   \n",
       "mt-Nd4                                170                          69   \n",
       "mt-Nd4l                                13                           4   \n",
       "mt-Nd5                                 24                           4   \n",
       "mt-Nd6                                  1                           2   \n",
       "\n",
       "               TTTGTTGCAAGGTCAG-1_10X37_2  TTTGTTGCACCGCTGA-1_10X37_2  \\\n",
       "gene                                                                    \n",
       "0610007P14Rik                           2                           1   \n",
       "0610009B22Rik                           2                           0   \n",
       "0610009L18Rik                           1                           0   \n",
       "0610009O20Rik                           0                           0   \n",
       "0610010F05Rik                           1                           0   \n",
       "...                                   ...                         ...   \n",
       "mt-Nd3                                 33                          12   \n",
       "mt-Nd4                                130                          67   \n",
       "mt-Nd4l                                 8                           5   \n",
       "mt-Nd5                                 18                          13   \n",
       "mt-Nd6                                  1                           1   \n",
       "\n",
       "               TTTGTTGGTCCATAGT-1_10X37_2  TTTGTTGGTTGCGAAG-1_10X37_2  \\\n",
       "gene                                                                    \n",
       "0610007P14Rik                           1                           0   \n",
       "0610009B22Rik                           0                           0   \n",
       "0610009L18Rik                           1                           0   \n",
       "0610009O20Rik                           0                           0   \n",
       "0610010F05Rik                           1                           0   \n",
       "...                                   ...                         ...   \n",
       "mt-Nd3                                  2                           4   \n",
       "mt-Nd4                                 33                          31   \n",
       "mt-Nd4l                                 9                           7   \n",
       "mt-Nd5                                  6                           2   \n",
       "mt-Nd6                                  0                           0   \n",
       "\n",
       "               TTTGTTGTCGAAGGAC-1_10X37_2  TTTGTTGTCTCTGCTG-1_10X37_2  \n",
       "gene                                                                   \n",
       "0610007P14Rik                           0                           0  \n",
       "0610009B22Rik                           0                           0  \n",
       "0610009L18Rik                           0                           0  \n",
       "0610009O20Rik                           0                           0  \n",
       "0610010F05Rik                           0                           0  \n",
       "...                                   ...                         ...  \n",
       "mt-Nd3                                  4                           1  \n",
       "mt-Nd4                                 38                          31  \n",
       "mt-Nd4l                                 3                           5  \n",
       "mt-Nd5                                  4                           6  \n",
       "mt-Nd6                                  0                           0  \n",
       "\n",
       "[27933 rows x 79613 columns]"
      ]
     },
     "execution_count": 98,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_orig"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 100,
   "id": "77b425c8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CAACAACAGACATATG-1_10X35_2</th>\n",
       "      <th>ATGACCATCGTGAGAG-1_10X51_1</th>\n",
       "      <th>AACCCAAAGAAATGGG-1_10X35_2</th>\n",
       "      <th>GAGGGATGTATGCTAC-1_10X35_1</th>\n",
       "      <th>AAGATAGGTAACACCT-1_10X37_2</th>\n",
       "      <th>CAGGGCTGTATCTTCT-1_10X51_1</th>\n",
       "      <th>CTGCTCATCCGATCGG-1_10X38_1</th>\n",
       "      <th>TGCTCCAAGAGGCGTT-1_10X51_1</th>\n",
       "      <th>TTGCCTGTCGTAGGGA-1_10X35_2</th>\n",
       "      <th>GCAACATAGATCGACG-1_10X51_2</th>\n",
       "      <th>...</th>\n",
       "      <th>CAAGAGGGTCAATCTG-1_10X36_2</th>\n",
       "      <th>TATACCTAGAGTAACT-1_10X35_1</th>\n",
       "      <th>GCACGTGGTTATTCCT-1_10X37_1</th>\n",
       "      <th>GATGATCTCAAGCCTA-1_10X36_1</th>\n",
       "      <th>GGTGTCGAGAAGCTGC-1_10X36_2</th>\n",
       "      <th>GAGAGGTCACATCATG-1_10X37_2</th>\n",
       "      <th>ATCAGGTTCCACTAGA-1_10X35_2</th>\n",
       "      <th>TCCACGTAGAGCATCG-1_10X37_1</th>\n",
       "      <th>CTGCGAGCACTACGGC-1_10X35_2</th>\n",
       "      <th>CTCTGGTTCGAAGCAG-1_10X37_1</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0610007P14Rik</th>\n",
       "      <td>3</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0610009B22Rik</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0610009L18Rik</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0610009O20Rik</th>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0610010F05Rik</th>\n",
       "      <td>5</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>6</td>\n",
       "      <td>7</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mt-Nd3</th>\n",
       "      <td>49</td>\n",
       "      <td>13</td>\n",
       "      <td>19</td>\n",
       "      <td>42</td>\n",
       "      <td>26</td>\n",
       "      <td>9</td>\n",
       "      <td>14</td>\n",
       "      <td>9</td>\n",
       "      <td>12</td>\n",
       "      <td>26</td>\n",
       "      <td>...</td>\n",
       "      <td>16</td>\n",
       "      <td>12</td>\n",
       "      <td>33</td>\n",
       "      <td>26</td>\n",
       "      <td>16</td>\n",
       "      <td>32</td>\n",
       "      <td>16</td>\n",
       "      <td>24</td>\n",
       "      <td>22</td>\n",
       "      <td>39</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mt-Nd4</th>\n",
       "      <td>252</td>\n",
       "      <td>74</td>\n",
       "      <td>233</td>\n",
       "      <td>591</td>\n",
       "      <td>340</td>\n",
       "      <td>82</td>\n",
       "      <td>43</td>\n",
       "      <td>89</td>\n",
       "      <td>119</td>\n",
       "      <td>39</td>\n",
       "      <td>...</td>\n",
       "      <td>230</td>\n",
       "      <td>71</td>\n",
       "      <td>147</td>\n",
       "      <td>148</td>\n",
       "      <td>75</td>\n",
       "      <td>106</td>\n",
       "      <td>59</td>\n",
       "      <td>95</td>\n",
       "      <td>87</td>\n",
       "      <td>126</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mt-Nd4l</th>\n",
       "      <td>22</td>\n",
       "      <td>2</td>\n",
       "      <td>7</td>\n",
       "      <td>21</td>\n",
       "      <td>16</td>\n",
       "      <td>4</td>\n",
       "      <td>3</td>\n",
       "      <td>2</td>\n",
       "      <td>9</td>\n",
       "      <td>8</td>\n",
       "      <td>...</td>\n",
       "      <td>8</td>\n",
       "      <td>6</td>\n",
       "      <td>9</td>\n",
       "      <td>4</td>\n",
       "      <td>6</td>\n",
       "      <td>14</td>\n",
       "      <td>5</td>\n",
       "      <td>5</td>\n",
       "      <td>4</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mt-Nd5</th>\n",
       "      <td>34</td>\n",
       "      <td>9</td>\n",
       "      <td>27</td>\n",
       "      <td>74</td>\n",
       "      <td>26</td>\n",
       "      <td>10</td>\n",
       "      <td>5</td>\n",
       "      <td>15</td>\n",
       "      <td>9</td>\n",
       "      <td>7</td>\n",
       "      <td>...</td>\n",
       "      <td>20</td>\n",
       "      <td>3</td>\n",
       "      <td>14</td>\n",
       "      <td>9</td>\n",
       "      <td>6</td>\n",
       "      <td>13</td>\n",
       "      <td>7</td>\n",
       "      <td>11</td>\n",
       "      <td>7</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mt-Nd6</th>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>27933 rows × 26824 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "               CAACAACAGACATATG-1_10X35_2  ATGACCATCGTGAGAG-1_10X51_1  \\\n",
       "gene                                                                    \n",
       "0610007P14Rik                           3                           2   \n",
       "0610009B22Rik                           1                           1   \n",
       "0610009L18Rik                           0                           0   \n",
       "0610009O20Rik                           2                           1   \n",
       "0610010F05Rik                           5                           0   \n",
       "...                                   ...                         ...   \n",
       "mt-Nd3                                 49                          13   \n",
       "mt-Nd4                                252                          74   \n",
       "mt-Nd4l                                22                           2   \n",
       "mt-Nd5                                 34                           9   \n",
       "mt-Nd6                                  4                           0   \n",
       "\n",
       "               AACCCAAAGAAATGGG-1_10X35_2  GAGGGATGTATGCTAC-1_10X35_1  \\\n",
       "gene                                                                    \n",
       "0610007P14Rik                           2                           0   \n",
       "0610009B22Rik                           0                           2   \n",
       "0610009L18Rik                           0                           2   \n",
       "0610009O20Rik                           1                           0   \n",
       "0610010F05Rik                           1                           6   \n",
       "...                                   ...                         ...   \n",
       "mt-Nd3                                 19                          42   \n",
       "mt-Nd4                                233                         591   \n",
       "mt-Nd4l                                 7                          21   \n",
       "mt-Nd5                                 27                          74   \n",
       "mt-Nd6                                  0                           3   \n",
       "\n",
       "               AAGATAGGTAACACCT-1_10X37_2  CAGGGCTGTATCTTCT-1_10X51_1  \\\n",
       "gene                                                                    \n",
       "0610007P14Rik                           1                           2   \n",
       "0610009B22Rik                           1                           0   \n",
       "0610009L18Rik                           0                           2   \n",
       "0610009O20Rik                           1                           0   \n",
       "0610010F05Rik                           7                           0   \n",
       "...                                   ...                         ...   \n",
       "mt-Nd3                                 26                           9   \n",
       "mt-Nd4                                340                          82   \n",
       "mt-Nd4l                                16                           4   \n",
       "mt-Nd5                                 26                          10   \n",
       "mt-Nd6                                  0                           1   \n",
       "\n",
       "               CTGCTCATCCGATCGG-1_10X38_1  TGCTCCAAGAGGCGTT-1_10X51_1  \\\n",
       "gene                                                                    \n",
       "0610007P14Rik                           2                           2   \n",
       "0610009B22Rik                           0                           3   \n",
       "0610009L18Rik                           0                           0   \n",
       "0610009O20Rik                           0                           0   \n",
       "0610010F05Rik                           0                           1   \n",
       "...                                   ...                         ...   \n",
       "mt-Nd3                                 14                           9   \n",
       "mt-Nd4                                 43                          89   \n",
       "mt-Nd4l                                 3                           2   \n",
       "mt-Nd5                                  5                          15   \n",
       "mt-Nd6                                  0                           1   \n",
       "\n",
       "               TTGCCTGTCGTAGGGA-1_10X35_2  GCAACATAGATCGACG-1_10X51_2  ...  \\\n",
       "gene                                                                   ...   \n",
       "0610007P14Rik                           2                           0  ...   \n",
       "0610009B22Rik                           0                           1  ...   \n",
       "0610009L18Rik                           0                           0  ...   \n",
       "0610009O20Rik                           0                           0  ...   \n",
       "0610010F05Rik                           0                           1  ...   \n",
       "...                                   ...                         ...  ...   \n",
       "mt-Nd3                                 12                          26  ...   \n",
       "mt-Nd4                                119                          39  ...   \n",
       "mt-Nd4l                                 9                           8  ...   \n",
       "mt-Nd5                                  9                           7  ...   \n",
       "mt-Nd6                                  1                           1  ...   \n",
       "\n",
       "               CAAGAGGGTCAATCTG-1_10X36_2  TATACCTAGAGTAACT-1_10X35_1  \\\n",
       "gene                                                                    \n",
       "0610007P14Rik                           2                           1   \n",
       "0610009B22Rik                           0                           1   \n",
       "0610009L18Rik                           0                           0   \n",
       "0610009O20Rik                           0                           0   \n",
       "0610010F05Rik                           0                           0   \n",
       "...                                   ...                         ...   \n",
       "mt-Nd3                                 16                          12   \n",
       "mt-Nd4                                230                          71   \n",
       "mt-Nd4l                                 8                           6   \n",
       "mt-Nd5                                 20                           3   \n",
       "mt-Nd6                                  0                           0   \n",
       "\n",
       "               GCACGTGGTTATTCCT-1_10X37_1  GATGATCTCAAGCCTA-1_10X36_1  \\\n",
       "gene                                                                    \n",
       "0610007P14Rik                           2                           0   \n",
       "0610009B22Rik                           0                           0   \n",
       "0610009L18Rik                           0                           0   \n",
       "0610009O20Rik                           1                           0   \n",
       "0610010F05Rik                           1                           0   \n",
       "...                                   ...                         ...   \n",
       "mt-Nd3                                 33                          26   \n",
       "mt-Nd4                                147                         148   \n",
       "mt-Nd4l                                 9                           4   \n",
       "mt-Nd5                                 14                           9   \n",
       "mt-Nd6                                  0                           0   \n",
       "\n",
       "               GGTGTCGAGAAGCTGC-1_10X36_2  GAGAGGTCACATCATG-1_10X37_2  \\\n",
       "gene                                                                    \n",
       "0610007P14Rik                           3                           3   \n",
       "0610009B22Rik                           0                           0   \n",
       "0610009L18Rik                           0                           0   \n",
       "0610009O20Rik                           0                           0   \n",
       "0610010F05Rik                           0                           0   \n",
       "...                                   ...                         ...   \n",
       "mt-Nd3                                 16                          32   \n",
       "mt-Nd4                                 75                         106   \n",
       "mt-Nd4l                                 6                          14   \n",
       "mt-Nd5                                  6                          13   \n",
       "mt-Nd6                                  1                           0   \n",
       "\n",
       "               ATCAGGTTCCACTAGA-1_10X35_2  TCCACGTAGAGCATCG-1_10X37_1  \\\n",
       "gene                                                                    \n",
       "0610007P14Rik                           1                           0   \n",
       "0610009B22Rik                           0                           0   \n",
       "0610009L18Rik                           0                           0   \n",
       "0610009O20Rik                           1                           0   \n",
       "0610010F05Rik                           0                           0   \n",
       "...                                   ...                         ...   \n",
       "mt-Nd3                                 16                          24   \n",
       "mt-Nd4                                 59                          95   \n",
       "mt-Nd4l                                 5                           5   \n",
       "mt-Nd5                                  7                          11   \n",
       "mt-Nd6                                  2                           0   \n",
       "\n",
       "               CTGCGAGCACTACGGC-1_10X35_2  CTCTGGTTCGAAGCAG-1_10X37_1  \n",
       "gene                                                                   \n",
       "0610007P14Rik                           3                           2  \n",
       "0610009B22Rik                           1                           2  \n",
       "0610009L18Rik                           1                           1  \n",
       "0610009O20Rik                           0                           0  \n",
       "0610010F05Rik                           0                           1  \n",
       "...                                   ...                         ...  \n",
       "mt-Nd3                                 22                          39  \n",
       "mt-Nd4                                 87                         126  \n",
       "mt-Nd4l                                 4                           6  \n",
       "mt-Nd5                                  7                           8  \n",
       "mt-Nd6                                  0                           0  \n",
       "\n",
       "[27933 rows x 26824 columns]"
      ]
     },
     "execution_count": 100,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = df_orig.reindex(columns=combined_metadata.columns)\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 110,
   "id": "7fc64a5f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_name</th>\n",
       "      <th>organism</th>\n",
       "      <th>donor_sex</th>\n",
       "      <th>cell_class</th>\n",
       "      <th>cell_subclass</th>\n",
       "      <th>cell_cluster</th>\n",
       "      <th>n_genes_by_counts</th>\n",
       "      <th>log1p_n_genes_by_counts</th>\n",
       "      <th>total_counts</th>\n",
       "      <th>log1p_total_counts</th>\n",
       "      <th>pct_counts_in_top_50_genes</th>\n",
       "      <th>pct_counts_in_top_100_genes</th>\n",
       "      <th>pct_counts_in_top_200_genes</th>\n",
       "      <th>pct_counts_in_top_500_genes</th>\n",
       "      <th>total_counts_mt</th>\n",
       "      <th>log1p_total_counts_mt</th>\n",
       "      <th>pct_counts_mt</th>\n",
       "      <th>n_counts</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>F2S4_151217_005_B01</th>\n",
       "      <td>F2S4_151217_005_B01</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>M</td>\n",
       "      <td>GABAergic</td>\n",
       "      <td>Pvalb</td>\n",
       "      <td>Pvalb Tpbg</td>\n",
       "      <td>8542</td>\n",
       "      <td>9.052868</td>\n",
       "      <td>1282648.0</td>\n",
       "      <td>14.064438</td>\n",
       "      <td>16.040254</td>\n",
       "      <td>22.379172</td>\n",
       "      <td>30.921110</td>\n",
       "      <td>45.899343</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1282648.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>F2S4_151217_005_C01</th>\n",
       "      <td>F2S4_151217_005_C01</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>M</td>\n",
       "      <td>Glutamatergic</td>\n",
       "      <td>L4</td>\n",
       "      <td>L4 IT VISp Rspo1</td>\n",
       "      <td>8111</td>\n",
       "      <td>9.001100</td>\n",
       "      <td>1129496.0</td>\n",
       "      <td>13.937283</td>\n",
       "      <td>14.022449</td>\n",
       "      <td>18.904626</td>\n",
       "      <td>25.959809</td>\n",
       "      <td>39.415102</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1129496.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>F2S4_151217_005_E01</th>\n",
       "      <td>F2S4_151217_005_E01</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>M</td>\n",
       "      <td>Glutamatergic</td>\n",
       "      <td>L4</td>\n",
       "      <td>L4 IT VISp Rspo1</td>\n",
       "      <td>8780</td>\n",
       "      <td>9.080346</td>\n",
       "      <td>1557742.0</td>\n",
       "      <td>14.258749</td>\n",
       "      <td>13.580490</td>\n",
       "      <td>19.555485</td>\n",
       "      <td>27.911361</td>\n",
       "      <td>42.299880</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1557742.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>F2S4_151217_005_F01</th>\n",
       "      <td>F2S4_151217_005_F01</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>M</td>\n",
       "      <td>Glutamatergic</td>\n",
       "      <td>L4</td>\n",
       "      <td>L4 IT VISp Rspo1</td>\n",
       "      <td>8498</td>\n",
       "      <td>9.047704</td>\n",
       "      <td>1306856.0</td>\n",
       "      <td>14.083136</td>\n",
       "      <td>13.945989</td>\n",
       "      <td>19.042113</td>\n",
       "      <td>26.350723</td>\n",
       "      <td>40.343236</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1306856.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>F2S4_151217_005_G01</th>\n",
       "      <td>F2S4_151217_005_G01</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>M</td>\n",
       "      <td>Glutamatergic</td>\n",
       "      <td>L4</td>\n",
       "      <td>L4 IT VISp Rspo1</td>\n",
       "      <td>7566</td>\n",
       "      <td>8.931552</td>\n",
       "      <td>1119824.0</td>\n",
       "      <td>13.928683</td>\n",
       "      <td>14.102752</td>\n",
       "      <td>19.419570</td>\n",
       "      <td>27.235976</td>\n",
       "      <td>41.983919</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1119824.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>F1S4_180124_317_D01</th>\n",
       "      <td>F1S4_180124_317_D01</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>F</td>\n",
       "      <td>GABAergic</td>\n",
       "      <td>Lamp5</td>\n",
       "      <td>Lamp5 Lsp1</td>\n",
       "      <td>8601</td>\n",
       "      <td>9.059750</td>\n",
       "      <td>1022536.0</td>\n",
       "      <td>13.837797</td>\n",
       "      <td>18.500473</td>\n",
       "      <td>24.796584</td>\n",
       "      <td>32.826326</td>\n",
       "      <td>47.307968</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1022536.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>F1S4_180124_317_E01</th>\n",
       "      <td>F1S4_180124_317_E01</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>F</td>\n",
       "      <td>GABAergic</td>\n",
       "      <td>Sncg</td>\n",
       "      <td>Sncg Slc17a8</td>\n",
       "      <td>9823</td>\n",
       "      <td>9.192584</td>\n",
       "      <td>724679.0</td>\n",
       "      <td>13.493485</td>\n",
       "      <td>18.947148</td>\n",
       "      <td>25.948316</td>\n",
       "      <td>34.913665</td>\n",
       "      <td>50.023942</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>724679.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>F1S4_180124_317_F01</th>\n",
       "      <td>F1S4_180124_317_F01</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>F</td>\n",
       "      <td>GABAergic</td>\n",
       "      <td>Sst</td>\n",
       "      <td>Sst Hpse Sema3c</td>\n",
       "      <td>8730</td>\n",
       "      <td>9.074635</td>\n",
       "      <td>980851.0</td>\n",
       "      <td>13.796177</td>\n",
       "      <td>18.740053</td>\n",
       "      <td>25.395294</td>\n",
       "      <td>33.757217</td>\n",
       "      <td>48.250346</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>980851.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>F1S4_180124_317_G01</th>\n",
       "      <td>F1S4_180124_317_G01</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>F</td>\n",
       "      <td>GABAergic</td>\n",
       "      <td>Sncg</td>\n",
       "      <td>Sncg Slc17a8</td>\n",
       "      <td>10633</td>\n",
       "      <td>9.271812</td>\n",
       "      <td>1029766.0</td>\n",
       "      <td>13.844843</td>\n",
       "      <td>17.725775</td>\n",
       "      <td>24.532078</td>\n",
       "      <td>33.706881</td>\n",
       "      <td>48.938594</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1029766.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>F1S4_180124_317_H01</th>\n",
       "      <td>F1S4_180124_317_H01</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>F</td>\n",
       "      <td>GABAergic</td>\n",
       "      <td>Sncg</td>\n",
       "      <td>Sncg Slc17a8</td>\n",
       "      <td>9773</td>\n",
       "      <td>9.187481</td>\n",
       "      <td>965509.0</td>\n",
       "      <td>13.780412</td>\n",
       "      <td>16.794975</td>\n",
       "      <td>23.122415</td>\n",
       "      <td>31.813479</td>\n",
       "      <td>46.717120</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>965509.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>21697 rows × 18 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                             sample_name      organism donor_sex  \\\n",
       "F2S4_151217_005_B01  F2S4_151217_005_B01  Mus musculus         M   \n",
       "F2S4_151217_005_C01  F2S4_151217_005_C01  Mus musculus         M   \n",
       "F2S4_151217_005_E01  F2S4_151217_005_E01  Mus musculus         M   \n",
       "F2S4_151217_005_F01  F2S4_151217_005_F01  Mus musculus         M   \n",
       "F2S4_151217_005_G01  F2S4_151217_005_G01  Mus musculus         M   \n",
       "...                                  ...           ...       ...   \n",
       "F1S4_180124_317_D01  F1S4_180124_317_D01  Mus musculus         F   \n",
       "F1S4_180124_317_E01  F1S4_180124_317_E01  Mus musculus         F   \n",
       "F1S4_180124_317_F01  F1S4_180124_317_F01  Mus musculus         F   \n",
       "F1S4_180124_317_G01  F1S4_180124_317_G01  Mus musculus         F   \n",
       "F1S4_180124_317_H01  F1S4_180124_317_H01  Mus musculus         F   \n",
       "\n",
       "                        cell_class cell_subclass      cell_cluster  \\\n",
       "F2S4_151217_005_B01      GABAergic         Pvalb        Pvalb Tpbg   \n",
       "F2S4_151217_005_C01  Glutamatergic            L4  L4 IT VISp Rspo1   \n",
       "F2S4_151217_005_E01  Glutamatergic            L4  L4 IT VISp Rspo1   \n",
       "F2S4_151217_005_F01  Glutamatergic            L4  L4 IT VISp Rspo1   \n",
       "F2S4_151217_005_G01  Glutamatergic            L4  L4 IT VISp Rspo1   \n",
       "...                            ...           ...               ...   \n",
       "F1S4_180124_317_D01      GABAergic         Lamp5        Lamp5 Lsp1   \n",
       "F1S4_180124_317_E01      GABAergic          Sncg      Sncg Slc17a8   \n",
       "F1S4_180124_317_F01      GABAergic           Sst   Sst Hpse Sema3c   \n",
       "F1S4_180124_317_G01      GABAergic          Sncg      Sncg Slc17a8   \n",
       "F1S4_180124_317_H01      GABAergic          Sncg      Sncg Slc17a8   \n",
       "\n",
       "                     n_genes_by_counts  log1p_n_genes_by_counts  total_counts  \\\n",
       "F2S4_151217_005_B01               8542                 9.052868     1282648.0   \n",
       "F2S4_151217_005_C01               8111                 9.001100     1129496.0   \n",
       "F2S4_151217_005_E01               8780                 9.080346     1557742.0   \n",
       "F2S4_151217_005_F01               8498                 9.047704     1306856.0   \n",
       "F2S4_151217_005_G01               7566                 8.931552     1119824.0   \n",
       "...                                ...                      ...           ...   \n",
       "F1S4_180124_317_D01               8601                 9.059750     1022536.0   \n",
       "F1S4_180124_317_E01               9823                 9.192584      724679.0   \n",
       "F1S4_180124_317_F01               8730                 9.074635      980851.0   \n",
       "F1S4_180124_317_G01              10633                 9.271812     1029766.0   \n",
       "F1S4_180124_317_H01               9773                 9.187481      965509.0   \n",
       "\n",
       "                     log1p_total_counts  pct_counts_in_top_50_genes  \\\n",
       "F2S4_151217_005_B01           14.064438                   16.040254   \n",
       "F2S4_151217_005_C01           13.937283                   14.022449   \n",
       "F2S4_151217_005_E01           14.258749                   13.580490   \n",
       "F2S4_151217_005_F01           14.083136                   13.945989   \n",
       "F2S4_151217_005_G01           13.928683                   14.102752   \n",
       "...                                 ...                         ...   \n",
       "F1S4_180124_317_D01           13.837797                   18.500473   \n",
       "F1S4_180124_317_E01           13.493485                   18.947148   \n",
       "F1S4_180124_317_F01           13.796177                   18.740053   \n",
       "F1S4_180124_317_G01           13.844843                   17.725775   \n",
       "F1S4_180124_317_H01           13.780412                   16.794975   \n",
       "\n",
       "                     pct_counts_in_top_100_genes  pct_counts_in_top_200_genes  \\\n",
       "F2S4_151217_005_B01                    22.379172                    30.921110   \n",
       "F2S4_151217_005_C01                    18.904626                    25.959809   \n",
       "F2S4_151217_005_E01                    19.555485                    27.911361   \n",
       "F2S4_151217_005_F01                    19.042113                    26.350723   \n",
       "F2S4_151217_005_G01                    19.419570                    27.235976   \n",
       "...                                          ...                          ...   \n",
       "F1S4_180124_317_D01                    24.796584                    32.826326   \n",
       "F1S4_180124_317_E01                    25.948316                    34.913665   \n",
       "F1S4_180124_317_F01                    25.395294                    33.757217   \n",
       "F1S4_180124_317_G01                    24.532078                    33.706881   \n",
       "F1S4_180124_317_H01                    23.122415                    31.813479   \n",
       "\n",
       "                     pct_counts_in_top_500_genes  total_counts_mt  \\\n",
       "F2S4_151217_005_B01                    45.899343              0.0   \n",
       "F2S4_151217_005_C01                    39.415102              0.0   \n",
       "F2S4_151217_005_E01                    42.299880              0.0   \n",
       "F2S4_151217_005_F01                    40.343236              0.0   \n",
       "F2S4_151217_005_G01                    41.983919              0.0   \n",
       "...                                          ...              ...   \n",
       "F1S4_180124_317_D01                    47.307968              0.0   \n",
       "F1S4_180124_317_E01                    50.023942              0.0   \n",
       "F1S4_180124_317_F01                    48.250346              0.0   \n",
       "F1S4_180124_317_G01                    48.938594              0.0   \n",
       "F1S4_180124_317_H01                    46.717120              0.0   \n",
       "\n",
       "                     log1p_total_counts_mt  pct_counts_mt   n_counts  \n",
       "F2S4_151217_005_B01                    0.0            0.0  1282648.0  \n",
       "F2S4_151217_005_C01                    0.0            0.0  1129496.0  \n",
       "F2S4_151217_005_E01                    0.0            0.0  1557742.0  \n",
       "F2S4_151217_005_F01                    0.0            0.0  1306856.0  \n",
       "F2S4_151217_005_G01                    0.0            0.0  1119824.0  \n",
       "...                                    ...            ...        ...  \n",
       "F1S4_180124_317_D01                    0.0            0.0  1022536.0  \n",
       "F1S4_180124_317_E01                    0.0            0.0   724679.0  \n",
       "F1S4_180124_317_F01                    0.0            0.0   980851.0  \n",
       "F1S4_180124_317_G01                    0.0            0.0  1029766.0  \n",
       "F1S4_180124_317_H01                    0.0            0.0   965509.0  \n",
       "\n",
       "[21697 rows x 18 columns]"
      ]
     },
     "execution_count": 110,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata_sc.obs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 119,
   "id": "5c9568da",
   "metadata": {},
   "outputs": [],
   "source": [
    "cells = df.columns.to_frame()\n",
    "cells = cells.rename(columns = {0:'sample_name'})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 121,
   "id": "23c6505b",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_sc_sd = sc.AnnData(df.T, obs=cells, var=df.index.to_frame())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 122,
   "id": "36ff0ad8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 26824 × 27933\n",
       "    obs: 'sample_name'\n",
       "    var: 'gene'"
      ]
     },
     "execution_count": 122,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata_sc_sd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 132,
   "id": "682bbd7a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "CAACAACAGACATATG-1_10X35_2           GABA\n",
       "ATGACCATCGTGAGAG-1_10X51_1           GABA\n",
       "AACCCAAAGAAATGGG-1_10X35_2           GABA\n",
       "GAGGGATGTATGCTAC-1_10X35_1           GABA\n",
       "AAGATAGGTAACACCT-1_10X37_2           GABA\n",
       "                                 ...     \n",
       "GAGAGGTCACATCATG-1_10X37_2    Nonneuronal\n",
       "ATCAGGTTCCACTAGA-1_10X35_2    Nonneuronal\n",
       "TCCACGTAGAGCATCG-1_10X37_1    Nonneuronal\n",
       "CTGCGAGCACTACGGC-1_10X35_2    Nonneuronal\n",
       "CTCTGGTTCGAAGCAG-1_10X37_1    Nonneuronal\n",
       "Name: cell_class, Length: 26824, dtype: object"
      ]
     },
     "execution_count": 132,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "combined_metadata.loc['cell_class']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 201,
   "id": "109bb479",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "CAACAACAGACATATG-1_10X35_2    Chat-Isl1\n",
       "ATGACCATCGTGAGAG-1_10X51_1    Chat-Isl1\n",
       "AACCCAAAGAAATGGG-1_10X35_2    Chat-Isl1\n",
       "GAGGGATGTATGCTAC-1_10X35_1    Chat-Isl1\n",
       "AAGATAGGTAACACCT-1_10X37_2    Chat-Isl1\n",
       "                                ...    \n",
       "GAGAGGTCACATCATG-1_10X37_2         OL_2\n",
       "ATCAGGTTCCACTAGA-1_10X35_2         OL_2\n",
       "TCCACGTAGAGCATCG-1_10X37_1         OL_2\n",
       "CTGCGAGCACTACGGC-1_10X35_2         OL_2\n",
       "CTCTGGTTCGAAGCAG-1_10X37_1         OL_2\n",
       "Name: markers, Length: 26824, dtype: object"
      ]
     },
     "execution_count": 201,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "combined_metadata.loc['markers']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 285,
   "id": "f8309d43",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "CAACAACAGACATATG-1_10X35_2    chat-isl1\n",
       "ATGACCATCGTGAGAG-1_10X51_1    chat-isl1\n",
       "AACCCAAAGAAATGGG-1_10X35_2    chat-isl1\n",
       "GAGGGATGTATGCTAC-1_10X35_1    chat-isl1\n",
       "AAGATAGGTAACACCT-1_10X37_2    chat-isl1\n",
       "                                ...    \n",
       "GAGAGGTCACATCATG-1_10X37_2         ol_2\n",
       "ATCAGGTTCCACTAGA-1_10X35_2         ol_2\n",
       "TCCACGTAGAGCATCG-1_10X37_1         ol_2\n",
       "CTGCGAGCACTACGGC-1_10X35_2         ol_2\n",
       "CTCTGGTTCGAAGCAG-1_10X37_1         ol_2\n",
       "Name: markers, Length: 26824, dtype: object"
      ]
     },
     "execution_count": 285,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#make lowercase\n",
    "combined_metadata.loc['markers'].str.lower()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 133,
   "id": "7451f09e",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_sc_sd.obs['cell_class'] = combined_metadata.loc['cell_class']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 286,
   "id": "7a2ef8dc",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_sc_sd.obs['cell_subclass'] = combined_metadata.loc['markers'].str.lower()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 287,
   "id": "fad04479",
   "metadata": {},
   "outputs": [],
   "source": [
    "output_path = '/bigdata/isaac/tangram_integration/'\n",
    "adata_sc_sd.write(output_path +'adata_sc_sd.h5ad', compression='gzip')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 127,
   "id": "c7ea1c71",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>mt</th>\n",
       "      <th>n_cells_by_counts</th>\n",
       "      <th>mean_counts</th>\n",
       "      <th>log1p_mean_counts</th>\n",
       "      <th>pct_dropout_by_counts</th>\n",
       "      <th>total_counts</th>\n",
       "      <th>log1p_total_counts</th>\n",
       "      <th>n_cells</th>\n",
       "      <th>highly_variable</th>\n",
       "      <th>highly_variable_rank</th>\n",
       "      <th>means</th>\n",
       "      <th>variances</th>\n",
       "      <th>variances_norm</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0610005C13Rik</th>\n",
       "      <td>False</td>\n",
       "      <td>872</td>\n",
       "      <td>0.524247</td>\n",
       "      <td>0.421501</td>\n",
       "      <td>96.237812</td>\n",
       "      <td>12151.0</td>\n",
       "      <td>9.405249</td>\n",
       "      <td>867</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.534452</td>\n",
       "      <td>29.752143</td>\n",
       "      <td>1.090297</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0610006L08Rik</th>\n",
       "      <td>False</td>\n",
       "      <td>29</td>\n",
       "      <td>0.049918</td>\n",
       "      <td>0.048712</td>\n",
       "      <td>99.874881</td>\n",
       "      <td>1157.0</td>\n",
       "      <td>7.054450</td>\n",
       "      <td>29</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.050422</td>\n",
       "      <td>5.631491</td>\n",
       "      <td>2.697598</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0610007P14Rik</th>\n",
       "      <td>False</td>\n",
       "      <td>21519</td>\n",
       "      <td>161.839111</td>\n",
       "      <td>5.092762</td>\n",
       "      <td>7.157649</td>\n",
       "      <td>3751107.0</td>\n",
       "      <td>15.137562</td>\n",
       "      <td>21344</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>164.495598</td>\n",
       "      <td>14562.080191</td>\n",
       "      <td>0.629189</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0610009B22Rik</th>\n",
       "      <td>False</td>\n",
       "      <td>21203</td>\n",
       "      <td>150.453445</td>\n",
       "      <td>5.020278</td>\n",
       "      <td>8.521011</td>\n",
       "      <td>3487210.0</td>\n",
       "      <td>15.064612</td>\n",
       "      <td>21044</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>153.827165</td>\n",
       "      <td>13897.739024</td>\n",
       "      <td>0.652549</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0610009E02Rik</th>\n",
       "      <td>False</td>\n",
       "      <td>4185</td>\n",
       "      <td>1.559280</td>\n",
       "      <td>0.939726</td>\n",
       "      <td>81.944085</td>\n",
       "      <td>36141.0</td>\n",
       "      <td>10.495211</td>\n",
       "      <td>4166</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.591464</td>\n",
       "      <td>51.485746</td>\n",
       "      <td>0.484592</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Zzef1</th>\n",
       "      <td>False</td>\n",
       "      <td>16165</td>\n",
       "      <td>38.780094</td>\n",
       "      <td>3.683367</td>\n",
       "      <td>30.257140</td>\n",
       "      <td>898845.0</td>\n",
       "      <td>13.708867</td>\n",
       "      <td>16034</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>38.959856</td>\n",
       "      <td>4371.811764</td>\n",
       "      <td>0.873219</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Zzz3</th>\n",
       "      <td>False</td>\n",
       "      <td>16354</td>\n",
       "      <td>50.861378</td>\n",
       "      <td>3.948574</td>\n",
       "      <td>29.441712</td>\n",
       "      <td>1178865.0</td>\n",
       "      <td>13.980063</td>\n",
       "      <td>16227</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>51.304835</td>\n",
       "      <td>6362.498978</td>\n",
       "      <td>0.978110</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>a</th>\n",
       "      <td>False</td>\n",
       "      <td>188</td>\n",
       "      <td>0.151092</td>\n",
       "      <td>0.140711</td>\n",
       "      <td>99.188886</td>\n",
       "      <td>3502.0</td>\n",
       "      <td>8.161375</td>\n",
       "      <td>185</td>\n",
       "      <td>True</td>\n",
       "      <td>3379.0</td>\n",
       "      <td>0.152878</td>\n",
       "      <td>16.621585</td>\n",
       "      <td>3.142282</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>l7Rn6</th>\n",
       "      <td>False</td>\n",
       "      <td>21904</td>\n",
       "      <td>208.119507</td>\n",
       "      <td>5.342906</td>\n",
       "      <td>5.496592</td>\n",
       "      <td>4823794.0</td>\n",
       "      <td>15.389071</td>\n",
       "      <td>21728</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>213.394617</td>\n",
       "      <td>18848.619805</td>\n",
       "      <td>0.581846</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>n-R5s136</th>\n",
       "      <td>False</td>\n",
       "      <td>39</td>\n",
       "      <td>0.002718</td>\n",
       "      <td>0.002714</td>\n",
       "      <td>99.831737</td>\n",
       "      <td>63.0</td>\n",
       "      <td>4.158883</td>\n",
       "      <td>39</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.002765</td>\n",
       "      <td>0.007183</td>\n",
       "      <td>0.369636</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>36826 rows × 13 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                  mt  n_cells_by_counts  mean_counts  log1p_mean_counts  \\\n",
       "0610005C13Rik  False                872     0.524247           0.421501   \n",
       "0610006L08Rik  False                 29     0.049918           0.048712   \n",
       "0610007P14Rik  False              21519   161.839111           5.092762   \n",
       "0610009B22Rik  False              21203   150.453445           5.020278   \n",
       "0610009E02Rik  False               4185     1.559280           0.939726   \n",
       "...              ...                ...          ...                ...   \n",
       "Zzef1          False              16165    38.780094           3.683367   \n",
       "Zzz3           False              16354    50.861378           3.948574   \n",
       "a              False                188     0.151092           0.140711   \n",
       "l7Rn6          False              21904   208.119507           5.342906   \n",
       "n-R5s136       False                 39     0.002718           0.002714   \n",
       "\n",
       "               pct_dropout_by_counts  total_counts  log1p_total_counts  \\\n",
       "0610005C13Rik              96.237812       12151.0            9.405249   \n",
       "0610006L08Rik              99.874881        1157.0            7.054450   \n",
       "0610007P14Rik               7.157649     3751107.0           15.137562   \n",
       "0610009B22Rik               8.521011     3487210.0           15.064612   \n",
       "0610009E02Rik              81.944085       36141.0           10.495211   \n",
       "...                              ...           ...                 ...   \n",
       "Zzef1                      30.257140      898845.0           13.708867   \n",
       "Zzz3                       29.441712     1178865.0           13.980063   \n",
       "a                          99.188886        3502.0            8.161375   \n",
       "l7Rn6                       5.496592     4823794.0           15.389071   \n",
       "n-R5s136                   99.831737          63.0            4.158883   \n",
       "\n",
       "               n_cells  highly_variable  highly_variable_rank       means  \\\n",
       "0610005C13Rik      867            False                   NaN    0.534452   \n",
       "0610006L08Rik       29            False                   NaN    0.050422   \n",
       "0610007P14Rik    21344            False                   NaN  164.495598   \n",
       "0610009B22Rik    21044            False                   NaN  153.827165   \n",
       "0610009E02Rik     4166            False                   NaN    1.591464   \n",
       "...                ...              ...                   ...         ...   \n",
       "Zzef1            16034            False                   NaN   38.959856   \n",
       "Zzz3             16227            False                   NaN   51.304835   \n",
       "a                  185             True                3379.0    0.152878   \n",
       "l7Rn6            21728            False                   NaN  213.394617   \n",
       "n-R5s136            39            False                   NaN    0.002765   \n",
       "\n",
       "                  variances  variances_norm  \n",
       "0610005C13Rik     29.752143        1.090297  \n",
       "0610006L08Rik      5.631491        2.697598  \n",
       "0610007P14Rik  14562.080191        0.629189  \n",
       "0610009B22Rik  13897.739024        0.652549  \n",
       "0610009E02Rik     51.485746        0.484592  \n",
       "...                     ...             ...  \n",
       "Zzef1           4371.811764        0.873219  \n",
       "Zzz3            6362.498978        0.978110  \n",
       "a                 16.621585        3.142282  \n",
       "l7Rn6          18848.619805        0.581846  \n",
       "n-R5s136           0.007183        0.369636  \n",
       "\n",
       "[36826 rows x 13 columns]"
      ]
     },
     "execution_count": 127,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata_sc.var"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 137,
   "id": "ceafce1e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "F2S4_151217_005_B01        GABAergic\n",
       "F2S4_151217_005_C01    Glutamatergic\n",
       "F2S4_151217_005_E01    Glutamatergic\n",
       "F2S4_151217_005_F01    Glutamatergic\n",
       "F2S4_151217_005_G01    Glutamatergic\n",
       "                           ...      \n",
       "F1S4_180124_317_D01        GABAergic\n",
       "F1S4_180124_317_E01        GABAergic\n",
       "F1S4_180124_317_F01        GABAergic\n",
       "F1S4_180124_317_G01        GABAergic\n",
       "F1S4_180124_317_H01        GABAergic\n",
       "Name: cell_class, Length: 21697, dtype: category\n",
       "Categories (4, object): ['Endothelial', 'GABAergic', 'Glutamatergic', 'Non-Neuronal']"
      ]
     },
     "execution_count": 137,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata_sc.obs['cell_class']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 139,
   "id": "69b760de",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 26824 × 27933\n",
       "    obs: 'sample_name', 'cell_class', 'cell_subclass'\n",
       "    var: 'gene'"
      ]
     },
     "execution_count": 139,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata_sc_sd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 140,
   "id": "2a8499ca",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1401"
      ]
     },
     "execution_count": 140,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#from tutorial\n",
    "sc.tl.rank_genes_groups(adata_sc, groupby=\"cell_subclass\", use_raw=False)\n",
    "markers_df = pd.DataFrame(adata_sc.uns[\"rank_genes_groups\"][\"names\"]).iloc[0:100, :]\n",
    "markers = list(np.unique(markers_df.melt().value.values))\n",
    "len(markers)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 141,
   "id": "4959f8ef",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['1110008P14Rik',\n",
       " '1700001L19Rik',\n",
       " '1700019D03Rik',\n",
       " '1700019L22Rik',\n",
       " '1700047M11Rik',\n",
       " '2010111I01Rik',\n",
       " '2010300C02Rik',\n",
       " '2610001J05Rik',\n",
       " '2610028E06Rik',\n",
       " '2810468N07Rik',\n",
       " '2900011O08Rik',\n",
       " '2900055J20Rik',\n",
       " '3110035E14Rik',\n",
       " '4632428N05Rik',\n",
       " '6330403A02Rik',\n",
       " '6330403K07Rik',\n",
       " '9130024F11Rik',\n",
       " '9430020K01Rik',\n",
       " 'A330050F15Rik',\n",
       " 'A530058N18Rik',\n",
       " 'A830018L16Rik',\n",
       " 'AF251705',\n",
       " 'ATP6',\n",
       " 'Abat',\n",
       " 'Abcb1a',\n",
       " 'Abcc9',\n",
       " 'Abcg2',\n",
       " 'Abracl',\n",
       " 'Ace2',\n",
       " 'Ache',\n",
       " 'Ackr3',\n",
       " 'Acot7',\n",
       " 'Acsbg1',\n",
       " 'Acsl3',\n",
       " 'Acta2',\n",
       " 'Actb',\n",
       " 'Actr3b',\n",
       " 'Acvrl1',\n",
       " 'Adamts19',\n",
       " 'Adap2',\n",
       " 'Adarb2',\n",
       " 'Adcy1',\n",
       " 'Adcy2',\n",
       " 'Adcyap1',\n",
       " 'Adora1',\n",
       " 'Adra1a',\n",
       " 'Adra1b',\n",
       " 'Afap1',\n",
       " 'Ahi1',\n",
       " 'Ahnak',\n",
       " 'Aif1',\n",
       " 'Ak4',\n",
       " 'Ak5',\n",
       " 'Alcam',\n",
       " 'Aldh1a2',\n",
       " 'Aldoc',\n",
       " 'Alox5ap',\n",
       " 'Amigo2',\n",
       " 'Anapc11',\n",
       " 'Ank1',\n",
       " 'Ankrd29',\n",
       " 'Ankrd33b',\n",
       " 'Ankrd35',\n",
       " 'Ano3',\n",
       " 'Anxa2',\n",
       " 'Anxa3',\n",
       " 'Anxa5',\n",
       " 'Ap1s2',\n",
       " 'Apod',\n",
       " 'Apoe',\n",
       " 'Apold1',\n",
       " 'Appl2',\n",
       " 'Aqp4',\n",
       " 'Araf',\n",
       " 'Arhgap25',\n",
       " 'Arhgap29',\n",
       " 'Arhgap31',\n",
       " 'Arhgap32',\n",
       " 'Arhgdib',\n",
       " 'Arhgdig',\n",
       " 'Arhgef10',\n",
       " 'Arhgef25',\n",
       " 'Arl15',\n",
       " 'Arl4c',\n",
       " 'Arpc1b',\n",
       " 'Arpp19',\n",
       " 'Arpp21',\n",
       " 'Art3',\n",
       " 'Arx',\n",
       " 'Asic1',\n",
       " 'Asic4',\n",
       " 'Asns',\n",
       " 'Aspa',\n",
       " 'Aspn',\n",
       " 'Atf4',\n",
       " 'Atox1',\n",
       " 'Atp13a5',\n",
       " 'Atp1a1',\n",
       " 'Atp1a2',\n",
       " 'Atp1b1',\n",
       " 'Atp1b2',\n",
       " 'Atp1b3',\n",
       " 'Atp2a3',\n",
       " 'Atp2b1',\n",
       " 'Atp2b2',\n",
       " 'Atp2b4',\n",
       " 'Atp5b',\n",
       " 'Atp5c1',\n",
       " 'Atp5g3',\n",
       " 'Atp5j',\n",
       " 'Atp6ap1l',\n",
       " 'Atp6ap2',\n",
       " 'Axl',\n",
       " 'B230216N24Rik',\n",
       " 'B2m',\n",
       " 'B3galt2',\n",
       " 'BC028528',\n",
       " 'BC030499',\n",
       " 'BC030500',\n",
       " 'Bai2',\n",
       " 'Baiap2',\n",
       " 'Bcam',\n",
       " 'Bcan',\n",
       " 'Bcas1',\n",
       " 'Bcat1',\n",
       " 'Bcl11a',\n",
       " 'Bcl11b',\n",
       " 'Bcl6',\n",
       " 'Bend5',\n",
       " 'Bex1',\n",
       " 'Bex4',\n",
       " 'Bgn',\n",
       " 'Bhlhe40',\n",
       " 'Bmp3',\n",
       " 'Bri3',\n",
       " 'Brinp1',\n",
       " 'Brinp3',\n",
       " 'Bsg',\n",
       " 'Btbd11',\n",
       " 'Btbd17',\n",
       " 'Btbd3',\n",
       " 'Btg1',\n",
       " 'Btg2',\n",
       " 'C1qa',\n",
       " 'C1qb',\n",
       " 'C1qc',\n",
       " 'C1ql3',\n",
       " 'C730002L08Rik',\n",
       " 'COX1',\n",
       " 'COX2',\n",
       " 'COX3',\n",
       " 'CYTB',\n",
       " 'Cacna1c',\n",
       " 'Cacna1e',\n",
       " 'Cacna2d1',\n",
       " 'Cacna2d2',\n",
       " 'Cacna2d3',\n",
       " 'Cacng3',\n",
       " 'Cacng5',\n",
       " 'Calb1',\n",
       " 'Calb2',\n",
       " 'Cald1',\n",
       " 'Caln1',\n",
       " 'Caly',\n",
       " 'Camk1',\n",
       " 'Camk2a',\n",
       " 'Camk2d',\n",
       " 'Camk2n1',\n",
       " 'Camk2n2',\n",
       " 'Camk4',\n",
       " 'Camkk2',\n",
       " 'Cap2',\n",
       " 'Car10',\n",
       " 'Car14',\n",
       " 'Car2',\n",
       " 'Car4',\n",
       " 'Carhsp1',\n",
       " 'Cartpt',\n",
       " 'Cav1',\n",
       " 'Cbln4',\n",
       " 'Ccdc109b',\n",
       " 'Ccdc136',\n",
       " 'Ccdc141',\n",
       " 'Cck',\n",
       " 'Cckbr',\n",
       " 'Ccl27a',\n",
       " 'Ccnh',\n",
       " 'Cd164',\n",
       " 'Cd274',\n",
       " 'Cd53',\n",
       " 'Cd63',\n",
       " 'Cd63-ps',\n",
       " 'Cd68',\n",
       " 'Cd81',\n",
       " 'Cd82',\n",
       " 'Cd9',\n",
       " 'Cd9-ps',\n",
       " 'Cdc42ep1',\n",
       " 'Cdh12',\n",
       " 'Cdh18',\n",
       " 'Cdh5',\n",
       " 'Cdh9',\n",
       " 'Cdk14',\n",
       " 'Cds1',\n",
       " 'Cdyl2',\n",
       " 'Celf4',\n",
       " 'Celf6',\n",
       " 'Cend1',\n",
       " 'Cep112',\n",
       " 'Cercam',\n",
       " 'Cers2',\n",
       " 'Cers6',\n",
       " 'Cetn3',\n",
       " 'Cfh',\n",
       " 'Cgnl1',\n",
       " 'Cgref1',\n",
       " 'Chchd10',\n",
       " 'Chd3os',\n",
       " 'Chga',\n",
       " 'Chgb',\n",
       " 'Chl1',\n",
       " 'Chn1',\n",
       " 'Chrna4',\n",
       " 'Chst1',\n",
       " 'Chst8',\n",
       " 'Ciapin1',\n",
       " 'Cib2',\n",
       " 'Cit',\n",
       " 'Cited2',\n",
       " 'Ck-ps1',\n",
       " 'Ck-ps3',\n",
       " 'Ckb',\n",
       " 'Ckmt1',\n",
       " 'Cldn10',\n",
       " 'Cldn11',\n",
       " 'Cldn5',\n",
       " 'Clic1',\n",
       " 'Clic4',\n",
       " 'Clk1',\n",
       " 'Clstn2',\n",
       " 'Clu',\n",
       " 'Cmbl',\n",
       " 'Cml1',\n",
       " 'Cmtm5',\n",
       " 'Cnih2',\n",
       " 'Cnih3',\n",
       " 'Cnksr2',\n",
       " 'Cnn2',\n",
       " 'Cnn3',\n",
       " 'Cnp',\n",
       " 'Cnr1',\n",
       " 'Cntn4',\n",
       " 'Cntnap2',\n",
       " 'Cntnap4',\n",
       " 'Cntnap5c',\n",
       " 'Cobl',\n",
       " 'Coch',\n",
       " 'Col1a1',\n",
       " 'Col1a2',\n",
       " 'Col5a1',\n",
       " 'Col6a1',\n",
       " 'Colec12',\n",
       " 'Coro1b',\n",
       " 'Cox4i2',\n",
       " 'Cox5a',\n",
       " 'Cox6a2',\n",
       " 'Cp',\n",
       " 'Cped1',\n",
       " 'Cplx1',\n",
       " 'Cplx2',\n",
       " 'Cplx3',\n",
       " 'Cpm',\n",
       " 'Cpne2',\n",
       " 'Cpne4',\n",
       " 'Cpne5',\n",
       " 'Cpne6',\n",
       " 'Cpne7',\n",
       " 'Cpne8',\n",
       " 'Cpne9',\n",
       " 'Cpq',\n",
       " 'Crh',\n",
       " 'Crim1',\n",
       " 'Crip1',\n",
       " 'Crip2',\n",
       " 'Crispld1',\n",
       " 'Crtac1',\n",
       " 'Cryab',\n",
       " 'Crybb1',\n",
       " 'Crym',\n",
       " 'Csdc2',\n",
       " 'Csf1r',\n",
       " 'Csmd3',\n",
       " 'Csrp1',\n",
       " 'Cst3',\n",
       " 'Ctgf',\n",
       " 'Ctla2a',\n",
       " 'Ctsd',\n",
       " 'Ctsh',\n",
       " 'Ctss',\n",
       " 'Ctsz',\n",
       " 'Ctxn1',\n",
       " 'Cux2',\n",
       " 'Cx3cr1',\n",
       " 'Cxcl12',\n",
       " 'Cxcl14',\n",
       " 'Cxcr4',\n",
       " 'Cxx1a',\n",
       " 'Cyba',\n",
       " 'Cyc1',\n",
       " 'Cycs',\n",
       " 'Cygb',\n",
       " 'Cyth2',\n",
       " 'Cyth4',\n",
       " 'Cyyr1',\n",
       " 'Daam2',\n",
       " 'Dach1',\n",
       " 'Dbi',\n",
       " 'Dbx2',\n",
       " 'Dclk2',\n",
       " 'Dcn',\n",
       " 'Dcx',\n",
       " 'Ddah1',\n",
       " 'Ddr1',\n",
       " 'Degs2',\n",
       " 'Dennd2a',\n",
       " 'Des',\n",
       " 'Dgkb',\n",
       " 'Dgkg',\n",
       " 'Dio2',\n",
       " 'Diras2',\n",
       " 'Disp2',\n",
       " 'Dkk3',\n",
       " 'Dkkl1',\n",
       " 'Dleu7',\n",
       " 'Dlg1',\n",
       " 'Dlx1',\n",
       " 'Dlx1as',\n",
       " 'Dlx2',\n",
       " 'Dlx5',\n",
       " 'Dlx6os1',\n",
       " 'Dnajc21',\n",
       " 'Dner',\n",
       " 'Dnm3',\n",
       " 'Dock10',\n",
       " 'Dock6',\n",
       " 'Dok5',\n",
       " 'Dpp6',\n",
       " 'Dpy19l1',\n",
       " 'Drd1',\n",
       " 'Dstn',\n",
       " 'Dtnb',\n",
       " 'Dtnbp1',\n",
       " 'Dusp14',\n",
       " 'Dync1i1',\n",
       " 'Dynll1',\n",
       " 'E530001K10Rik',\n",
       " 'Ebf3',\n",
       " 'Ece1',\n",
       " 'Eef1a1',\n",
       " 'Eef1e1',\n",
       " 'Efemp1',\n",
       " 'Efr3a',\n",
       " 'Egfl7',\n",
       " 'Egln3',\n",
       " 'Egr1',\n",
       " 'Egr3',\n",
       " 'Eif1',\n",
       " 'Elavl2',\n",
       " 'Elfn1',\n",
       " 'Elmo1',\n",
       " 'Elmod1',\n",
       " 'Elovl1',\n",
       " 'Eltd1',\n",
       " 'Emcn',\n",
       " 'Emp3',\n",
       " 'Emr1',\n",
       " 'Emx2',\n",
       " 'Enc1',\n",
       " 'Eng',\n",
       " 'Enho',\n",
       " 'Eno1',\n",
       " 'Eno1b',\n",
       " 'Enpep',\n",
       " 'Enpp2',\n",
       " 'Entpd1',\n",
       " 'Eogt',\n",
       " 'Epas1',\n",
       " 'Epha4',\n",
       " 'Ephx4',\n",
       " 'Erbb4',\n",
       " 'Ermn',\n",
       " 'Esam',\n",
       " 'Esrrg',\n",
       " 'Etl4',\n",
       " 'Ets1',\n",
       " 'Etv1',\n",
       " 'Etv5',\n",
       " 'Evi2a',\n",
       " 'Extl1',\n",
       " 'F3',\n",
       " 'Fa2h',\n",
       " 'Fam107a',\n",
       " 'Fam131a',\n",
       " 'Fam135b',\n",
       " 'Fam159b',\n",
       " 'Fam163b',\n",
       " 'Fam184a',\n",
       " 'Fam19a1',\n",
       " 'Fam19a2',\n",
       " 'Fam20a',\n",
       " 'Fam210b',\n",
       " 'Fam212b',\n",
       " 'Fam84b',\n",
       " 'Fat3',\n",
       " 'Fbxw7',\n",
       " 'Fcer1g',\n",
       " 'Fcgr3',\n",
       " 'Fcgrt',\n",
       " 'Fcrls',\n",
       " 'Fdps',\n",
       " 'Fermt2',\n",
       " 'Fezf2',\n",
       " 'Fgf1',\n",
       " 'Fgf12',\n",
       " 'Fgf13',\n",
       " 'Fgf9',\n",
       " 'Fgfr3',\n",
       " 'Fhl1',\n",
       " 'Fhl2',\n",
       " 'Filip1l',\n",
       " 'Fkbp1a',\n",
       " 'Flna',\n",
       " 'Flt1',\n",
       " 'Flt3',\n",
       " 'Fmnl1',\n",
       " 'Fn1',\n",
       " 'Fndc5',\n",
       " 'Foxc1',\n",
       " 'Foxp1',\n",
       " 'Foxp2',\n",
       " 'Frmd7',\n",
       " 'Frmd8',\n",
       " 'Frmpd4',\n",
       " 'Fstl4',\n",
       " 'Fstl5',\n",
       " 'Fth1',\n",
       " 'Ftl1',\n",
       " 'Ftl2',\n",
       " 'Fut9',\n",
       " 'Fxyd1',\n",
       " 'Fxyd5',\n",
       " 'Fxyd6',\n",
       " 'Fxyd7',\n",
       " 'Fyb',\n",
       " 'G0s2',\n",
       " 'G630016G05Rik',\n",
       " 'Gaa',\n",
       " 'Gab1',\n",
       " 'Gabra5',\n",
       " 'Gabrb2',\n",
       " 'Gabrd',\n",
       " 'Gad1',\n",
       " 'Gad1-ps',\n",
       " 'Gad2',\n",
       " 'Gadd45a',\n",
       " 'Gal3st1',\n",
       " 'Galnt14',\n",
       " 'Galnt9',\n",
       " 'Galntl6',\n",
       " 'Gap43',\n",
       " 'Garnl3',\n",
       " 'Gas5',\n",
       " 'Gatm',\n",
       " 'Gda',\n",
       " 'Gfra2',\n",
       " 'Gfra4',\n",
       " 'Ggt5',\n",
       " 'Gja1',\n",
       " 'Gjb1',\n",
       " 'Gjb2',\n",
       " 'Gjb6',\n",
       " 'Gjc1',\n",
       " 'Gjc3',\n",
       " 'Glra2',\n",
       " 'Gltp',\n",
       " 'Glul',\n",
       " 'Gm10076',\n",
       " 'Gm10080',\n",
       " 'Gm10116',\n",
       " 'Gm10182',\n",
       " 'Gm10257',\n",
       " 'Gm10269',\n",
       " 'Gm11223',\n",
       " 'Gm11469',\n",
       " 'Gm11549',\n",
       " 'Gm12164',\n",
       " 'Gm12222',\n",
       " 'Gm12791',\n",
       " 'Gm12892',\n",
       " 'Gm13456',\n",
       " 'Gm13629',\n",
       " 'Gm13861',\n",
       " 'Gm13889',\n",
       " 'Gm14204',\n",
       " 'Gm14303',\n",
       " 'Gm15427',\n",
       " 'Gm15612',\n",
       " 'Gm17750',\n",
       " 'Gm1821',\n",
       " 'Gm19410',\n",
       " 'Gm21399',\n",
       " 'Gm2694',\n",
       " 'Gm27199',\n",
       " 'Gm2830',\n",
       " 'Gm3134',\n",
       " 'Gm3764',\n",
       " 'Gm4149',\n",
       " 'Gm4617',\n",
       " 'Gm5087',\n",
       " 'Gm5454',\n",
       " 'Gm5514',\n",
       " 'Gm5526',\n",
       " 'Gm5905',\n",
       " 'Gm6421',\n",
       " 'Gm6485',\n",
       " 'Gm6581',\n",
       " 'Gm6863',\n",
       " 'Gm7204',\n",
       " 'Gm7554',\n",
       " 'Gm7618',\n",
       " 'Gm7665',\n",
       " 'Gm7676',\n",
       " 'Gm7808',\n",
       " 'Gm8814',\n",
       " 'Gm8894',\n",
       " 'Gm8960',\n",
       " 'Gmpr',\n",
       " 'Gnao1',\n",
       " 'Gng11',\n",
       " 'Gng12',\n",
       " 'Gng13',\n",
       " 'Gng2',\n",
       " 'Gng4',\n",
       " 'Gng5',\n",
       " 'Golga7b',\n",
       " 'Got1',\n",
       " 'Gpd1',\n",
       " 'Gper1',\n",
       " 'Gpm6b',\n",
       " 'Gpr116',\n",
       " 'Gpr12',\n",
       " 'Gpr123',\n",
       " 'Gpr176',\n",
       " 'Gpr183',\n",
       " 'Gpr22',\n",
       " 'Gpr34',\n",
       " 'Gpr37',\n",
       " 'Gpr37l1',\n",
       " 'Gpr88',\n",
       " 'Gprasp2',\n",
       " 'Gpsm1',\n",
       " 'Gpx4',\n",
       " 'Gpx8',\n",
       " 'Grb14',\n",
       " 'Gria3',\n",
       " 'Grik1',\n",
       " 'Grik3',\n",
       " 'Grin3a',\n",
       " 'Grip1',\n",
       " 'Grm1',\n",
       " 'Grm2',\n",
       " 'Grm3',\n",
       " 'Grn',\n",
       " 'Grpr',\n",
       " 'Gsn',\n",
       " 'Gsta4',\n",
       " 'Gstm1',\n",
       " 'Gstm2',\n",
       " 'Gstm2-ps1',\n",
       " 'Gstm5',\n",
       " 'Gsto1',\n",
       " 'Gstp1',\n",
       " 'Gucy1b3',\n",
       " 'H3f3a',\n",
       " 'H3f3a-ps2',\n",
       " 'H3f3b',\n",
       " 'Hap1',\n",
       " 'Hapln1',\n",
       " 'Hcn1',\n",
       " 'Hdac9',\n",
       " 'Hepacam',\n",
       " 'Hexb',\n",
       " 'Higd1b',\n",
       " 'Hist1h2bc',\n",
       " 'Hist3h2a',\n",
       " 'Hivep2',\n",
       " 'Hk2',\n",
       " 'Hlf',\n",
       " 'Hmha1',\n",
       " 'Hnrnpa2b1',\n",
       " 'Homer1',\n",
       " 'Hpca',\n",
       " 'Hpcal4',\n",
       " 'Hpgd',\n",
       " 'Hpgds',\n",
       " 'Hs3st2',\n",
       " 'Hs3st4',\n",
       " 'Hsp25-ps1',\n",
       " 'Hspa12a',\n",
       " 'Hspb1',\n",
       " 'Hspb2',\n",
       " 'Htr3a',\n",
       " 'Htra1',\n",
       " 'Icam2',\n",
       " 'Id1',\n",
       " 'Id2',\n",
       " 'Id3',\n",
       " 'Id4',\n",
       " 'Ier5',\n",
       " 'Ifi27',\n",
       " 'Ifit2',\n",
       " 'Ifitm1',\n",
       " 'Ifitm10',\n",
       " 'Ifitm2',\n",
       " 'Ifitm3',\n",
       " 'Ifngr1',\n",
       " 'Ift57',\n",
       " 'Igf1',\n",
       " 'Igf2',\n",
       " 'Igfbp2',\n",
       " 'Igfbp4',\n",
       " 'Igfbp6',\n",
       " 'Igfbp7',\n",
       " 'Igsf11',\n",
       " 'Igsf21',\n",
       " 'Igsf3',\n",
       " 'Il11ra1',\n",
       " 'Il18',\n",
       " 'Inafm1',\n",
       " 'Inpp5d',\n",
       " 'Inpp5f',\n",
       " 'Inpp5j',\n",
       " 'Ipcef1',\n",
       " 'Islr',\n",
       " 'Isoc1',\n",
       " 'Isyna1',\n",
       " 'Itga1',\n",
       " 'Itgb1',\n",
       " 'Itgb5',\n",
       " 'Itih5',\n",
       " 'Itm2a',\n",
       " 'Itm2b',\n",
       " 'Itpka',\n",
       " 'Ivns1abp',\n",
       " 'Jam2',\n",
       " 'Jam3',\n",
       " 'Jun',\n",
       " 'Junb',\n",
       " 'Kalrn',\n",
       " 'Kank3',\n",
       " 'Kcna6',\n",
       " 'Kcnab3',\n",
       " 'Kcnc1',\n",
       " 'Kcnc2',\n",
       " 'Kcnh6',\n",
       " 'Kcnh7',\n",
       " 'Kcnip1',\n",
       " 'Kcnip3',\n",
       " 'Kcnip4',\n",
       " 'Kcnj10',\n",
       " 'Kcnj8',\n",
       " 'Kcnk2',\n",
       " 'Kcnk3',\n",
       " 'Kcnk4',\n",
       " 'Kcnma1',\n",
       " 'Kcnmb2',\n",
       " 'Kcnmb4',\n",
       " 'Kcnn2',\n",
       " 'Kcns3',\n",
       " 'Kcnt2',\n",
       " 'Kcnv1',\n",
       " 'Kctd1',\n",
       " 'Kctd12',\n",
       " 'Kctd16',\n",
       " 'Kctd4',\n",
       " 'Kctd6',\n",
       " 'Kctd8',\n",
       " 'Kdr',\n",
       " 'Kit',\n",
       " 'Klf10',\n",
       " 'Klf2',\n",
       " 'Klf5',\n",
       " 'Klf9',\n",
       " 'Klhl13',\n",
       " 'Krt1',\n",
       " 'Krt12',\n",
       " 'LOC100861618',\n",
       " 'LOC101056001',\n",
       " 'LOC102632463',\n",
       " 'LOC102635661',\n",
       " 'LOC102637129',\n",
       " 'LOC105244017',\n",
       " 'LOC105246395',\n",
       " 'Lamb2',\n",
       " 'Lamc1',\n",
       " 'Lamp5',\n",
       " 'Lancl1',\n",
       " 'Laptm4a',\n",
       " 'Laptm4b',\n",
       " 'Laptm5',\n",
       " 'Lcat',\n",
       " 'Ldb2',\n",
       " 'Ldha',\n",
       " 'Ldhb',\n",
       " 'Lef1',\n",
       " 'Lgals1',\n",
       " 'Lgi2',\n",
       " 'Lgmn',\n",
       " 'Lgr5',\n",
       " 'Lgr6',\n",
       " 'Lhfp',\n",
       " 'Lhx1os',\n",
       " 'Lhx2',\n",
       " 'Lhx5',\n",
       " 'Lhx6',\n",
       " 'Limch1',\n",
       " 'Lims2',\n",
       " 'Lin7a',\n",
       " 'Lin7b',\n",
       " 'Lingo1',\n",
       " 'Litaf',\n",
       " 'Lmo2',\n",
       " 'Lmo3',\n",
       " 'Lmo4',\n",
       " 'Lmod1',\n",
       " 'Lpar1',\n",
       " 'Lpcat2',\n",
       " 'Lpgat1',\n",
       " 'Lppr1',\n",
       " 'Lrrc4c',\n",
       " 'Lrrtm4',\n",
       " 'Lsr',\n",
       " 'Ltbp4',\n",
       " 'Ltc4s',\n",
       " 'Lum',\n",
       " 'Lurap1l',\n",
       " 'Luzp2',\n",
       " 'Ly6a',\n",
       " 'Ly6c1',\n",
       " 'Ly6c2',\n",
       " 'Ly6e',\n",
       " 'Ly6e-ps1',\n",
       " 'Ly6h',\n",
       " 'Ly86',\n",
       " 'Lyn',\n",
       " 'Lypd1',\n",
       " 'Lypd6',\n",
       " 'Lypd6b',\n",
       " 'Lzts1',\n",
       " 'Maf',\n",
       " 'Mafb',\n",
       " 'Mag',\n",
       " 'Maged1',\n",
       " 'Magi1',\n",
       " 'Magi3',\n",
       " 'Mal',\n",
       " 'Mal2',\n",
       " 'Malat1',\n",
       " 'Man1c1',\n",
       " 'Map3k1',\n",
       " 'Mapk11',\n",
       " 'Mapk3',\n",
       " 'Mapk4',\n",
       " 'Marc2',\n",
       " 'Marcks',\n",
       " 'Marcksl1',\n",
       " 'Masp1',\n",
       " 'Mast3',\n",
       " 'Matn2',\n",
       " 'Mcam',\n",
       " 'Mctp1',\n",
       " 'Mdh1',\n",
       " 'Mdk',\n",
       " 'Me3',\n",
       " 'Med10',\n",
       " 'Mef2c',\n",
       " 'Meis1',\n",
       " 'Meis2',\n",
       " 'Mertk',\n",
       " 'Mfge8',\n",
       " 'Mgat4c',\n",
       " 'Mgat5',\n",
       " 'Mgll',\n",
       " 'Mgp',\n",
       " 'Mgst3',\n",
       " 'Mical2',\n",
       " 'Mir143hg',\n",
       " 'Mkl2',\n",
       " 'Mlc1',\n",
       " 'Mmd2',\n",
       " 'Mmp16',\n",
       " 'Mmp17',\n",
       " 'Mobp',\n",
       " 'Mog',\n",
       " 'Morf4l2',\n",
       " 'Mpc2',\n",
       " 'Mpi',\n",
       " 'Mpped1',\n",
       " 'Msi2',\n",
       " 'Msmo1',\n",
       " 'Msn',\n",
       " 'Mt1',\n",
       " 'Mt2',\n",
       " 'Mt3',\n",
       " 'Mtfp1',\n",
       " 'Mustn1',\n",
       " 'Myh11',\n",
       " 'Myh7',\n",
       " 'Myl12a',\n",
       " 'Myl4',\n",
       " 'Myl6',\n",
       " 'Myl9',\n",
       " 'Mylk',\n",
       " 'Myo16',\n",
       " 'Myo1b',\n",
       " 'Myrf',\n",
       " 'N28178',\n",
       " 'ND4',\n",
       " 'Naca',\n",
       " 'Nacc2',\n",
       " 'Nap1l5',\n",
       " 'Nat8',\n",
       " 'Nav3',\n",
       " 'Nbl1',\n",
       " 'Ncald',\n",
       " 'Ndn',\n",
       " 'Ndnf',\n",
       " 'Ndrg1',\n",
       " 'Ndrg2',\n",
       " 'Ndrg3',\n",
       " 'Ndufa12',\n",
       " 'Ndufa13',\n",
       " 'Ndufa4l2',\n",
       " 'Ndufa6',\n",
       " 'Necab1',\n",
       " 'Necab2',\n",
       " 'Nefl',\n",
       " 'Nefm',\n",
       " 'Nek7',\n",
       " 'Nell2',\n",
       " 'Neto1',\n",
       " 'Neto2',\n",
       " 'Neurod1',\n",
       " 'Neurod2',\n",
       " 'Neurod6',\n",
       " 'Nfia',\n",
       " 'Nfib',\n",
       " 'Nfix',\n",
       " 'Nfkbia',\n",
       " 'Ngef',\n",
       " 'Ngf',\n",
       " 'Ngfrap1',\n",
       " 'Nhlh2',\n",
       " 'Nipsnap3b',\n",
       " 'Nkain4',\n",
       " 'Nkx6-2',\n",
       " 'Nmnat2',\n",
       " 'Nog',\n",
       " 'Nos1ap',\n",
       " 'Nostrin',\n",
       " 'Notch3',\n",
       " 'Nov',\n",
       " 'Npas1',\n",
       " 'Npas3',\n",
       " 'Npc1',\n",
       " 'Npr3',\n",
       " 'Nptx1',\n",
       " 'Npy',\n",
       " 'Npy1r',\n",
       " 'Nr2f2',\n",
       " 'Nr3c2',\n",
       " 'Nrep',\n",
       " 'Nrgn',\n",
       " 'Nrip2',\n",
       " 'Nrip3',\n",
       " 'Nrn1',\n",
       " 'Nrp1',\n",
       " 'Nrsn2',\n",
       " 'Nrxn1',\n",
       " 'Nrxn3',\n",
       " 'Nsg2',\n",
       " 'Nt5dc2',\n",
       " 'Ntsr2',\n",
       " 'Nudt4',\n",
       " 'Nupr1',\n",
       " 'Nxph1',\n",
       " 'Nxph3',\n",
       " 'Nxph4',\n",
       " 'Nyap2',\n",
       " 'Ociad2',\n",
       " 'Ogfrl1',\n",
       " 'Ogn',\n",
       " 'Olfm1',\n",
       " 'Olfm3',\n",
       " 'Olfml3',\n",
       " 'Olig1',\n",
       " 'Olig2',\n",
       " 'Opalin',\n",
       " 'Oprm1',\n",
       " 'Osbp2',\n",
       " 'Osbpl1a',\n",
       " 'Osbpl8',\n",
       " 'P2ry12',\n",
       " 'P2ry13',\n",
       " 'P2ry14',\n",
       " 'P2ry6',\n",
       " 'P4ha1',\n",
       " 'Pacsin2',\n",
       " 'Pafah1b3',\n",
       " 'Paip2',\n",
       " 'Pak1',\n",
       " 'Palld',\n",
       " 'Pamr1',\n",
       " 'Paqr5',\n",
       " 'Parm1',\n",
       " 'Pax6',\n",
       " 'Pbx1',\n",
       " 'Pbx3',\n",
       " 'Pcbp3',\n",
       " 'Pcca',\n",
       " 'Pcdh17',\n",
       " 'Pcdh20',\n",
       " 'Pcolce',\n",
       " 'Pcp4',\n",
       " 'Pcp4l1',\n",
       " 'Pcsk2',\n",
       " 'Pcsk5',\n",
       " 'Pde11a',\n",
       " 'Pde1a',\n",
       " 'Pde3a',\n",
       " 'Pdgfra',\n",
       " 'Pdgfrb',\n",
       " 'Pdlim3',\n",
       " 'Pdp1',\n",
       " 'Pecam1',\n",
       " 'Penk',\n",
       " 'Pex5l',\n",
       " 'Pfn1',\n",
       " 'Pglyrp1',\n",
       " 'Pgm2l1',\n",
       " 'Pgrmc1',\n",
       " 'Phactr1',\n",
       " 'Phgdh',\n",
       " 'Phlda1',\n",
       " 'Phldb1',\n",
       " 'Phldb2',\n",
       " 'Phyhd1',\n",
       " 'Phyhip',\n",
       " 'Phyhipl',\n",
       " 'Pip5k1b',\n",
       " 'Pitpnc1',\n",
       " 'Pkig',\n",
       " 'Pla1a',\n",
       " 'Pla2g16',\n",
       " 'Pla2g7',\n",
       " 'Plat',\n",
       " 'Plcd4',\n",
       " 'Plcxd2',\n",
       " 'Pld4',\n",
       " 'Pld5',\n",
       " 'Plekha2',\n",
       " 'Plekhb1',\n",
       " 'Plk2',\n",
       " 'Pllp',\n",
       " 'Pln',\n",
       " 'Plp1',\n",
       " 'Pls3',\n",
       " 'Pltp',\n",
       " 'Plxnb3',\n",
       " 'Pnck',\n",
       " 'Pnoc',\n",
       " 'Podxl',\n",
       " 'Polr2h',\n",
       " 'Pou3f1',\n",
       " 'Ppap2b',\n",
       " 'Ppargc1a',\n",
       " 'Ppia',\n",
       " 'Ppp1r12a',\n",
       " 'Ppp1r14a',\n",
       " 'Ppp1r1b',\n",
       " 'Ppp3ca',\n",
       " 'Prdx2',\n",
       " 'Prdx6',\n",
       " 'Prelp',\n",
       " 'Prex2',\n",
       " 'Prkca',\n",
       " 'Prkcb',\n",
       " 'Prkcdbp',\n",
       " 'Prkcg',\n",
       " 'Prmt2',\n",
       " ...]"
      ]
     },
     "execution_count": 141,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "markers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 148,
   "id": "af64743b",
   "metadata": {},
   "outputs": [],
   "source": [
    "#concat marker dfs for each class to get markers_sd\n",
    "markers_df = pd.read_feather('/bigdata/isaac/df_marker_2024-05-19.feather')\n",
    "markers = sorted(markers_df.index)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 156,
   "id": "08632c94",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "262"
      ]
     },
     "execution_count": 156,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(markers)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fb033a2a",
   "metadata": {},
   "source": [
    "# Load subsetted visium spatial data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 207,
   "id": "978416a3",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_st_test_subset = sc.read_h5ad('/bigdata/isaac/tangram_integration/adata_st_test_subset.h5ad')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 208,
   "id": "c0594647",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 582 × 15091\n",
       "    obs: 'in_tissue', 'array_row', 'array_col', 'thing', 'n_genes_by_counts', 'log1p_n_genes_by_counts', 'total_counts', 'log1p_total_counts', 'pct_counts_in_top_50_genes', 'pct_counts_in_top_100_genes', 'pct_counts_in_top_200_genes', 'pct_counts_in_top_500_genes', 'total_counts_mt', 'log1p_total_counts_mt', 'pct_counts_mt', 'n_counts', 'clusters'\n",
       "    var: 'gene_ids', 'feature_types', 'genome', 'mt', 'n_cells_by_counts', 'mean_counts', 'log1p_mean_counts', 'pct_dropout_by_counts', 'total_counts', 'log1p_total_counts', 'n_cells', 'highly_variable', 'means', 'dispersions', 'dispersions_norm'\n",
       "    uns: 'spatial'\n",
       "    obsm: 'spatial'"
      ]
     },
     "execution_count": 208,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata_st_test_subset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 209,
   "id": "05577673",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "INFO:root:249 training genes are saved in `uns``training_genes` of both single cell and spatial Anndatas.\n",
      "INFO:root:14928 overlapped genes are saved in `uns``overlap_genes` of both single cell and spatial Anndatas.\n",
      "INFO:root:uniform based density prior is calculated and saved in `obs``uniform_density` of the spatial Anndata.\n",
      "INFO:root:rna count based density prior is calculated and saved in `obs``rna_count_based_density` of the spatial Anndata.\n"
     ]
    }
   ],
   "source": [
    "tg.pp_adatas(adata_sc_sd, adata_st_test_subset, genes=markers, gene_to_lowercase = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 210,
   "id": "70dd0521",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "INFO:root:Allocate tensors for mapping.\n",
      "INFO:root:Begin training with 249 genes and rna_count_based density_prior in clusters mode...\n",
      "INFO:root:Printing scores every 100 epochs.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Score: 0.313, KL reg: 0.030\n",
      "Score: 0.580, KL reg: 0.001\n",
      "Score: 0.581, KL reg: 0.001\n",
      "Score: 0.581, KL reg: 0.001\n",
      "Score: 0.581, KL reg: 0.001\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "INFO:root:Saving results..\n"
     ]
    }
   ],
   "source": [
    "ad_map = tg.map_cells_to_space(adata_sc_sd, adata_st_test_subset,\n",
    "    #mode=\"cells\",\n",
    "    mode=\"clusters\",\n",
    "    cluster_label='cell_subclass',  # .obs field w cell types\n",
    "    density_prior='rna_count_based',\n",
    "    num_epochs=500,\n",
    "    # device=\"cuda:0\",\n",
    "    device='cpu',\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 211,
   "id": "25b74db0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 103 × 582\n",
       "    obs: 'cell_subclass', 'cluster_density'\n",
       "    var: 'in_tissue', 'array_row', 'array_col', 'thing', 'n_genes_by_counts', 'log1p_n_genes_by_counts', 'total_counts', 'log1p_total_counts', 'pct_counts_in_top_50_genes', 'pct_counts_in_top_100_genes', 'pct_counts_in_top_200_genes', 'pct_counts_in_top_500_genes', 'total_counts_mt', 'log1p_total_counts_mt', 'pct_counts_mt', 'n_counts', 'clusters', 'uniform_density', 'rna_count_based_density'\n",
       "    uns: 'train_genes_df', 'training_history'"
      ]
     },
     "execution_count": 211,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ad_map"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 288,
   "id": "0bd37a10",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "INFO:root:spatial prediction dataframe is saved in `obsm` `tangram_ct_pred` of the spatial AnnData.\n"
     ]
    },
    {
     "ename": "ValueError",
     "evalue": "Key tmem100 could be found in both .var_names and .obs.columns",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mValueError\u001b[0m                                Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[288], line 3\u001b[0m\n\u001b[1;32m      1\u001b[0m tg\u001b[38;5;241m.\u001b[39mproject_cell_annotations(ad_map, adata_st_test_subset, annotation\u001b[38;5;241m=\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mcell_subclass\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n\u001b[1;32m      2\u001b[0m annotation_list \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mlist\u001b[39m(pd\u001b[38;5;241m.\u001b[39munique(adata_sc_sd\u001b[38;5;241m.\u001b[39mobs[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mcell_subclass\u001b[39m\u001b[38;5;124m'\u001b[39m]))\n\u001b[0;32m----> 3\u001b[0m tg\u001b[38;5;241m.\u001b[39mplot_cell_annotation_sc(adata_st_test_subset, annotation_list,perc\u001b[38;5;241m=\u001b[39m\u001b[38;5;241m0.02\u001b[39m)\n",
      "File \u001b[0;32m~/anaconda3/envs/scanpy_env/lib/python3.11/site-packages/tangram/plot_utils.py:208\u001b[0m, in \u001b[0;36mplot_cell_annotation_sc\u001b[0;34m(adata_sp, annotation_list, x, y, spot_size, scale_factor, perc, alpha_img, bw, ax)\u001b[0m\n\u001b[1;32m    205\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mspatial\u001b[39m\u001b[38;5;124m'\u001b[39m \u001b[38;5;129;01min\u001b[39;00m adata_sp\u001b[38;5;241m.\u001b[39muns\u001b[38;5;241m.\u001b[39mkeys() \u001b[38;5;129;01mand\u001b[39;00m spot_size \u001b[38;5;241m!=\u001b[39m \u001b[38;5;28;01mNone\u001b[39;00m \u001b[38;5;129;01mand\u001b[39;00m scale_factor \u001b[38;5;241m!=\u001b[39m \u001b[38;5;28;01mNone\u001b[39;00m:\n\u001b[1;32m    206\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mValueError\u001b[39;00m(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mSpot Size and Scale Factor should be None when ad_sp.uns[\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mspatial\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124m] exists\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n\u001b[0;32m--> 208\u001b[0m sc\u001b[38;5;241m.\u001b[39mpl\u001b[38;5;241m.\u001b[39mspatial(\n\u001b[1;32m    209\u001b[0m     adata_sp, color\u001b[38;5;241m=\u001b[39mannotation_list, cmap\u001b[38;5;241m=\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mviridis\u001b[39m\u001b[38;5;124m\"\u001b[39m, show\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mFalse\u001b[39;00m, frameon\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mFalse\u001b[39;00m, spot_size\u001b[38;5;241m=\u001b[39mspot_size,\n\u001b[1;32m    210\u001b[0m     scale_factor\u001b[38;5;241m=\u001b[39mscale_factor, alpha_img\u001b[38;5;241m=\u001b[39malpha_img, bw\u001b[38;5;241m=\u001b[39mbw, ax\u001b[38;5;241m=\u001b[39max\n\u001b[1;32m    211\u001b[0m )\n\u001b[1;32m    213\u001b[0m adata_sp\u001b[38;5;241m.\u001b[39mobs\u001b[38;5;241m.\u001b[39mdrop(annotation_list, inplace\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mTrue\u001b[39;00m, errors\u001b[38;5;241m=\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mignore\u001b[39m\u001b[38;5;124m\"\u001b[39m, axis\u001b[38;5;241m=\u001b[39m\u001b[38;5;241m1\u001b[39m)\n",
      "File \u001b[0;32m~/anaconda3/envs/scanpy_env/lib/python3.11/site-packages/scanpy/plotting/_tools/scatterplots.py:1023\u001b[0m, in \u001b[0;36mspatial\u001b[0;34m(adata, basis, img, img_key, library_id, crop_coord, alpha_img, bw, size, scale_factor, spot_size, na_color, show, return_fig, save, **kwargs)\u001b[0m\n\u001b[1;32m   1020\u001b[0m     cmap_img \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;01mNone\u001b[39;00m\n\u001b[1;32m   1021\u001b[0m circle_radius \u001b[38;5;241m=\u001b[39m size \u001b[38;5;241m*\u001b[39m scale_factor \u001b[38;5;241m*\u001b[39m spot_size \u001b[38;5;241m*\u001b[39m \u001b[38;5;241m0.5\u001b[39m\n\u001b[0;32m-> 1023\u001b[0m axs \u001b[38;5;241m=\u001b[39m embedding(\n\u001b[1;32m   1024\u001b[0m     adata,\n\u001b[1;32m   1025\u001b[0m     basis\u001b[38;5;241m=\u001b[39mbasis,\n\u001b[1;32m   1026\u001b[0m     scale_factor\u001b[38;5;241m=\u001b[39mscale_factor,\n\u001b[1;32m   1027\u001b[0m     size\u001b[38;5;241m=\u001b[39mcircle_radius,\n\u001b[1;32m   1028\u001b[0m     na_color\u001b[38;5;241m=\u001b[39mna_color,\n\u001b[1;32m   1029\u001b[0m     show\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mFalse\u001b[39;00m,\n\u001b[1;32m   1030\u001b[0m     save\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mFalse\u001b[39;00m,\n\u001b[1;32m   1031\u001b[0m     \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mkwargs,\n\u001b[1;32m   1032\u001b[0m )\n\u001b[1;32m   1033\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28misinstance\u001b[39m(axs, \u001b[38;5;28mlist\u001b[39m):\n\u001b[1;32m   1034\u001b[0m     axs \u001b[38;5;241m=\u001b[39m [axs]\n",
      "File \u001b[0;32m~/anaconda3/envs/scanpy_env/lib/python3.11/site-packages/scanpy/plotting/_tools/scatterplots.py:278\u001b[0m, in \u001b[0;36membedding\u001b[0;34m(adata, basis, color, mask_obs, gene_symbols, use_raw, sort_order, edges, edges_width, edges_color, neighbors_key, arrows, arrows_kwds, groups, components, dimensions, layer, projection, scale_factor, color_map, cmap, palette, na_color, na_in_legend, size, frameon, legend_fontsize, legend_fontweight, legend_loc, legend_fontoutline, colorbar_loc, vmax, vmin, vcenter, norm, add_outline, outline_width, outline_color, ncols, hspace, wspace, title, show, save, ax, return_fig, marker, **kwargs)\u001b[0m\n\u001b[1;32m    276\u001b[0m \u001b[38;5;28;01mfor\u001b[39;00m count, (value_to_plot, dims) \u001b[38;5;129;01min\u001b[39;00m \u001b[38;5;28menumerate\u001b[39m(\u001b[38;5;28mzip\u001b[39m(color, dimensions)):\n\u001b[1;32m    277\u001b[0m     kwargs_scatter \u001b[38;5;241m=\u001b[39m kwargs\u001b[38;5;241m.\u001b[39mcopy()  \u001b[38;5;66;03m# is potentially mutated for each plot\u001b[39;00m\n\u001b[0;32m--> 278\u001b[0m     color_source_vector \u001b[38;5;241m=\u001b[39m _get_color_source_vector(\n\u001b[1;32m    279\u001b[0m         adata,\n\u001b[1;32m    280\u001b[0m         value_to_plot,\n\u001b[1;32m    281\u001b[0m         layer\u001b[38;5;241m=\u001b[39mlayer,\n\u001b[1;32m    282\u001b[0m         mask_obs\u001b[38;5;241m=\u001b[39mmask_obs,\n\u001b[1;32m    283\u001b[0m         use_raw\u001b[38;5;241m=\u001b[39muse_raw,\n\u001b[1;32m    284\u001b[0m         gene_symbols\u001b[38;5;241m=\u001b[39mgene_symbols,\n\u001b[1;32m    285\u001b[0m         groups\u001b[38;5;241m=\u001b[39mgroups,\n\u001b[1;32m    286\u001b[0m     )\n\u001b[1;32m    287\u001b[0m     color_vector, color_type \u001b[38;5;241m=\u001b[39m _color_vector(\n\u001b[1;32m    288\u001b[0m         adata,\n\u001b[1;32m    289\u001b[0m         value_to_plot,\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m    292\u001b[0m         na_color\u001b[38;5;241m=\u001b[39mna_color,\n\u001b[1;32m    293\u001b[0m     )\n\u001b[1;32m    295\u001b[0m     \u001b[38;5;66;03m# Order points\u001b[39;00m\n",
      "File \u001b[0;32m~/anaconda3/envs/scanpy_env/lib/python3.11/site-packages/scanpy/plotting/_tools/scatterplots.py:1203\u001b[0m, in \u001b[0;36m_get_color_source_vector\u001b[0;34m(adata, value_to_plot, mask_obs, use_raw, gene_symbols, layer, groups)\u001b[0m\n\u001b[1;32m   1201\u001b[0m     values \u001b[38;5;241m=\u001b[39m adata\u001b[38;5;241m.\u001b[39mraw\u001b[38;5;241m.\u001b[39mobs_vector(value_to_plot)\n\u001b[1;32m   1202\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[0;32m-> 1203\u001b[0m     values \u001b[38;5;241m=\u001b[39m adata\u001b[38;5;241m.\u001b[39mobs_vector(value_to_plot, layer\u001b[38;5;241m=\u001b[39mlayer)\n\u001b[1;32m   1204\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m mask_obs \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m:\n\u001b[1;32m   1205\u001b[0m     values[\u001b[38;5;241m~\u001b[39mmask_obs] \u001b[38;5;241m=\u001b[39m np\u001b[38;5;241m.\u001b[39mnan\n",
      "File \u001b[0;32m~/anaconda3/envs/scanpy_env/lib/python3.11/site-packages/anndata/_core/anndata.py:1365\u001b[0m, in \u001b[0;36mAnnData.obs_vector\u001b[0;34m(self, k, layer)\u001b[0m\n\u001b[1;32m   1359\u001b[0m         warnings\u001b[38;5;241m.\u001b[39mwarn(\n\u001b[1;32m   1360\u001b[0m             \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mIn a future version of AnnData, access to `.X` by passing\u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[1;32m   1361\u001b[0m             \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124m `layer=\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mX\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124m` will be removed. Instead pass `layer=None`.\u001b[39m\u001b[38;5;124m\"\u001b[39m,\n\u001b[1;32m   1362\u001b[0m             \u001b[38;5;167;01mFutureWarning\u001b[39;00m,\n\u001b[1;32m   1363\u001b[0m         )\n\u001b[1;32m   1364\u001b[0m         layer \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;01mNone\u001b[39;00m\n\u001b[0;32m-> 1365\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m get_vector(\u001b[38;5;28mself\u001b[39m, k, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mobs\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mvar\u001b[39m\u001b[38;5;124m\"\u001b[39m, layer\u001b[38;5;241m=\u001b[39mlayer)\n",
      "File \u001b[0;32m~/anaconda3/envs/scanpy_env/lib/python3.11/site-packages/anndata/_core/index.py:214\u001b[0m, in \u001b[0;36mget_vector\u001b[0;34m(adata, k, coldim, idxdim, layer)\u001b[0m\n\u001b[1;32m    211\u001b[0m in_idx \u001b[38;5;241m=\u001b[39m k \u001b[38;5;129;01min\u001b[39;00m idx\n\u001b[1;32m    213\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m (in_col \u001b[38;5;241m+\u001b[39m in_idx) \u001b[38;5;241m==\u001b[39m \u001b[38;5;241m2\u001b[39m:\n\u001b[0;32m--> 214\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mValueError\u001b[39;00m(\n\u001b[1;32m    215\u001b[0m         \u001b[38;5;124mf\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mKey \u001b[39m\u001b[38;5;132;01m{\u001b[39;00mk\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m could be found in both .\u001b[39m\u001b[38;5;132;01m{\u001b[39;00midxdim\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m_names and .\u001b[39m\u001b[38;5;132;01m{\u001b[39;00mcoldim\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m.columns\u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[1;32m    216\u001b[0m     )\n\u001b[1;32m    217\u001b[0m \u001b[38;5;28;01melif\u001b[39;00m (in_col \u001b[38;5;241m+\u001b[39m in_idx) \u001b[38;5;241m==\u001b[39m \u001b[38;5;241m0\u001b[39m:\n\u001b[1;32m    218\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mKeyError\u001b[39;00m(\n\u001b[1;32m    219\u001b[0m         \u001b[38;5;124mf\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mCould not find key \u001b[39m\u001b[38;5;132;01m{\u001b[39;00mk\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m in .\u001b[39m\u001b[38;5;132;01m{\u001b[39;00midxdim\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m_names or .\u001b[39m\u001b[38;5;132;01m{\u001b[39;00mcoldim\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m.columns.\u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[1;32m    220\u001b[0m     )\n",
      "\u001b[0;31mValueError\u001b[0m: Key tmem100 could be found in both .var_names and .obs.columns"
     ]
    }
   ],
   "source": [
    "tg.project_cell_annotations(ad_map, adata_st_test_subset, annotation=\"cell_subclass\")\n",
    "annotation_list = list(pd.unique(adata_sc_sd.obs['cell_subclass']))\n",
    "tg.plot_cell_annotation_sc(adata_st_test_subset, annotation_list,perc=0.02)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 278,
   "id": "31966c40",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['chat-isl1',\n",
       " 'sncg-htr3a',\n",
       " 'vip-cbln2',\n",
       " 'cpne4-pde3a',\n",
       " 'reln-ndnf',\n",
       " 'dab1-nfib',\n",
       " 'kit-pdlim5',\n",
       " 'hcrtr2-id2',\n",
       " 'pthlh-pvalb',\n",
       " 'rbp4-kitl',\n",
       " 'moxd1-vwc2',\n",
       " 'maf-mafb',\n",
       " 'npy-sst',\n",
       " 'spon1-tox',\n",
       " 'gm28884-fam159b',\n",
       " 'st18-gm17660',\n",
       " 'calcr-peg10',\n",
       " 'prlr-pde1c',\n",
       " 'cbln1-grp',\n",
       " 'tmem100-',\n",
       " 'unc13c-crtac1',\n",
       " 'igfbp6-rprm',\n",
       " 'wfs1-prok2',\n",
       " 'jsrp1-col18a1',\n",
       " 'lpl-',\n",
       " 'rmst-mgat4c',\n",
       " 'oprk1-trhde',\n",
       " 'lmo1-chn2',\n",
       " 'gal-asb4',\n",
       " 'igsf1-a230065h16rik',\n",
       " 'foxp2-gpr88',\n",
       " 'npy2r-',\n",
       " 'nwd2-',\n",
       " 'lypd1',\n",
       " 'hs3st4-car10',\n",
       " 'bc039966-ngb',\n",
       " 'col23a1',\n",
       " 'col6a1',\n",
       " 'myh7-',\n",
       " 'htr1f-tshz1',\n",
       " 'col11a1-pde11a',\n",
       " 'pax6-npnt',\n",
       " 'calca',\n",
       " 'dsc3-mfge8',\n",
       " 'nrn1-slc17a7',\n",
       " 'pou3f2-zfhx3',\n",
       " 'cyp26b1-dach1',\n",
       " 'ptn-ndst4',\n",
       " 'megf11-gpc5',\n",
       " 'bc048546-dlx1',\n",
       " 'sim1-arhgdib',\n",
       " 'gm11549-prox1',\n",
       " 'il33-fbln1',\n",
       " 'cd44-cplx3',\n",
       " 'dcn-arhgap6',\n",
       " '2410022m11rik-cd24a',\n",
       " 'pde11a-vgll3',\n",
       " 'lypd6-',\n",
       " 'bdnf-pcdh11x',\n",
       " 'igfbp4-mgp',\n",
       " 'cpne2-adgrg6',\n",
       " 'tac1-angpt1',\n",
       " 'sema5a-medag',\n",
       " 'snca-foxp2',\n",
       " 'spon1-tmem215',\n",
       " 'rxfp1-fam3c',\n",
       " 'moxd1-',\n",
       " 'nppc-plk5',\n",
       " 'rnf152-ecel1',\n",
       " 'rab3b-',\n",
       " 'reln-bc048546',\n",
       " 'gfod2-',\n",
       " 'foxb1-bc048546',\n",
       " 'otp-sim1',\n",
       " 'zfhx3-gabrq',\n",
       " 'lpl-zic5',\n",
       " 'pgr15l-',\n",
       " 'grp-tcf4',\n",
       " 'tshz1-noa1',\n",
       " 'nr4a2-zfhx4',\n",
       " 'h2-q2-medag',\n",
       " 'ebf1-',\n",
       " 'hmbs-cd27',\n",
       " 'lamb3-col18a1',\n",
       " 'cartpt-tac1',\n",
       " 'bhlhe22-thrsp',\n",
       " 'mylk-lmo3',\n",
       " 'lmo1',\n",
       " 'gpr88-sparcl1',\n",
       " 'foxp2-sema3c',\n",
       " 'ndnf-sox6',\n",
       " 'pbx3-4930523c07rik',\n",
       " 'epend',\n",
       " 'proliferating_cells',\n",
       " 'microglia_1',\n",
       " 'microglia_2',\n",
       " 'peri_1',\n",
       " 'peri_2',\n",
       " 'vsm',\n",
       " 'opc',\n",
       " 'astro',\n",
       " 'ol_1',\n",
       " 'ol_2']"
      ]
     },
     "execution_count": 278,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "annotation_list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 228,
   "id": "8eb9a244",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(array([38]),)"
      ]
     },
     "execution_count": 228,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.where(adata_st_test_subset.obs.columns == 'tmem100')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 231,
   "id": "82bd24d1",
   "metadata": {},
   "outputs": [
    {
     "ename": "TypeError",
     "evalue": "Index does not support mutable operations",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mTypeError\u001b[0m                                 Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[231], line 1\u001b[0m\n\u001b[0;32m----> 1\u001b[0m adata_st_test_subset\u001b[38;5;241m.\u001b[39mobs\u001b[38;5;241m.\u001b[39mcolumns[\u001b[38;5;241m38\u001b[39m] \u001b[38;5;241m=\u001b[39m \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mtmem100-\u001b[39m\u001b[38;5;124m'\u001b[39m\n",
      "File \u001b[0;32m~/anaconda3/envs/scanpy_env/lib/python3.11/site-packages/pandas/core/indexes/base.py:5371\u001b[0m, in \u001b[0;36mIndex.__setitem__\u001b[0;34m(self, key, value)\u001b[0m\n\u001b[1;32m   5369\u001b[0m \u001b[38;5;129m@final\u001b[39m\n\u001b[1;32m   5370\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21m__setitem__\u001b[39m(\u001b[38;5;28mself\u001b[39m, key, value) \u001b[38;5;241m-\u001b[39m\u001b[38;5;241m>\u001b[39m \u001b[38;5;28;01mNone\u001b[39;00m:\n\u001b[0;32m-> 5371\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mTypeError\u001b[39;00m(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mIndex does not support mutable operations\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n",
      "\u001b[0;31mTypeError\u001b[0m: Index does not support mutable operations"
     ]
    }
   ],
   "source": [
    "adata_st_test_subset.obs.columns[38] = 'tmem100-'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 219,
   "id": "53042a8f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['in_tissue',\n",
       " 'array_row',\n",
       " 'array_col',\n",
       " 'thing',\n",
       " 'n_genes_by_counts',\n",
       " 'log1p_n_genes_by_counts',\n",
       " 'total_counts',\n",
       " 'log1p_total_counts',\n",
       " 'pct_counts_in_top_50_genes',\n",
       " 'pct_counts_in_top_100_genes',\n",
       " 'pct_counts_in_top_200_genes',\n",
       " 'pct_counts_in_top_500_genes',\n",
       " 'total_counts_mt',\n",
       " 'log1p_total_counts_mt',\n",
       " 'pct_counts_mt',\n",
       " 'n_counts',\n",
       " 'clusters',\n",
       " 'uniform_density',\n",
       " 'rna_count_based_density',\n",
       " 'chat-isl1',\n",
       " 'sncg-htr3a',\n",
       " 'vip-cbln2',\n",
       " 'cpne4-pde3a',\n",
       " 'reln-ndnf',\n",
       " 'dab1-nfib',\n",
       " 'kit-pdlim5',\n",
       " 'hcrtr2-id2',\n",
       " 'pthlh-pvalb',\n",
       " 'rbp4-kitl',\n",
       " 'moxd1-vwc2',\n",
       " 'maf-mafb',\n",
       " 'npy-sst',\n",
       " 'spon1-tox',\n",
       " 'gm28884-fam159b',\n",
       " 'st18-gm17660',\n",
       " 'calcr-peg10',\n",
       " 'prlr-pde1c',\n",
       " 'cbln1-grp',\n",
       " 'tmem100',\n",
       " 'unc13c-crtac1',\n",
       " 'igfbp6-rprm',\n",
       " 'wfs1-prok2',\n",
       " 'jsrp1-col18a1',\n",
       " 'lpl',\n",
       " 'rmst-mgat4c',\n",
       " 'oprk1-trhde',\n",
       " 'lmo1-chn2',\n",
       " 'gal-asb4',\n",
       " 'igsf1-a230065h16rik',\n",
       " 'foxp2-gpr88',\n",
       " 'npy2r',\n",
       " 'nwd2',\n",
       " 'lypd1',\n",
       " 'hs3st4-car10',\n",
       " 'bc039966-ngb',\n",
       " 'col23a1',\n",
       " 'col6a1',\n",
       " 'myh7',\n",
       " 'htr1f-tshz1',\n",
       " 'col11a1-pde11a',\n",
       " 'pax6-npnt',\n",
       " 'calca',\n",
       " 'dsc3-mfge8',\n",
       " 'nrn1-slc17a7',\n",
       " 'pou3f2-zfhx3',\n",
       " 'cyp26b1-dach1',\n",
       " 'ptn-ndst4',\n",
       " 'megf11-gpc5',\n",
       " 'bc048546-dlx1',\n",
       " 'sim1-arhgdib',\n",
       " 'gm11549-prox1',\n",
       " 'il33-fbln1',\n",
       " 'cd44-cplx3',\n",
       " 'dcn-arhgap6',\n",
       " '2410022m11rik-cd24a',\n",
       " 'pde11a-vgll3',\n",
       " 'lypd6',\n",
       " 'bdnf-pcdh11x',\n",
       " 'igfbp4-mgp',\n",
       " 'cpne2-adgrg6',\n",
       " 'tac1-angpt1',\n",
       " 'sema5a-medag',\n",
       " 'snca-foxp2',\n",
       " 'spon1-tmem215',\n",
       " 'rxfp1-fam3c',\n",
       " 'moxd1',\n",
       " 'nppc-plk5',\n",
       " 'rnf152-ecel1',\n",
       " 'rab3b',\n",
       " 'reln-bc048546',\n",
       " 'gfod2',\n",
       " 'foxb1-bc048546',\n",
       " 'otp-sim1',\n",
       " 'zfhx3-gabrq',\n",
       " 'lpl-zic5',\n",
       " 'pgr15l',\n",
       " 'grp-tcf4',\n",
       " 'tshz1-noa1',\n",
       " 'nr4a2-zfhx4',\n",
       " 'h2-q2-medag',\n",
       " 'ebf1',\n",
       " 'hmbs-cd27',\n",
       " 'lamb3-col18a1',\n",
       " 'cartpt-tac1',\n",
       " 'bhlhe22-thrsp',\n",
       " 'mylk-lmo3',\n",
       " 'lmo1',\n",
       " 'gpr88-sparcl1',\n",
       " 'foxp2-sema3c',\n",
       " 'ndnf-sox6',\n",
       " 'pbx3-4930523c07rik',\n",
       " 'epend',\n",
       " 'proliferating_cells',\n",
       " 'microglia_1',\n",
       " 'microglia_2',\n",
       " 'peri_1',\n",
       " 'peri_2',\n",
       " 'vsm',\n",
       " 'opc',\n",
       " 'astro',\n",
       " 'ol_1',\n",
       " 'ol_2']"
      ]
     },
     "execution_count": 219,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "list(adata_st_test_subset.obs.columns)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 214,
   "id": "aef5dacb",
   "metadata": {},
   "outputs": [],
   "source": [
    "tg.plot_training_scores(ad_map, bins=20, alpha=.5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 279,
   "id": "c57c8b58",
   "metadata": {},
   "outputs": [
    {
     "ename": "KeyError",
     "evalue": "'Could not find key chat-isl1 in .var_names or .obs.columns.'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[279], line 1\u001b[0m\n\u001b[0;32m----> 1\u001b[0m sc\u001b[38;5;241m.\u001b[39mpl\u001b[38;5;241m.\u001b[39mspatial(\n\u001b[1;32m      2\u001b[0m         adata_st_test_subset, color\u001b[38;5;241m=\u001b[39mannotation_list, cmap\u001b[38;5;241m=\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mviridis\u001b[39m\u001b[38;5;124m\"\u001b[39m, show\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mFalse\u001b[39;00m, frameon\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mFalse\u001b[39;00m, spot_size\u001b[38;5;241m=\u001b[39m\u001b[38;5;241m50\u001b[39m,\n\u001b[1;32m      3\u001b[0m         scale_factor\u001b[38;5;241m=\u001b[39m\u001b[38;5;241m.32\u001b[39m, alpha_img\u001b[38;5;241m=\u001b[39m\u001b[38;5;241m1.0\u001b[39m, bw\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mFalse\u001b[39;00m, ax\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mNone\u001b[39;00m\n\u001b[1;32m      4\u001b[0m     )\n",
      "File \u001b[0;32m~/anaconda3/envs/scanpy_env/lib/python3.11/site-packages/scanpy/plotting/_tools/scatterplots.py:1023\u001b[0m, in \u001b[0;36mspatial\u001b[0;34m(adata, basis, img, img_key, library_id, crop_coord, alpha_img, bw, size, scale_factor, spot_size, na_color, show, return_fig, save, **kwargs)\u001b[0m\n\u001b[1;32m   1020\u001b[0m     cmap_img \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;01mNone\u001b[39;00m\n\u001b[1;32m   1021\u001b[0m circle_radius \u001b[38;5;241m=\u001b[39m size \u001b[38;5;241m*\u001b[39m scale_factor \u001b[38;5;241m*\u001b[39m spot_size \u001b[38;5;241m*\u001b[39m \u001b[38;5;241m0.5\u001b[39m\n\u001b[0;32m-> 1023\u001b[0m axs \u001b[38;5;241m=\u001b[39m embedding(\n\u001b[1;32m   1024\u001b[0m     adata,\n\u001b[1;32m   1025\u001b[0m     basis\u001b[38;5;241m=\u001b[39mbasis,\n\u001b[1;32m   1026\u001b[0m     scale_factor\u001b[38;5;241m=\u001b[39mscale_factor,\n\u001b[1;32m   1027\u001b[0m     size\u001b[38;5;241m=\u001b[39mcircle_radius,\n\u001b[1;32m   1028\u001b[0m     na_color\u001b[38;5;241m=\u001b[39mna_color,\n\u001b[1;32m   1029\u001b[0m     show\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mFalse\u001b[39;00m,\n\u001b[1;32m   1030\u001b[0m     save\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mFalse\u001b[39;00m,\n\u001b[1;32m   1031\u001b[0m     \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mkwargs,\n\u001b[1;32m   1032\u001b[0m )\n\u001b[1;32m   1033\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28misinstance\u001b[39m(axs, \u001b[38;5;28mlist\u001b[39m):\n\u001b[1;32m   1034\u001b[0m     axs \u001b[38;5;241m=\u001b[39m [axs]\n",
      "File \u001b[0;32m~/anaconda3/envs/scanpy_env/lib/python3.11/site-packages/scanpy/plotting/_tools/scatterplots.py:278\u001b[0m, in \u001b[0;36membedding\u001b[0;34m(adata, basis, color, mask_obs, gene_symbols, use_raw, sort_order, edges, edges_width, edges_color, neighbors_key, arrows, arrows_kwds, groups, components, dimensions, layer, projection, scale_factor, color_map, cmap, palette, na_color, na_in_legend, size, frameon, legend_fontsize, legend_fontweight, legend_loc, legend_fontoutline, colorbar_loc, vmax, vmin, vcenter, norm, add_outline, outline_width, outline_color, ncols, hspace, wspace, title, show, save, ax, return_fig, marker, **kwargs)\u001b[0m\n\u001b[1;32m    276\u001b[0m \u001b[38;5;28;01mfor\u001b[39;00m count, (value_to_plot, dims) \u001b[38;5;129;01min\u001b[39;00m \u001b[38;5;28menumerate\u001b[39m(\u001b[38;5;28mzip\u001b[39m(color, dimensions)):\n\u001b[1;32m    277\u001b[0m     kwargs_scatter \u001b[38;5;241m=\u001b[39m kwargs\u001b[38;5;241m.\u001b[39mcopy()  \u001b[38;5;66;03m# is potentially mutated for each plot\u001b[39;00m\n\u001b[0;32m--> 278\u001b[0m     color_source_vector \u001b[38;5;241m=\u001b[39m _get_color_source_vector(\n\u001b[1;32m    279\u001b[0m         adata,\n\u001b[1;32m    280\u001b[0m         value_to_plot,\n\u001b[1;32m    281\u001b[0m         layer\u001b[38;5;241m=\u001b[39mlayer,\n\u001b[1;32m    282\u001b[0m         mask_obs\u001b[38;5;241m=\u001b[39mmask_obs,\n\u001b[1;32m    283\u001b[0m         use_raw\u001b[38;5;241m=\u001b[39muse_raw,\n\u001b[1;32m    284\u001b[0m         gene_symbols\u001b[38;5;241m=\u001b[39mgene_symbols,\n\u001b[1;32m    285\u001b[0m         groups\u001b[38;5;241m=\u001b[39mgroups,\n\u001b[1;32m    286\u001b[0m     )\n\u001b[1;32m    287\u001b[0m     color_vector, color_type \u001b[38;5;241m=\u001b[39m _color_vector(\n\u001b[1;32m    288\u001b[0m         adata,\n\u001b[1;32m    289\u001b[0m         value_to_plot,\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m    292\u001b[0m         na_color\u001b[38;5;241m=\u001b[39mna_color,\n\u001b[1;32m    293\u001b[0m     )\n\u001b[1;32m    295\u001b[0m     \u001b[38;5;66;03m# Order points\u001b[39;00m\n",
      "File \u001b[0;32m~/anaconda3/envs/scanpy_env/lib/python3.11/site-packages/scanpy/plotting/_tools/scatterplots.py:1203\u001b[0m, in \u001b[0;36m_get_color_source_vector\u001b[0;34m(adata, value_to_plot, mask_obs, use_raw, gene_symbols, layer, groups)\u001b[0m\n\u001b[1;32m   1201\u001b[0m     values \u001b[38;5;241m=\u001b[39m adata\u001b[38;5;241m.\u001b[39mraw\u001b[38;5;241m.\u001b[39mobs_vector(value_to_plot)\n\u001b[1;32m   1202\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[0;32m-> 1203\u001b[0m     values \u001b[38;5;241m=\u001b[39m adata\u001b[38;5;241m.\u001b[39mobs_vector(value_to_plot, layer\u001b[38;5;241m=\u001b[39mlayer)\n\u001b[1;32m   1204\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m mask_obs \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m:\n\u001b[1;32m   1205\u001b[0m     values[\u001b[38;5;241m~\u001b[39mmask_obs] \u001b[38;5;241m=\u001b[39m np\u001b[38;5;241m.\u001b[39mnan\n",
      "File \u001b[0;32m~/anaconda3/envs/scanpy_env/lib/python3.11/site-packages/anndata/_core/anndata.py:1365\u001b[0m, in \u001b[0;36mAnnData.obs_vector\u001b[0;34m(self, k, layer)\u001b[0m\n\u001b[1;32m   1359\u001b[0m         warnings\u001b[38;5;241m.\u001b[39mwarn(\n\u001b[1;32m   1360\u001b[0m             \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mIn a future version of AnnData, access to `.X` by passing\u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[1;32m   1361\u001b[0m             \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124m `layer=\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mX\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124m` will be removed. Instead pass `layer=None`.\u001b[39m\u001b[38;5;124m\"\u001b[39m,\n\u001b[1;32m   1362\u001b[0m             \u001b[38;5;167;01mFutureWarning\u001b[39;00m,\n\u001b[1;32m   1363\u001b[0m         )\n\u001b[1;32m   1364\u001b[0m         layer \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;01mNone\u001b[39;00m\n\u001b[0;32m-> 1365\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m get_vector(\u001b[38;5;28mself\u001b[39m, k, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mobs\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mvar\u001b[39m\u001b[38;5;124m\"\u001b[39m, layer\u001b[38;5;241m=\u001b[39mlayer)\n",
      "File \u001b[0;32m~/anaconda3/envs/scanpy_env/lib/python3.11/site-packages/anndata/_core/index.py:218\u001b[0m, in \u001b[0;36mget_vector\u001b[0;34m(adata, k, coldim, idxdim, layer)\u001b[0m\n\u001b[1;32m    214\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mValueError\u001b[39;00m(\n\u001b[1;32m    215\u001b[0m         \u001b[38;5;124mf\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mKey \u001b[39m\u001b[38;5;132;01m{\u001b[39;00mk\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m could be found in both .\u001b[39m\u001b[38;5;132;01m{\u001b[39;00midxdim\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m_names and .\u001b[39m\u001b[38;5;132;01m{\u001b[39;00mcoldim\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m.columns\u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[1;32m    216\u001b[0m     )\n\u001b[1;32m    217\u001b[0m \u001b[38;5;28;01melif\u001b[39;00m (in_col \u001b[38;5;241m+\u001b[39m in_idx) \u001b[38;5;241m==\u001b[39m \u001b[38;5;241m0\u001b[39m:\n\u001b[0;32m--> 218\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mKeyError\u001b[39;00m(\n\u001b[1;32m    219\u001b[0m         \u001b[38;5;124mf\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mCould not find key \u001b[39m\u001b[38;5;132;01m{\u001b[39;00mk\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m in .\u001b[39m\u001b[38;5;132;01m{\u001b[39;00midxdim\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m_names or .\u001b[39m\u001b[38;5;132;01m{\u001b[39;00mcoldim\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m.columns.\u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[1;32m    220\u001b[0m     )\n\u001b[1;32m    221\u001b[0m \u001b[38;5;28;01melif\u001b[39;00m in_col:\n\u001b[1;32m    222\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28mgetattr\u001b[39m(adata, coldim)[k]\u001b[38;5;241m.\u001b[39mvalues\n",
      "\u001b[0;31mKeyError\u001b[0m: 'Could not find key chat-isl1 in .var_names or .obs.columns.'"
     ]
    }
   ],
   "source": [
    "sc.pl.spatial(\n",
    "        adata_st_test_subset, color=annotation_list, cmap=\"viridis\", show=False, frameon=False, spot_size=50,\n",
    "        scale_factor=.32, alpha_img=1.0, bw=False, ax=None\n",
    "    )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 216,
   "id": "86709ea6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['in_tissue',\n",
       " 'array_row',\n",
       " 'array_col',\n",
       " 'thing',\n",
       " 'n_genes_by_counts',\n",
       " 'log1p_n_genes_by_counts',\n",
       " 'total_counts',\n",
       " 'log1p_total_counts',\n",
       " 'pct_counts_in_top_50_genes',\n",
       " 'pct_counts_in_top_100_genes',\n",
       " 'pct_counts_in_top_200_genes',\n",
       " 'pct_counts_in_top_500_genes',\n",
       " 'total_counts_mt',\n",
       " 'log1p_total_counts_mt',\n",
       " 'pct_counts_mt',\n",
       " 'n_counts',\n",
       " 'clusters',\n",
       " 'uniform_density',\n",
       " 'rna_count_based_density',\n",
       " 'chat-isl1',\n",
       " 'sncg-htr3a',\n",
       " 'vip-cbln2',\n",
       " 'cpne4-pde3a',\n",
       " 'reln-ndnf',\n",
       " 'dab1-nfib',\n",
       " 'kit-pdlim5',\n",
       " 'hcrtr2-id2',\n",
       " 'pthlh-pvalb',\n",
       " 'rbp4-kitl',\n",
       " 'moxd1-vwc2',\n",
       " 'maf-mafb',\n",
       " 'npy-sst',\n",
       " 'spon1-tox',\n",
       " 'gm28884-fam159b',\n",
       " 'st18-gm17660',\n",
       " 'calcr-peg10',\n",
       " 'prlr-pde1c',\n",
       " 'cbln1-grp',\n",
       " 'tmem100',\n",
       " 'unc13c-crtac1',\n",
       " 'igfbp6-rprm',\n",
       " 'wfs1-prok2',\n",
       " 'jsrp1-col18a1',\n",
       " 'lpl',\n",
       " 'rmst-mgat4c',\n",
       " 'oprk1-trhde',\n",
       " 'lmo1-chn2',\n",
       " 'gal-asb4',\n",
       " 'igsf1-a230065h16rik',\n",
       " 'foxp2-gpr88',\n",
       " 'npy2r',\n",
       " 'nwd2',\n",
       " 'lypd1',\n",
       " 'hs3st4-car10',\n",
       " 'bc039966-ngb',\n",
       " 'col23a1',\n",
       " 'col6a1',\n",
       " 'myh7',\n",
       " 'htr1f-tshz1',\n",
       " 'col11a1-pde11a',\n",
       " 'pax6-npnt',\n",
       " 'calca',\n",
       " 'dsc3-mfge8',\n",
       " 'nrn1-slc17a7',\n",
       " 'pou3f2-zfhx3',\n",
       " 'cyp26b1-dach1',\n",
       " 'ptn-ndst4',\n",
       " 'megf11-gpc5',\n",
       " 'bc048546-dlx1',\n",
       " 'sim1-arhgdib',\n",
       " 'gm11549-prox1',\n",
       " 'il33-fbln1',\n",
       " 'cd44-cplx3',\n",
       " 'dcn-arhgap6',\n",
       " '2410022m11rik-cd24a',\n",
       " 'pde11a-vgll3',\n",
       " 'lypd6',\n",
       " 'bdnf-pcdh11x',\n",
       " 'igfbp4-mgp',\n",
       " 'cpne2-adgrg6',\n",
       " 'tac1-angpt1',\n",
       " 'sema5a-medag',\n",
       " 'snca-foxp2',\n",
       " 'spon1-tmem215',\n",
       " 'rxfp1-fam3c',\n",
       " 'moxd1',\n",
       " 'nppc-plk5',\n",
       " 'rnf152-ecel1',\n",
       " 'rab3b',\n",
       " 'reln-bc048546',\n",
       " 'gfod2',\n",
       " 'foxb1-bc048546',\n",
       " 'otp-sim1',\n",
       " 'zfhx3-gabrq',\n",
       " 'lpl-zic5',\n",
       " 'pgr15l',\n",
       " 'grp-tcf4',\n",
       " 'tshz1-noa1',\n",
       " 'nr4a2-zfhx4',\n",
       " 'h2-q2-medag',\n",
       " 'ebf1',\n",
       " 'hmbs-cd27',\n",
       " 'lamb3-col18a1',\n",
       " 'cartpt-tac1',\n",
       " 'bhlhe22-thrsp',\n",
       " 'mylk-lmo3',\n",
       " 'lmo1',\n",
       " 'gpr88-sparcl1',\n",
       " 'foxp2-sema3c',\n",
       " 'ndnf-sox6',\n",
       " 'pbx3-4930523c07rik',\n",
       " 'epend',\n",
       " 'proliferating_cells',\n",
       " 'microglia_1',\n",
       " 'microglia_2',\n",
       " 'peri_1',\n",
       " 'peri_2',\n",
       " 'vsm',\n",
       " 'opc',\n",
       " 'astro',\n",
       " 'ol_1',\n",
       " 'ol_2']"
      ]
     },
     "execution_count": 216,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "list(adata_st_test_subset.obs.columns)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "302b4c50",
   "metadata": {},
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "967f8b1f",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7934bad6",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 182,
   "id": "2b68f602",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "INFO:root:257 training genes are saved in `uns``training_genes` of both single cell and spatial Anndatas.\n",
      "INFO:root:15080 overlapped genes are saved in `uns``overlap_genes` of both single cell and spatial Anndatas.\n",
      "INFO:root:uniform based density prior is calculated and saved in `obs``uniform_density` of the spatial Anndata.\n",
      "INFO:root:rna count based density prior is calculated and saved in `obs``rna_count_based_density` of the spatial Anndata.\n",
      "INFO:root:Allocate tensors for mapping.\n",
      "INFO:root:Begin training with 257 genes and rna_count_based density_prior in clusters mode...\n",
      "INFO:root:Printing scores every 100 epochs.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Score: 0.300, KL reg: 0.026\n",
      "Score: 0.576, KL reg: 0.001\n",
      "Score: 0.577, KL reg: 0.001\n",
      "Score: 0.578, KL reg: 0.001\n",
      "Score: 0.578, KL reg: 0.001\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "INFO:root:Saving results..\n",
      "INFO:root:spatial prediction dataframe is saved in `obsm` `tangram_ct_pred` of the spatial AnnData.\n"
     ]
    }
   ],
   "source": [
    "#repeat but with full spatial data...\n",
    "adata_st_test = sc.read_h5ad('/bigdata/isaac/tangram_integration/adata_st_test.h5ad')\n",
    "tg.pp_adatas(adata_sc_sd, adata_st_test, genes=markers)\n",
    "ad_map = tg.map_cells_to_space(adata_sc_sd, adata_st_test,\n",
    "    #mode=\"cells\",\n",
    "    mode=\"clusters\",\n",
    "    cluster_label='cell_subclass',  # .obs field w cell types\n",
    "    density_prior='rna_count_based',\n",
    "    num_epochs=500,\n",
    "    # device=\"cuda:0\",\n",
    "    device='cpu',\n",
    ")\n",
    "tg.project_cell_annotations(ad_map, adata_st_test, annotation=\"cell_subclass\")\n",
    "annotation_list = list(pd.unique(adata_sc_sd.obs['cell_subclass']))\n",
    "tg.plot_cell_annotation_sc(adata_st_test, annotation_list,perc=0.02)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2f22e8cf",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
